// @0no-co/graphqlsp/dist/graphqlsp.js
var e = require("./chunks/api-chunk.js");

var t = require("node:fs/promises");

var n = require("path");

var r = require("@gql.tada/internal");

var i = require("graphql");

var statFile = (e, n) => t.stat(e).then(n).catch(() => !1);

var swapWrite = async (e, n) => {
  if (!await statFile(e, e => e.isFile())) {
    await t.writeFile(e, n);
  } else {
    var r = e + ".tmp";
    await t.writeFile(r, n);
    try {
      await t.rename(r, e);
    } catch (e) {
      await t.unlink(r);
      throw e;
    } finally {
      await (async e => {
        try {
          var n = new Date;
          await t.utimes(e, n, n);
        } catch (e) {}
      })(e);
    }
  }
};

var toErrorMessage = e => e instanceof Error ? e.message : `${e}`;

var a = {};

var loadSchema = (e, i) => {
  var o = {
    config: null,
    load: new Map,
    write: new Map
  };
  var s = new Map;
  var u = null;
  var c = null;
  var l = 0;
  var d = {
    errors: o,
    outputLocations: s,
    get current() {
      return u && u.current;
    },
    get multi() {
      return u ? u.multi : a;
    },
    get version() {
      return u ? u.version : 0;
    },
    checkStale() {
      var e = Date.now();
      if (!c || e - l < 5e3) {
        return;
      }
      l = e;
      c();
    }
  };
  (async () => {
    var a = e.project.getProjectName();
    var l = await r.resolveTypeScriptRootDir(a) || n.dirname(a);
    i("Got root-directory to resolve schema from: " + l);
    var d;
    try {
      d = r.parseConfig(e.config, l);
    } catch (e) {
      o.config = `${toErrorMessage(e)}. Update the GraphQLSP plugin's entry in "compilerOptions.plugins" in your tsconfig.json.`;
      i(`Found invalid configuration: ${e}`);
      return;
    }
    var f = e.config.tadaDisablePreprocessing ?? !1;
    i('Resolving schema from "schema" config: ' + JSON.stringify(d));
    u = r.loadRef(d);
    var setLoadError = (e, t) => {
      o.load.set(e || null, `Failed to load the GraphQL schema${e ? ` "${e}"` : ""}: ${toErrorMessage(t)}`);
    };
    var persistSchema = e => {
      if (e && e.tadaOutputLocation) {
        !async function saveTadaIntrospection(e, i, a, o, s, u) {
          try {
            var c = i;
            if (await statFile(c, e => e.isDirectory())) {
              c = n.join(c, "introspection.d.ts");
            } else if (!await statFile(c, e => !!e)) {
              await t.mkdir(n.dirname(c), {
                recursive: !0
              });
              if (await statFile(c, e => e.isDirectory())) {
                c = n.join(c, "introspection.d.ts");
              }
            }
            var l = await t.readFile(c, "utf8").catch(() => {});
            var d = r.minifyIntrospection(e);
            var f = r.outputIntrospectionFile(d, {
              fileType: i,
              shouldPreprocess: !a,
              preamble: l ? r.extractIntrospectionHeader(l) || void 0 : void 0
            });
            await swapWrite(c, f);
            o.write.delete(i);
            s.set(c, Date.now());
            u(`Introspection saved to path @ ${c}`);
          } catch (e) {
            o.write.set(i, `Failed to write the typings file to "${i}": ${toErrorMessage(e)}`);
            u(`Failed to write introspection @ ${e}`);
          }
        }(e.introspection, n.resolve(l, e.tadaOutputLocation), f, o, s, i);
      }
    };
    var persistAll = () => {
      if (u.current) {
        persistSchema(u.current);
      }
      for (var e in u.multi) {
        persistSchema(u.multi[e]);
      }
    };
    var v = ("schemas" in d ? d.schemas : [ d ]).map(e => e.schema).filter(e => "string" == typeof e && !r.getURLConfig(e)).map(e => n.resolve(l, e));
    var p = new Map;
    var recordMtimes = async () => {
      await Promise.all(v.map(async e => {
        try {
          p.set(e, (await t.stat(e)).mtimeMs);
        } catch (e) {}
      }));
    };
    var revalidate = e => u.load({
      rootPath: l,
      reload: e
    }).then(() => {
      o.load.clear();
    }).catch(e => {
      o.load.clear();
      setLoadError(null, e);
    }).then(recordMtimes);
    c = () => {
      (async () => {
        var e = !1;
        for (var n of v) {
          try {
            var r = p.get(n);
            var a = (await t.stat(n)).mtimeMs;
            if (void 0 !== r && a > r) {
              e = !0;
            }
          } catch (e) {}
        }
        if (!e) {
          return;
        }
        i("Schema files changed without watcher events; reloading...");
        await revalidate(!0);
        persistAll();
      })().catch(e => {
        i(`Failed to check schemas for staleness: ${e}`);
      });
    };
    try {
      i("Loading schema...");
      await u.load({
        rootPath: l
      });
      o.load.clear();
    } catch (e) {
      setLoadError(null, e);
      i(`Failed to load schema: ${e}`);
    }
    await recordMtimes();
    persistAll();
    u.autoupdate({
      rootPath: l
    }, (e, t) => {
      revalidate();
      if (!t) {
        return;
      }
      var n = t.name ? e.multi[t.name] : e.current;
      if (n) {
        persistSchema(n);
      }
    }, (e, t) => {
      o.load.delete(null);
      setLoadError(t && t.name, e);
      recordMtimes();
      i(`Failed to load schema${t && t.name ? ` "${t.name}"` : ""} while watching: ${e}`);
    });
  })().catch(e => {
    !function setUnattributedError(e, t) {
      e.load.set(null, `Failed to load the GraphQL schema: ${toErrorMessage(t)}`);
    }(o, e);
    i(`Unexpected error while loading schema: ${e}`);
  });
  return d;
};

function getDefinitionState(e) {
  var t;
  forEachState(e, e => {
    switch (e.kind) {
     case "Query":
     case "ShortQuery":
     case "Mutation":
     case "Subscription":
     case "FragmentDefinition":
      t = e;
    }
  });
  return t;
}

function getFieldDef(e, t, n) {
  if (n === i.SchemaMetaFieldDef.name && e.getQueryType() === t) {
    return i.SchemaMetaFieldDef;
  }
  if (n === i.TypeMetaFieldDef.name && e.getQueryType() === t) {
    return i.TypeMetaFieldDef;
  }
  if (n === i.TypeNameMetaFieldDef.name && i.isCompositeType(t)) {
    return i.TypeNameMetaFieldDef;
  }
  if ("getFields" in t) {
    return t.getFields()[n];
  }
  return null;
}

function forEachState(e, t) {
  var n = [];
  var r = e;
  while (null == r ? void 0 : r.kind) {
    n.push(r);
    r = r.prevState;
  }
  for (var i = n.length - 1; i >= 0; i--) {
    t(n[i]);
  }
}

function objectValues(e) {
  var t = Object.keys(e);
  var n = t.length;
  var r = new Array(n);
  for (var i = 0; i < n; ++i) {
    r[i] = e[t[i]];
  }
  return r;
}

function hintList$1(e, t) {
  return function filterAndSortList$1(e, t) {
    if (!t) {
      return filterNonEmpty$1(e, e => !e.isDeprecated);
    }
    var n = e.map(e => ({
      proximity: getProximity$1(normalizeText$1(e.label), t),
      entry: e
    }));
    return filterNonEmpty$1(filterNonEmpty$1(n, e => e.proximity <= 2), e => !e.entry.isDeprecated).sort((e, t) => (e.entry.isDeprecated ? 1 : 0) - (t.entry.isDeprecated ? 1 : 0) || e.proximity - t.proximity || e.entry.label.length - t.entry.label.length).map(e => e.entry);
  }(t, normalizeText$1(e.string));
}

function filterNonEmpty$1(e, t) {
  var n = e.filter(t);
  return 0 === n.length ? e : n;
}

function normalizeText$1(e) {
  return e.toLowerCase().replaceAll(/\W/g, "");
}

function getProximity$1(e, t) {
  var n = function lexicalDistance$1(e, t) {
    var n;
    var r;
    var i = [];
    var a = e.length;
    var o = t.length;
    for (n = 0; n <= a; n++) {
      i[n] = [ n ];
    }
    for (r = 1; r <= o; r++) {
      i[0][r] = r;
    }
    for (n = 1; n <= a; n++) {
      for (r = 1; r <= o; r++) {
        var s = e[n - 1] === t[r - 1] ? 0 : 1;
        i[n][r] = Math.min(i[n - 1][r] + 1, i[n][r - 1] + 1, i[n - 1][r - 1] + s);
        if (n > 1 && r > 1 && e[n - 1] === t[r - 2] && e[n - 2] === t[r - 1]) {
          i[n][r] = Math.min(i[n][r], i[n - 2][r - 2] + s);
        }
      }
    }
    return i[a][o];
  }(t, e);
  if (e.length > t.length) {
    n -= e.length - t.length - 1;
    n += 0 === e.indexOf(t) ? 0 : .5;
  }
  return n;
}

var o;

!function(e) {
  e.is = function is(e) {
    return "string" == typeof e;
  };
}(o || (o = {}));

var s;

!function(e) {
  e.is = function is(e) {
    return "string" == typeof e;
  };
}(s || (s = {}));

var u;

!function(e) {
  e.MIN_VALUE = -2147483648;
  e.MAX_VALUE = 2147483647;
  e.is = function is(t) {
    return "number" == typeof t && e.MIN_VALUE <= t && t <= e.MAX_VALUE;
  };
}(u || (u = {}));

var c;

!function(e) {
  e.MIN_VALUE = 0;
  e.MAX_VALUE = 2147483647;
  e.is = function is(t) {
    return "number" == typeof t && e.MIN_VALUE <= t && t <= e.MAX_VALUE;
  };
}(c || (c = {}));

var l;

!function(e) {
  e.create = function create(e, t) {
    if (e === Number.MAX_VALUE) {
      e = c.MAX_VALUE;
    }
    if (t === Number.MAX_VALUE) {
      t = c.MAX_VALUE;
    }
    return {
      line: e,
      character: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Ce.objectLiteral(t) && Ce.uinteger(t.line) && Ce.uinteger(t.character);
  };
}(l || (l = {}));

var d;

!function(e) {
  e.create = function create(e, t, n, r) {
    if (Ce.uinteger(e) && Ce.uinteger(t) && Ce.uinteger(n) && Ce.uinteger(r)) {
      return {
        start: l.create(e, t),
        end: l.create(n, r)
      };
    } else if (l.is(e) && l.is(t)) {
      return {
        start: e,
        end: t
      };
    } else {
      throw new Error(`Range#create called with invalid arguments[${e}, ${t}, ${n}, ${r}]`);
    }
  };
  e.is = function is(e) {
    var t = e;
    return Ce.objectLiteral(t) && l.is(t.start) && l.is(t.end);
  };
}(d || (d = {}));

var f;

!function(e) {
  e.create = function create(e, t) {
    return {
      uri: e,
      range: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Ce.objectLiteral(t) && d.is(t.range) && (Ce.string(t.uri) || Ce.undefined(t.uri));
  };
}(f || (f = {}));

var v;

!function(e) {
  e.create = function create(e, t, n, r) {
    return {
      targetUri: e,
      targetRange: t,
      targetSelectionRange: n,
      originSelectionRange: r
    };
  };
  e.is = function is(e) {
    var t = e;
    return Ce.objectLiteral(t) && d.is(t.targetRange) && Ce.string(t.targetUri) && d.is(t.targetSelectionRange) && (d.is(t.originSelectionRange) || Ce.undefined(t.originSelectionRange));
  };
}(v || (v = {}));

var p;

!function(e) {
  e.create = function create(e, t, n, r) {
    return {
      red: e,
      green: t,
      blue: n,
      alpha: r
    };
  };
  e.is = function is(e) {
    var t = e;
    return Ce.objectLiteral(t) && Ce.numberRange(t.red, 0, 1) && Ce.numberRange(t.green, 0, 1) && Ce.numberRange(t.blue, 0, 1) && Ce.numberRange(t.alpha, 0, 1);
  };
}(p || (p = {}));

var g;

!function(e) {
  e.create = function create(e, t) {
    return {
      range: e,
      color: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Ce.objectLiteral(t) && d.is(t.range) && p.is(t.color);
  };
}(g || (g = {}));

var m;

!function(e) {
  e.create = function create(e, t, n) {
    return {
      label: e,
      textEdit: t,
      additionalTextEdits: n
    };
  };
  e.is = function is(e) {
    var t = e;
    return Ce.objectLiteral(t) && Ce.string(t.label) && (Ce.undefined(t.textEdit) || N.is(t)) && (Ce.undefined(t.additionalTextEdits) || Ce.typedArray(t.additionalTextEdits, N.is));
  };
}(m || (m = {}));

var E;

!function(e) {
  e.Comment = "comment";
  e.Imports = "imports";
  e.Region = "region";
}(E || (E = {}));

var T;

!function(e) {
  e.create = function create(e, t, n, r, i, a) {
    var o = {
      startLine: e,
      endLine: t
    };
    if (Ce.defined(n)) {
      o.startCharacter = n;
    }
    if (Ce.defined(r)) {
      o.endCharacter = r;
    }
    if (Ce.defined(i)) {
      o.kind = i;
    }
    if (Ce.defined(a)) {
      o.collapsedText = a;
    }
    return o;
  };
  e.is = function is(e) {
    var t = e;
    return Ce.objectLiteral(t) && Ce.uinteger(t.startLine) && Ce.uinteger(t.startLine) && (Ce.undefined(t.startCharacter) || Ce.uinteger(t.startCharacter)) && (Ce.undefined(t.endCharacter) || Ce.uinteger(t.endCharacter)) && (Ce.undefined(t.kind) || Ce.string(t.kind));
  };
}(T || (T = {}));

var y;

!function(e) {
  e.create = function create(e, t) {
    return {
      location: e,
      message: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Ce.defined(t) && f.is(t.location) && Ce.string(t.message);
  };
}(y || (y = {}));

var h;

!function(e) {
  e.Error = 1;
  e.Warning = 2;
  e.Information = 3;
  e.Hint = 4;
}(h || (h = {}));

var I;

!function(e) {
  e.Unnecessary = 1;
  e.Deprecated = 2;
}(I || (I = {}));

var S;

!function(e) {
  e.is = function is(e) {
    var t = e;
    return Ce.objectLiteral(t) && Ce.string(t.href);
  };
}(S || (S = {}));

var b;

!function(e) {
  e.create = function create(e, t, n, r, i, a) {
    var o = {
      range: e,
      message: t
    };
    if (Ce.defined(n)) {
      o.severity = n;
    }
    if (Ce.defined(r)) {
      o.code = r;
    }
    if (Ce.defined(i)) {
      o.source = i;
    }
    if (Ce.defined(a)) {
      o.relatedInformation = a;
    }
    return o;
  };
  e.is = function is(e) {
    var t;
    var n = e;
    return Ce.defined(n) && d.is(n.range) && Ce.string(n.message) && (Ce.number(n.severity) || Ce.undefined(n.severity)) && (Ce.integer(n.code) || Ce.string(n.code) || Ce.undefined(n.code)) && (Ce.undefined(n.codeDescription) || Ce.string(null === (t = n.codeDescription) || void 0 === t ? void 0 : t.href)) && (Ce.string(n.source) || Ce.undefined(n.source)) && (Ce.undefined(n.relatedInformation) || Ce.typedArray(n.relatedInformation, y.is));
  };
}(b || (b = {}));

var D;

!function(e) {
  e.create = function create(e, t, ...n) {
    var r = {
      title: e,
      command: t
    };
    if (Ce.defined(n) && n.length > 0) {
      r.arguments = n;
    }
    return r;
  };
  e.is = function is(e) {
    var t = e;
    return Ce.defined(t) && Ce.string(t.title) && Ce.string(t.command);
  };
}(D || (D = {}));

var N;

!function(e) {
  e.replace = function replace(e, t) {
    return {
      range: e,
      newText: t
    };
  };
  e.insert = function insert(e, t) {
    return {
      range: {
        start: e,
        end: e
      },
      newText: t
    };
  };
  e.del = function del(e) {
    return {
      range: e,
      newText: ""
    };
  };
  e.is = function is(e) {
    var t = e;
    return Ce.objectLiteral(t) && Ce.string(t.newText) && d.is(t.range);
  };
}(N || (N = {}));

var L;

!function(e) {
  e.create = function create(e, t, n) {
    var r = {
      label: e
    };
    if (void 0 !== t) {
      r.needsConfirmation = t;
    }
    if (void 0 !== n) {
      r.description = n;
    }
    return r;
  };
  e.is = function is(e) {
    var t = e;
    return Ce.objectLiteral(t) && Ce.string(t.label) && (Ce.boolean(t.needsConfirmation) || void 0 === t.needsConfirmation) && (Ce.string(t.description) || void 0 === t.description);
  };
}(L || (L = {}));

var F;

!function(e) {
  e.is = function is(e) {
    return Ce.string(e);
  };
}(F || (F = {}));

var A;

!function(e) {
  e.replace = function replace(e, t, n) {
    return {
      range: e,
      newText: t,
      annotationId: n
    };
  };
  e.insert = function insert(e, t, n) {
    return {
      range: {
        start: e,
        end: e
      },
      newText: t,
      annotationId: n
    };
  };
  e.del = function del(e, t) {
    return {
      range: e,
      newText: "",
      annotationId: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return N.is(t) && (L.is(t.annotationId) || F.is(t.annotationId));
  };
}(A || (A = {}));

var k;

!function(e) {
  e.create = function create(e, t) {
    return {
      textDocument: e,
      edits: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Ce.defined(t) && R.is(t.textDocument) && Array.isArray(t.edits);
  };
}(k || (k = {}));

var _;

!function(e) {
  e.create = function create(e, t, n) {
    var r = {
      kind: "create",
      uri: e
    };
    if (void 0 !== t && (void 0 !== t.overwrite || void 0 !== t.ignoreIfExists)) {
      r.options = t;
    }
    if (void 0 !== n) {
      r.annotationId = n;
    }
    return r;
  };
  e.is = function is(e) {
    var t = e;
    return t && "create" === t.kind && Ce.string(t.uri) && (void 0 === t.options || (void 0 === t.options.overwrite || Ce.boolean(t.options.overwrite)) && (void 0 === t.options.ignoreIfExists || Ce.boolean(t.options.ignoreIfExists))) && (void 0 === t.annotationId || F.is(t.annotationId));
  };
}(_ || (_ = {}));

var C;

!function(e) {
  e.create = function create(e, t, n, r) {
    var i = {
      kind: "rename",
      oldUri: e,
      newUri: t
    };
    if (void 0 !== n && (void 0 !== n.overwrite || void 0 !== n.ignoreIfExists)) {
      i.options = n;
    }
    if (void 0 !== r) {
      i.annotationId = r;
    }
    return i;
  };
  e.is = function is(e) {
    var t = e;
    return t && "rename" === t.kind && Ce.string(t.oldUri) && Ce.string(t.newUri) && (void 0 === t.options || (void 0 === t.options.overwrite || Ce.boolean(t.options.overwrite)) && (void 0 === t.options.ignoreIfExists || Ce.boolean(t.options.ignoreIfExists))) && (void 0 === t.annotationId || F.is(t.annotationId));
  };
}(C || (C = {}));

var O;

!function(e) {
  e.create = function create(e, t, n) {
    var r = {
      kind: "delete",
      uri: e
    };
    if (void 0 !== t && (void 0 !== t.recursive || void 0 !== t.ignoreIfNotExists)) {
      r.options = t;
    }
    if (void 0 !== n) {
      r.annotationId = n;
    }
    return r;
  };
  e.is = function is(e) {
    var t = e;
    return t && "delete" === t.kind && Ce.string(t.uri) && (void 0 === t.options || (void 0 === t.options.recursive || Ce.boolean(t.options.recursive)) && (void 0 === t.options.ignoreIfNotExists || Ce.boolean(t.options.ignoreIfNotExists))) && (void 0 === t.annotationId || F.is(t.annotationId));
  };
}(O || (O = {}));

var M;

!function(e) {
  e.is = function is(e) {
    return e && (void 0 !== e.changes || void 0 !== e.documentChanges) && (void 0 === e.documentChanges || e.documentChanges.every(e => {
      if (Ce.string(e.kind)) {
        return _.is(e) || C.is(e) || O.is(e);
      } else {
        return k.is(e);
      }
    }));
  };
}(M || (M = {}));

var P;

!function(e) {
  e.create = function create(e) {
    return {
      uri: e
    };
  };
  e.is = function is(e) {
    var t = e;
    return Ce.defined(t) && Ce.string(t.uri);
  };
}(P || (P = {}));

var x;

!function(e) {
  e.create = function create(e, t) {
    return {
      uri: e,
      version: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Ce.defined(t) && Ce.string(t.uri) && Ce.integer(t.version);
  };
}(x || (x = {}));

var R;

!function(e) {
  e.create = function create(e, t) {
    return {
      uri: e,
      version: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Ce.defined(t) && Ce.string(t.uri) && (null === t.version || Ce.integer(t.version));
  };
}(R || (R = {}));

var w;

!function(e) {
  e.create = function create(e, t, n, r) {
    return {
      uri: e,
      languageId: t,
      version: n,
      text: r
    };
  };
  e.is = function is(e) {
    var t = e;
    return Ce.defined(t) && Ce.string(t.uri) && Ce.string(t.languageId) && Ce.integer(t.version) && Ce.string(t.text);
  };
}(w || (w = {}));

var U;

!function(e) {
  e.PlainText = "plaintext";
  e.Markdown = "markdown";
  e.is = function is(t) {
    return t === e.PlainText || t === e.Markdown;
  };
}(U || (U = {}));

var j;

!function(e) {
  e.is = function is(e) {
    var t = e;
    return Ce.objectLiteral(e) && U.is(t.kind) && Ce.string(t.value);
  };
}(j || (j = {}));

var V;

!function(e) {
  e.Text = 1;
  e.Method = 2;
  e.Function = 3;
  e.Constructor = 4;
  e.Field = 5;
  e.Variable = 6;
  e.Class = 7;
  e.Interface = 8;
  e.Module = 9;
  e.Property = 10;
  e.Unit = 11;
  e.Value = 12;
  e.Enum = 13;
  e.Keyword = 14;
  e.Snippet = 15;
  e.Color = 16;
  e.File = 17;
  e.Reference = 18;
  e.Folder = 19;
  e.EnumMember = 20;
  e.Constant = 21;
  e.Struct = 22;
  e.Event = 23;
  e.Operator = 24;
  e.TypeParameter = 25;
}(V || (V = {}));

var $;

!function(e) {
  e.PlainText = 1;
  e.Snippet = 2;
}($ || ($ = {}));

var G;

!function(e) {
  e.Deprecated = 1;
}(G || (G = {}));

var Y;

!function(e) {
  e.create = function create(e, t, n) {
    return {
      newText: e,
      insert: t,
      replace: n
    };
  };
  e.is = function is(e) {
    var t = e;
    return t && Ce.string(t.newText) && d.is(t.insert) && d.is(t.replace);
  };
}(Y || (Y = {}));

var Q;

!function(e) {
  e.asIs = 1;
  e.adjustIndentation = 2;
}(Q || (Q = {}));

var B;

!function(e) {
  e.is = function is(e) {
    var t = e;
    return t && (Ce.string(t.detail) || void 0 === t.detail) && (Ce.string(t.description) || void 0 === t.description);
  };
}(B || (B = {}));

var K;

!function(e) {
  e.create = function create(e) {
    return {
      label: e
    };
  };
}(K || (K = {}));

var X;

!function(e) {
  e.create = function create(e, t) {
    return {
      items: e ? e : [],
      isIncomplete: !!t
    };
  };
}(X || (X = {}));

var J;

!function(e) {
  e.fromPlainText = function fromPlainText(e) {
    return e.replace(/[\\`*_{}[\]()#+\-.!]/g, "\\$&");
  };
  e.is = function is(e) {
    var t = e;
    return Ce.string(t) || Ce.objectLiteral(t) && Ce.string(t.language) && Ce.string(t.value);
  };
}(J || (J = {}));

var q;

!function(e) {
  e.is = function is(e) {
    var t = e;
    return !!t && Ce.objectLiteral(t) && (j.is(t.contents) || J.is(t.contents) || Ce.typedArray(t.contents, J.is)) && (void 0 === e.range || d.is(e.range));
  };
}(q || (q = {}));

var H;

!function(e) {
  e.create = function create(e, t) {
    return t ? {
      label: e,
      documentation: t
    } : {
      label: e
    };
  };
}(H || (H = {}));

var z;

!function(e) {
  e.create = function create(e, t, ...n) {
    var r = {
      label: e
    };
    if (Ce.defined(t)) {
      r.documentation = t;
    }
    if (Ce.defined(n)) {
      r.parameters = n;
    } else {
      r.parameters = [];
    }
    return r;
  };
}(z || (z = {}));

var W;

!function(e) {
  e.Text = 1;
  e.Read = 2;
  e.Write = 3;
}(W || (W = {}));

var Z;

!function(e) {
  e.create = function create(e, t) {
    var n = {
      range: e
    };
    if (Ce.number(t)) {
      n.kind = t;
    }
    return n;
  };
}(Z || (Z = {}));

var ee;

!function(e) {
  e.File = 1;
  e.Module = 2;
  e.Namespace = 3;
  e.Package = 4;
  e.Class = 5;
  e.Method = 6;
  e.Property = 7;
  e.Field = 8;
  e.Constructor = 9;
  e.Enum = 10;
  e.Interface = 11;
  e.Function = 12;
  e.Variable = 13;
  e.Constant = 14;
  e.String = 15;
  e.Number = 16;
  e.Boolean = 17;
  e.Array = 18;
  e.Object = 19;
  e.Key = 20;
  e.Null = 21;
  e.EnumMember = 22;
  e.Struct = 23;
  e.Event = 24;
  e.Operator = 25;
  e.TypeParameter = 26;
}(ee || (ee = {}));

var te;

!function(e) {
  e.Deprecated = 1;
}(te || (te = {}));

var ne;

!function(e) {
  e.create = function create(e, t, n, r, i) {
    var a = {
      name: e,
      kind: t,
      location: {
        uri: r,
        range: n
      }
    };
    if (i) {
      a.containerName = i;
    }
    return a;
  };
}(ne || (ne = {}));

var re;

!function(e) {
  e.create = function create(e, t, n, r) {
    return void 0 !== r ? {
      name: e,
      kind: t,
      location: {
        uri: n,
        range: r
      }
    } : {
      name: e,
      kind: t,
      location: {
        uri: n
      }
    };
  };
}(re || (re = {}));

var ie;

!function(e) {
  e.create = function create(e, t, n, r, i, a) {
    var o = {
      name: e,
      detail: t,
      kind: n,
      range: r,
      selectionRange: i
    };
    if (void 0 !== a) {
      o.children = a;
    }
    return o;
  };
  e.is = function is(e) {
    var t = e;
    return t && Ce.string(t.name) && Ce.number(t.kind) && d.is(t.range) && d.is(t.selectionRange) && (void 0 === t.detail || Ce.string(t.detail)) && (void 0 === t.deprecated || Ce.boolean(t.deprecated)) && (void 0 === t.children || Array.isArray(t.children)) && (void 0 === t.tags || Array.isArray(t.tags));
  };
}(ie || (ie = {}));

var ae;

!function(e) {
  e.Empty = "";
  e.QuickFix = "quickfix";
  e.Refactor = "refactor";
  e.RefactorExtract = "refactor.extract";
  e.RefactorInline = "refactor.inline";
  e.RefactorRewrite = "refactor.rewrite";
  e.Source = "source";
  e.SourceOrganizeImports = "source.organizeImports";
  e.SourceFixAll = "source.fixAll";
}(ae || (ae = {}));

var oe;

!function(e) {
  e.Invoked = 1;
  e.Automatic = 2;
}(oe || (oe = {}));

var se;

!function(e) {
  e.create = function create(e, t, n) {
    var r = {
      diagnostics: e
    };
    if (null != t) {
      r.only = t;
    }
    if (null != n) {
      r.triggerKind = n;
    }
    return r;
  };
  e.is = function is(e) {
    var t = e;
    return Ce.defined(t) && Ce.typedArray(t.diagnostics, b.is) && (void 0 === t.only || Ce.typedArray(t.only, Ce.string)) && (void 0 === t.triggerKind || t.triggerKind === oe.Invoked || t.triggerKind === oe.Automatic);
  };
}(se || (se = {}));

var ue;

!function(e) {
  e.create = function create(e, t, n) {
    var r = {
      title: e
    };
    var i = !0;
    if ("string" == typeof t) {
      i = !1;
      r.kind = t;
    } else if (D.is(t)) {
      r.command = t;
    } else {
      r.edit = t;
    }
    if (i && void 0 !== n) {
      r.kind = n;
    }
    return r;
  };
  e.is = function is(e) {
    var t = e;
    return t && Ce.string(t.title) && (void 0 === t.diagnostics || Ce.typedArray(t.diagnostics, b.is)) && (void 0 === t.kind || Ce.string(t.kind)) && (void 0 !== t.edit || void 0 !== t.command) && (void 0 === t.command || D.is(t.command)) && (void 0 === t.isPreferred || Ce.boolean(t.isPreferred)) && (void 0 === t.edit || M.is(t.edit));
  };
}(ue || (ue = {}));

var ce;

!function(e) {
  e.create = function create(e, t) {
    var n = {
      range: e
    };
    if (Ce.defined(t)) {
      n.data = t;
    }
    return n;
  };
  e.is = function is(e) {
    var t = e;
    return Ce.defined(t) && d.is(t.range) && (Ce.undefined(t.command) || D.is(t.command));
  };
}(ce || (ce = {}));

var le;

!function(e) {
  e.create = function create(e, t) {
    return {
      tabSize: e,
      insertSpaces: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Ce.defined(t) && Ce.uinteger(t.tabSize) && Ce.boolean(t.insertSpaces);
  };
}(le || (le = {}));

var de;

!function(e) {
  e.create = function create(e, t, n) {
    return {
      range: e,
      target: t,
      data: n
    };
  };
  e.is = function is(e) {
    var t = e;
    return Ce.defined(t) && d.is(t.range) && (Ce.undefined(t.target) || Ce.string(t.target));
  };
}(de || (de = {}));

var fe;

!function(e) {
  e.create = function create(e, t) {
    return {
      range: e,
      parent: t
    };
  };
  e.is = function is(t) {
    var n = t;
    return Ce.objectLiteral(n) && d.is(n.range) && (void 0 === n.parent || e.is(n.parent));
  };
}(fe || (fe = {}));

var ve;

!function(e) {
  e.namespace = "namespace";
  e.type = "type";
  e.class = "class";
  e.enum = "enum";
  e.interface = "interface";
  e.struct = "struct";
  e.typeParameter = "typeParameter";
  e.parameter = "parameter";
  e.variable = "variable";
  e.property = "property";
  e.enumMember = "enumMember";
  e.event = "event";
  e.function = "function";
  e.method = "method";
  e.macro = "macro";
  e.keyword = "keyword";
  e.modifier = "modifier";
  e.comment = "comment";
  e.string = "string";
  e.number = "number";
  e.regexp = "regexp";
  e.operator = "operator";
  e.decorator = "decorator";
}(ve || (ve = {}));

var pe;

!function(e) {
  e.declaration = "declaration";
  e.definition = "definition";
  e.readonly = "readonly";
  e.static = "static";
  e.deprecated = "deprecated";
  e.abstract = "abstract";
  e.async = "async";
  e.modification = "modification";
  e.documentation = "documentation";
  e.defaultLibrary = "defaultLibrary";
}(pe || (pe = {}));

var ge;

!function(e) {
  e.is = function is(e) {
    var t = e;
    return Ce.objectLiteral(t) && (void 0 === t.resultId || "string" == typeof t.resultId) && Array.isArray(t.data) && (0 === t.data.length || "number" == typeof t.data[0]);
  };
}(ge || (ge = {}));

var me;

!function(e) {
  e.create = function create(e, t) {
    return {
      range: e,
      text: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return null != t && d.is(t.range) && Ce.string(t.text);
  };
}(me || (me = {}));

var Ee;

!function(e) {
  e.create = function create(e, t, n) {
    return {
      range: e,
      variableName: t,
      caseSensitiveLookup: n
    };
  };
  e.is = function is(e) {
    var t = e;
    return null != t && d.is(t.range) && Ce.boolean(t.caseSensitiveLookup) && (Ce.string(t.variableName) || void 0 === t.variableName);
  };
}(Ee || (Ee = {}));

var Te;

!function(e) {
  e.create = function create(e, t) {
    return {
      range: e,
      expression: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return null != t && d.is(t.range) && (Ce.string(t.expression) || void 0 === t.expression);
  };
}(Te || (Te = {}));

var ye;

!function(e) {
  e.create = function create(e, t) {
    return {
      frameId: e,
      stoppedLocation: t
    };
  };
  e.is = function is(e) {
    return Ce.defined(e) && d.is(e.stoppedLocation);
  };
}(ye || (ye = {}));

var he;

!function(e) {
  e.Type = 1;
  e.Parameter = 2;
  e.is = function is(e) {
    return 1 === e || 2 === e;
  };
}(he || (he = {}));

var Ie;

!function(e) {
  e.create = function create(e) {
    return {
      value: e
    };
  };
  e.is = function is(e) {
    var t = e;
    return Ce.objectLiteral(t) && (void 0 === t.tooltip || Ce.string(t.tooltip) || j.is(t.tooltip)) && (void 0 === t.location || f.is(t.location)) && (void 0 === t.command || D.is(t.command));
  };
}(Ie || (Ie = {}));

var Se;

!function(e) {
  e.create = function create(e, t, n) {
    var r = {
      position: e,
      label: t
    };
    if (void 0 !== n) {
      r.kind = n;
    }
    return r;
  };
  e.is = function is(e) {
    var t = e;
    return Ce.objectLiteral(t) && l.is(t.position) && (Ce.string(t.label) || Ce.typedArray(t.label, Ie.is)) && (void 0 === t.kind || he.is(t.kind)) && void 0 === t.textEdits || Ce.typedArray(t.textEdits, N.is) && (void 0 === t.tooltip || Ce.string(t.tooltip) || j.is(t.tooltip)) && (void 0 === t.paddingLeft || Ce.boolean(t.paddingLeft)) && (void 0 === t.paddingRight || Ce.boolean(t.paddingRight));
  };
}(Se || (Se = {}));

var be;

!function(e) {
  e.createSnippet = function createSnippet(e) {
    return {
      kind: "snippet",
      value: e
    };
  };
}(be || (be = {}));

var De;

!function(e) {
  e.create = function create(e, t, n, r) {
    return {
      insertText: e,
      filterText: t,
      range: n,
      command: r
    };
  };
}(De || (De = {}));

var Ne;

!function(e) {
  e.create = function create(e) {
    return {
      items: e
    };
  };
}(Ne || (Ne = {}));

var Le;

!function(e) {
  e.Invoked = 0;
  e.Automatic = 1;
}(Le || (Le = {}));

var Fe;

!function(e) {
  e.create = function create(e, t) {
    return {
      range: e,
      text: t
    };
  };
}(Fe || (Fe = {}));

var Ae;

!function(e) {
  e.create = function create(e, t) {
    return {
      triggerKind: e,
      selectedCompletionInfo: t
    };
  };
}(Ae || (Ae = {}));

var ke;

!function(e) {
  e.is = function is(e) {
    var t = e;
    return Ce.objectLiteral(t) && s.is(t.uri) && Ce.string(t.name);
  };
}(ke || (ke = {}));

var _e;

!function(e) {
  e.create = function create(e, t, n, r) {
    return new FullTextDocument(e, t, n, r);
  };
  e.is = function is(e) {
    var t = e;
    return Ce.defined(t) && Ce.string(t.uri) && (Ce.undefined(t.languageId) || Ce.string(t.languageId)) && Ce.uinteger(t.lineCount) && Ce.func(t.getText) && Ce.func(t.positionAt) && Ce.func(t.offsetAt) ? !0 : !1;
  };
  e.applyEdits = function applyEdits(e, t) {
    var n = e.getText();
    var r = mergeSort(t, (e, t) => {
      var n = e.range.start.line - t.range.start.line;
      if (0 === n) {
        return e.range.start.character - t.range.start.character;
      }
      return n;
    });
    var i = n.length;
    for (var a = r.length - 1; a >= 0; a--) {
      var o = r[a];
      var s = e.offsetAt(o.range.start);
      var u = e.offsetAt(o.range.end);
      if (u <= i) {
        n = n.substring(0, s) + o.newText + n.substring(u, n.length);
      } else {
        throw new Error("Overlapping edit");
      }
      i = s;
    }
    return n;
  };
  function mergeSort(e, t) {
    if (e.length <= 1) {
      return e;
    }
    var n = e.length / 2 | 0;
    var r = e.slice(0, n);
    var i = e.slice(n);
    mergeSort(r, t);
    mergeSort(i, t);
    var a = 0;
    var o = 0;
    var s = 0;
    while (a < r.length && o < i.length) {
      if (t(r[a], i[o]) <= 0) {
        e[s++] = r[a++];
      } else {
        e[s++] = i[o++];
      }
    }
    while (a < r.length) {
      e[s++] = r[a++];
    }
    while (o < i.length) {
      e[s++] = i[o++];
    }
    return e;
  }
}(_e || (_e = {}));

class FullTextDocument {
  constructor(e, t, n, r) {
    this._uri = e;
    this._languageId = t;
    this._version = n;
    this._content = r;
    this._lineOffsets = void 0;
  }
  get uri() {
    return this._uri;
  }
  get languageId() {
    return this._languageId;
  }
  get version() {
    return this._version;
  }
  getText(e) {
    if (e) {
      var t = this.offsetAt(e.start);
      var n = this.offsetAt(e.end);
      return this._content.substring(t, n);
    }
    return this._content;
  }
  update(e, t) {
    this._content = e.text;
    this._version = t;
    this._lineOffsets = void 0;
  }
  getLineOffsets() {
    if (void 0 === this._lineOffsets) {
      var e = [];
      var t = this._content;
      var n = !0;
      for (var r = 0; r < t.length; r++) {
        if (n) {
          e.push(r);
          n = !1;
        }
        var i = t.charAt(r);
        n = "\r" === i || "\n" === i;
        if ("\r" === i && r + 1 < t.length && "\n" === t.charAt(r + 1)) {
          r++;
        }
      }
      if (n && t.length > 0) {
        e.push(t.length);
      }
      this._lineOffsets = e;
    }
    return this._lineOffsets;
  }
  positionAt(e) {
    e = Math.max(Math.min(e, this._content.length), 0);
    var t = this.getLineOffsets();
    var n = 0, r = t.length;
    if (0 === r) {
      return l.create(0, e);
    }
    while (n < r) {
      var i = Math.floor((n + r) / 2);
      if (t[i] > e) {
        r = i;
      } else {
        n = i + 1;
      }
    }
    var a = n - 1;
    return l.create(a, e - t[a]);
  }
  offsetAt(e) {
    var t = this.getLineOffsets();
    if (e.line >= t.length) {
      return this._content.length;
    } else if (e.line < 0) {
      return 0;
    }
    var n = t[e.line];
    return Math.max(Math.min(n + e.character, e.line + 1 < t.length ? t[e.line + 1] : this._content.length), n);
  }
  get lineCount() {
    return this.getLineOffsets().length;
  }
}

var Ce;

!function(e) {
  var t = Object.prototype.toString;
  e.defined = function defined(e) {
    return void 0 !== e;
  };
  e.undefined = function undefined$1(e) {
    return void 0 === e;
  };
  e.boolean = function boolean(e) {
    return !0 === e || !1 === e;
  };
  e.string = function string(e) {
    return "[object String]" === t.call(e);
  };
  e.number = function number(e) {
    return "[object Number]" === t.call(e);
  };
  e.numberRange = function numberRange(e, n, r) {
    return "[object Number]" === t.call(e) && n <= e && e <= r;
  };
  e.integer = function integer(e) {
    return "[object Number]" === t.call(e) && -2147483648 <= e && e <= 2147483647;
  };
  e.uinteger = function uinteger(e) {
    return "[object Number]" === t.call(e) && 0 <= e && e <= 2147483647;
  };
  e.func = function func(e) {
    return "[object Function]" === t.call(e);
  };
  e.objectLiteral = function objectLiteral(e) {
    return null !== e && "object" == typeof e;
  };
  e.typedArray = function typedArray(e, t) {
    return Array.isArray(e) && e.every(t);
  };
}(Ce || (Ce = {}));

var Oe;

!function(e) {
  e.Text = 1;
  e.Method = 2;
  e.Function = 3;
  e.Constructor = 4;
  e.Field = 5;
  e.Variable = 6;
  e.Class = 7;
  e.Interface = 8;
  e.Module = 9;
  e.Property = 10;
  e.Unit = 11;
  e.Value = 12;
  e.Enum = 13;
  e.Keyword = 14;
  e.Snippet = 15;
  e.Color = 16;
  e.File = 17;
  e.Reference = 18;
  e.Folder = 19;
  e.EnumMember = 20;
  e.Constant = 21;
  e.Struct = 22;
  e.Event = 23;
  e.Operator = 24;
  e.TypeParameter = 25;
}(Oe || (Oe = {}));

var Me = Object.assign(Object.assign({}, i.Kind), {
  ALIASED_FIELD: "AliasedField",
  ARGUMENTS: "Arguments",
  SHORT_QUERY: "ShortQuery",
  QUERY: "Query",
  MUTATION: "Mutation",
  SUBSCRIPTION: "Subscription",
  TYPE_CONDITION: "TypeCondition",
  INVALID: "Invalid",
  COMMENT: "Comment",
  SCHEMA_DEF: "SchemaDef",
  SCALAR_DEF: "ScalarDef",
  OBJECT_TYPE_DEF: "ObjectTypeDef",
  OBJECT_VALUE: "ObjectValue",
  LIST_VALUE: "ListValue",
  INTERFACE_DEF: "InterfaceDef",
  UNION_DEF: "UnionDef",
  ENUM_DEF: "EnumDef",
  ENUM_VALUE: "EnumValue",
  FIELD_DEF: "FieldDef",
  INPUT_DEF: "InputDef",
  INPUT_VALUE_DEF: "InputValueDef",
  ARGUMENTS_DEF: "ArgumentsDef",
  EXTEND_DEF: "ExtendDef",
  EXTENSION_DEFINITION: "ExtensionDefinition",
  DIRECTIVE_DEF: "DirectiveDef",
  IMPLEMENTS: "Implements",
  VARIABLE_DEFINITIONS: "VariableDefinitions",
  TYPE: "Type"
});

var Pe = {
  command: "editor.action.triggerSuggest",
  title: "Suggestions"
};

var xe = [ i.Kind.SCHEMA_DEFINITION, i.Kind.OPERATION_TYPE_DEFINITION, i.Kind.SCALAR_TYPE_DEFINITION, i.Kind.OBJECT_TYPE_DEFINITION, i.Kind.INTERFACE_TYPE_DEFINITION, i.Kind.UNION_TYPE_DEFINITION, i.Kind.ENUM_TYPE_DEFINITION, i.Kind.INPUT_OBJECT_TYPE_DEFINITION, i.Kind.DIRECTIVE_DEFINITION, i.Kind.SCHEMA_EXTENSION, i.Kind.SCALAR_TYPE_EXTENSION, i.Kind.OBJECT_TYPE_EXTENSION, i.Kind.INTERFACE_TYPE_EXTENSION, i.Kind.UNION_TYPE_EXTENSION, i.Kind.ENUM_TYPE_EXTENSION, i.Kind.INPUT_OBJECT_TYPE_EXTENSION ];

function getAutocompleteSuggestions(e, t, n, r, a, o) {
  var s;
  var u = Object.assign(Object.assign({}, o), {
    schema: e
  });
  var c = r || getTokenAtPosition(t, n, 1);
  var l = "Invalid" === c.state.kind ? c.state.prevState : c.state;
  var d = (null == o ? void 0 : o.mode) || function getDocumentMode(e, t) {
    if (null == t ? void 0 : t.endsWith(".graphqls")) {
      return we.TYPE_SYSTEM;
    }
    return (e => {
      var t = !1;
      if (e) {
        try {
          i.visit(i.parse(e), {
            enter(e) {
              if ("Document" === e.kind) {
                return;
              }
              if (xe.includes(e.kind)) {
                t = !0;
                return i.BREAK;
              }
              return !1;
            }
          });
        } catch (e) {
          return t;
        }
      }
      return t;
    })(e) ? we.TYPE_SYSTEM : we.EXECUTABLE;
  }(t, null == o ? void 0 : o.uri);
  if (!l) {
    return [];
  }
  var {kind: f, step: v, prevState: p} = l;
  var g = getTypeInfo(e, c.state);
  if (f === Me.DOCUMENT) {
    if (d === we.TYPE_SYSTEM) {
      return function getSuggestionsForTypeSystemDefinitions(e) {
        return hintList$1(e, [ {
          label: "extend",
          kind: Oe.Function
        }, {
          label: "type",
          kind: Oe.Function
        }, {
          label: "interface",
          kind: Oe.Function
        }, {
          label: "union",
          kind: Oe.Function
        }, {
          label: "input",
          kind: Oe.Function
        }, {
          label: "scalar",
          kind: Oe.Function
        }, {
          label: "schema",
          kind: Oe.Function
        } ]);
      }(c);
    }
    return function getSuggestionsForExecutableDefinitions(e) {
      return hintList$1(e, [ {
        label: "query",
        kind: Oe.Function
      }, {
        label: "mutation",
        kind: Oe.Function
      }, {
        label: "subscription",
        kind: Oe.Function
      }, {
        label: "fragment",
        kind: Oe.Function
      }, {
        label: "{",
        kind: Oe.Constructor
      } ]);
    }(c);
  }
  if (f === Me.EXTEND_DEF) {
    return function getSuggestionsForExtensionDefinitions(e) {
      return hintList$1(e, [ {
        label: "type",
        kind: Oe.Function
      }, {
        label: "interface",
        kind: Oe.Function
      }, {
        label: "union",
        kind: Oe.Function
      }, {
        label: "input",
        kind: Oe.Function
      }, {
        label: "scalar",
        kind: Oe.Function
      }, {
        label: "schema",
        kind: Oe.Function
      } ]);
    }(c);
  }
  if ((null === (s = null == p ? void 0 : p.prevState) || void 0 === s ? void 0 : s.kind) === Me.EXTENSION_DEFINITION && l.name) {
    return hintList$1(c, []);
  }
  if ((null == p ? void 0 : p.kind) === i.Kind.SCALAR_TYPE_EXTENSION) {
    return hintList$1(c, Object.values(e.getTypeMap()).filter(i.isScalarType).map(e => ({
      label: e.name,
      kind: Oe.Function
    })));
  }
  if ((null == p ? void 0 : p.kind) === i.Kind.OBJECT_TYPE_EXTENSION) {
    return hintList$1(c, Object.values(e.getTypeMap()).filter(e => i.isObjectType(e) && !e.name.startsWith("__")).map(e => ({
      label: e.name,
      kind: Oe.Function
    })));
  }
  if ((null == p ? void 0 : p.kind) === i.Kind.INTERFACE_TYPE_EXTENSION) {
    return hintList$1(c, Object.values(e.getTypeMap()).filter(i.isInterfaceType).map(e => ({
      label: e.name,
      kind: Oe.Function
    })));
  }
  if ((null == p ? void 0 : p.kind) === i.Kind.UNION_TYPE_EXTENSION) {
    return hintList$1(c, Object.values(e.getTypeMap()).filter(i.isUnionType).map(e => ({
      label: e.name,
      kind: Oe.Function
    })));
  }
  if ((null == p ? void 0 : p.kind) === i.Kind.ENUM_TYPE_EXTENSION) {
    return hintList$1(c, Object.values(e.getTypeMap()).filter(e => i.isEnumType(e) && !e.name.startsWith("__")).map(e => ({
      label: e.name,
      kind: Oe.Function
    })));
  }
  if ((null == p ? void 0 : p.kind) === i.Kind.INPUT_OBJECT_TYPE_EXTENSION) {
    return hintList$1(c, Object.values(e.getTypeMap()).filter(i.isInputObjectType).map(e => ({
      label: e.name,
      kind: Oe.Function
    })));
  }
  if (f === Me.IMPLEMENTS || f === Me.NAMED_TYPE && (null == p ? void 0 : p.kind) === Me.IMPLEMENTS) {
    return function getSuggestionsForImplements(e, t, n, r, a) {
      if (t.needsSeparator) {
        return [];
      }
      var o = n.getTypeMap();
      var s = objectValues(o).filter(i.isInterfaceType);
      var u = s.map(({name: e}) => e);
      var c = new Set;
      runOnlineParser$1(r, (e, t) => {
        var r, o, s, l, d;
        if (t.name) {
          if (t.kind === Me.INTERFACE_DEF && !u.includes(t.name)) {
            c.add(t.name);
          }
          if (t.kind === Me.NAMED_TYPE && (null === (r = t.prevState) || void 0 === r ? void 0 : r.kind) === Me.IMPLEMENTS) {
            if (a.interfaceDef) {
              if (null === (o = a.interfaceDef) || void 0 === o ? void 0 : o.getInterfaces().find(({name: e}) => e === t.name)) {
                return;
              }
              var f = n.getType(t.name);
              var v = null === (s = a.interfaceDef) || void 0 === s ? void 0 : s.toConfig();
              a.interfaceDef = new i.GraphQLInterfaceType(Object.assign(Object.assign({}, v), {
                interfaces: [ ...v.interfaces, f || new i.GraphQLInterfaceType({
                  name: t.name,
                  fields: {}
                }) ]
              }));
            } else if (a.objectTypeDef) {
              if (null === (l = a.objectTypeDef) || void 0 === l ? void 0 : l.getInterfaces().find(({name: e}) => e === t.name)) {
                return;
              }
              var p = n.getType(t.name);
              var g = null === (d = a.objectTypeDef) || void 0 === d ? void 0 : d.toConfig();
              a.objectTypeDef = new i.GraphQLObjectType(Object.assign(Object.assign({}, g), {
                interfaces: [ ...g.interfaces, p || new i.GraphQLInterfaceType({
                  name: t.name,
                  fields: {}
                }) ]
              }));
            }
          }
        }
      });
      var l = a.interfaceDef || a.objectTypeDef;
      var d = ((null == l ? void 0 : l.getInterfaces()) || []).map(({name: e}) => e);
      var f = s.concat([ ...c ].map(e => ({
        name: e
      }))).filter(({name: e}) => e !== (null == l ? void 0 : l.name) && !d.includes(e));
      return hintList$1(e, f.map(e => {
        var t = {
          label: e.name,
          kind: Oe.Interface,
          type: e
        };
        if (null == e ? void 0 : e.description) {
          t.documentation = e.description;
        }
        return t;
      }));
    }(c, l, e, t, g);
  }
  if (f === Me.SELECTION_SET || f === Me.FIELD || f === Me.ALIASED_FIELD) {
    return function getSuggestionsForFieldNames(e, t, n) {
      var r;
      if (t.parentType) {
        var {parentType: a} = t;
        var o = [];
        if ("getFields" in a) {
          o = objectValues(a.getFields());
        }
        if (i.isCompositeType(a)) {
          o.push(i.TypeNameMetaFieldDef);
        }
        if (a === (null === (r = null == n ? void 0 : n.schema) || void 0 === r ? void 0 : r.getQueryType())) {
          o.push(i.SchemaMetaFieldDef, i.TypeMetaFieldDef);
        }
        return hintList$1(e, o.map((e, t) => {
          var r;
          var i = {
            sortText: String(t) + e.name,
            label: e.name,
            detail: String(e.type),
            documentation: null !== (r = e.description) && void 0 !== r ? r : void 0,
            deprecated: Boolean(e.deprecationReason),
            isDeprecated: Boolean(e.deprecationReason),
            deprecationReason: e.deprecationReason,
            kind: Oe.Field,
            type: e.type
          };
          if (null == n ? void 0 : n.fillLeafsOnComplete) {
            var a = getInsertText(e);
            if (a) {
              i.insertText = e.name + a;
              i.insertTextFormat = $.Snippet;
              i.command = Pe;
            }
          }
          return i;
        }));
      }
      return [];
    }(c, g, u);
  }
  if (f === Me.ARGUMENTS || f === Me.ARGUMENT && 0 === v) {
    var {argDefs: m} = g;
    if (m) {
      return hintList$1(c, m.map(e => {
        var t;
        return {
          label: e.name,
          insertText: e.name + ": ",
          command: Pe,
          detail: String(e.type),
          documentation: null !== (t = e.description) && void 0 !== t ? t : void 0,
          kind: Oe.Variable,
          type: e.type
        };
      }));
    }
  }
  if ((f === Me.OBJECT_VALUE || f === Me.OBJECT_FIELD && 0 === v) && g.objectFieldDefs) {
    var E = objectValues(g.objectFieldDefs);
    var T = f === Me.OBJECT_VALUE ? Oe.Value : Oe.Field;
    return hintList$1(c, E.map(e => {
      var t;
      return {
        label: e.name,
        detail: String(e.type),
        documentation: null !== (t = e.description) && void 0 !== t ? t : void 0,
        kind: T,
        type: e.type
      };
    }));
  }
  if (f === Me.ENUM_VALUE || f === Me.LIST_VALUE && 1 === v || f === Me.OBJECT_FIELD && 2 === v || f === Me.ARGUMENT && 2 === v) {
    return function getSuggestionsForInputValues(e, t, n, r) {
      var a = i.getNamedType(t.inputType);
      var o = getVariableCompletions(n, r, e).filter(e => e.detail === a.name);
      if (a instanceof i.GraphQLEnumType) {
        return hintList$1(e, a.getValues().map(e => {
          var t;
          return {
            label: e.name,
            detail: String(a),
            documentation: null !== (t = e.description) && void 0 !== t ? t : void 0,
            deprecated: Boolean(e.deprecationReason),
            isDeprecated: Boolean(e.deprecationReason),
            deprecationReason: e.deprecationReason,
            kind: Oe.EnumMember,
            type: a
          };
        }).concat(o));
      }
      if (a === i.GraphQLBoolean) {
        return hintList$1(e, o.concat([ {
          label: "true",
          detail: String(i.GraphQLBoolean),
          documentation: "Not false.",
          kind: Oe.Variable,
          type: i.GraphQLBoolean
        }, {
          label: "false",
          detail: String(i.GraphQLBoolean),
          documentation: "Not true.",
          kind: Oe.Variable,
          type: i.GraphQLBoolean
        } ]));
      }
      return o;
    }(c, g, t, e);
  }
  if (f === Me.VARIABLE && 1 === v) {
    var y = i.getNamedType(g.inputType);
    return hintList$1(c, getVariableCompletions(t, e, c).filter(e => e.detail === (null == y ? void 0 : y.name)));
  }
  if (f === Me.TYPE_CONDITION && 1 === v || f === Me.NAMED_TYPE && null != p && p.kind === Me.TYPE_CONDITION) {
    return function getSuggestionsForFragmentTypeConditions(e, t, n, r) {
      var a;
      if (t.parentType) {
        if (i.isAbstractType(t.parentType)) {
          var o = i.assertAbstractType(t.parentType);
          var s = n.getPossibleTypes(o);
          var u = Object.create(null);
          for (var c of s) {
            for (var l of c.getInterfaces()) {
              u[l.name] = l;
            }
          }
          a = s.concat(objectValues(u));
        } else {
          a = [ t.parentType ];
        }
      } else {
        a = objectValues(n.getTypeMap()).filter(e => i.isCompositeType(e) && !e.name.startsWith("__"));
      }
      return hintList$1(e, a.map(e => {
        var t = i.getNamedType(e);
        return {
          label: String(e),
          documentation: (null == t ? void 0 : t.description) || "",
          kind: Oe.Field
        };
      }));
    }(c, g, e);
  }
  if (f === Me.FRAGMENT_SPREAD && 1 === v) {
    return function getSuggestionsForFragmentSpread$1(e, t, n, r, a) {
      if (!r) {
        return [];
      }
      var o = n.getTypeMap();
      var s = getDefinitionState(e.state);
      var u = function getFragmentDefinitions(e) {
        var t = [];
        runOnlineParser$1(e, (e, n) => {
          if (n.kind === Me.FRAGMENT_DEFINITION && n.name && n.type) {
            t.push({
              kind: Me.FRAGMENT_DEFINITION,
              name: {
                kind: i.Kind.NAME,
                value: n.name
              },
              selectionSet: {
                kind: Me.SELECTION_SET,
                selections: []
              },
              typeCondition: {
                kind: Me.NAMED_TYPE,
                name: {
                  kind: i.Kind.NAME,
                  value: n.type
                }
              }
            });
          }
        });
        return t;
      }(r);
      if (a && a.length > 0) {
        u.push(...a);
      }
      var c = u.filter(e => o[e.typeCondition.name.value] && !(s && s.kind === Me.FRAGMENT_DEFINITION && s.name === e.name.value) && i.isCompositeType(t.parentType) && i.isCompositeType(o[e.typeCondition.name.value]) && i.doTypesOverlap(n, t.parentType, o[e.typeCondition.name.value]));
      return hintList$1(e, c.map(e => ({
        label: e.name.value,
        detail: String(o[e.typeCondition.name.value]),
        documentation: `fragment ${e.name.value} on ${e.typeCondition.name.value}`,
        kind: Oe.Field,
        type: o[e.typeCondition.name.value]
      })));
    }(c, g, e, t, Array.isArray(a) ? a : (e => {
      var t = [];
      if (e) {
        try {
          i.visit(i.parse(e), {
            FragmentDefinition(e) {
              t.push(e);
            }
          });
        } catch (e) {
          return [];
        }
      }
      return t;
    })(a));
  }
  var h = unwrapType(l);
  if (d === we.TYPE_SYSTEM && !h.needsAdvance && f === Me.NAMED_TYPE || f === Me.LIST_TYPE) {
    if (h.kind === Me.FIELD_DEF) {
      return hintList$1(c, Object.values(e.getTypeMap()).filter(e => i.isOutputType(e) && !e.name.startsWith("__")).map(e => ({
        label: e.name,
        kind: Oe.Function
      })));
    }
    if (h.kind === Me.INPUT_VALUE_DEF) {
      return hintList$1(c, Object.values(e.getTypeMap()).filter(e => i.isInputType(e) && !e.name.startsWith("__")).map(e => ({
        label: e.name,
        kind: Oe.Function
      })));
    }
  }
  if (f === Me.VARIABLE_DEFINITION && 2 === v || f === Me.LIST_TYPE && 1 === v || f === Me.NAMED_TYPE && p && (p.kind === Me.VARIABLE_DEFINITION || p.kind === Me.LIST_TYPE || p.kind === Me.NON_NULL_TYPE)) {
    return function getSuggestionsForVariableDefinition(e, t, n) {
      var r = t.getTypeMap();
      var a = objectValues(r).filter(i.isInputType);
      return hintList$1(e, a.map(e => ({
        label: e.name,
        documentation: e.description,
        kind: Oe.Variable
      })));
    }(c, e);
  }
  if (f === Me.DIRECTIVE) {
    return function getSuggestionsForDirective(e, t, n, r) {
      var a;
      if (null === (a = t.prevState) || void 0 === a ? void 0 : a.kind) {
        var o = n.getDirectives().filter(e => function canUseDirective(e, t) {
          if (!(null == e ? void 0 : e.kind)) {
            return !1;
          }
          var {kind: n, prevState: r} = e;
          var {locations: a} = t;
          switch (n) {
           case Me.QUERY:
            return a.includes(i.DirectiveLocation.QUERY);

           case Me.MUTATION:
            return a.includes(i.DirectiveLocation.MUTATION);

           case Me.SUBSCRIPTION:
            return a.includes(i.DirectiveLocation.SUBSCRIPTION);

           case Me.FIELD:
           case Me.ALIASED_FIELD:
            return a.includes(i.DirectiveLocation.FIELD);

           case Me.FRAGMENT_DEFINITION:
            return a.includes(i.DirectiveLocation.FRAGMENT_DEFINITION);

           case Me.FRAGMENT_SPREAD:
            return a.includes(i.DirectiveLocation.FRAGMENT_SPREAD);

           case Me.INLINE_FRAGMENT:
            return a.includes(i.DirectiveLocation.INLINE_FRAGMENT);

           case Me.SCHEMA_DEF:
            return a.includes(i.DirectiveLocation.SCHEMA);

           case Me.SCALAR_DEF:
            return a.includes(i.DirectiveLocation.SCALAR);

           case Me.OBJECT_TYPE_DEF:
            return a.includes(i.DirectiveLocation.OBJECT);

           case Me.FIELD_DEF:
            return a.includes(i.DirectiveLocation.FIELD_DEFINITION);

           case Me.INTERFACE_DEF:
            return a.includes(i.DirectiveLocation.INTERFACE);

           case Me.UNION_DEF:
            return a.includes(i.DirectiveLocation.UNION);

           case Me.ENUM_DEF:
            return a.includes(i.DirectiveLocation.ENUM);

           case Me.ENUM_VALUE:
            return a.includes(i.DirectiveLocation.ENUM_VALUE);

           case Me.INPUT_DEF:
            return a.includes(i.DirectiveLocation.INPUT_OBJECT);

           case Me.INPUT_VALUE_DEF:
            switch (null == r ? void 0 : r.kind) {
             case Me.ARGUMENTS_DEF:
              return a.includes(i.DirectiveLocation.ARGUMENT_DEFINITION);

             case Me.INPUT_DEF:
              return a.includes(i.DirectiveLocation.INPUT_FIELD_DEFINITION);
            }
          }
          return !1;
        }(t.prevState, e));
        return hintList$1(e, o.map(e => ({
          label: e.name,
          documentation: e.description || "",
          kind: Oe.Function
        })));
      }
      return [];
    }(c, l, e);
  }
  return [];
}

var Re = " {\n  $1\n}";

var getInsertText = e => {
  var {type: t} = e;
  if (i.isCompositeType(t)) {
    return Re;
  }
  if (i.isListType(t) && i.isCompositeType(t.ofType)) {
    return Re;
  }
  if (i.isNonNullType(t)) {
    if (i.isCompositeType(t.ofType)) {
      return Re;
    }
    if (i.isListType(t.ofType) && i.isCompositeType(t.ofType.ofType)) {
      return Re;
    }
  }
  return null;
};

var getParentDefinition$1 = (e, t) => {
  var n, r, i, a, o, s, u, c, l, d;
  if ((null === (n = e.prevState) || void 0 === n ? void 0 : n.kind) === t) {
    return e.prevState;
  }
  if ((null === (i = null === (r = e.prevState) || void 0 === r ? void 0 : r.prevState) || void 0 === i ? void 0 : i.kind) === t) {
    return e.prevState.prevState;
  }
  if ((null === (s = null === (o = null === (a = e.prevState) || void 0 === a ? void 0 : a.prevState) || void 0 === o ? void 0 : o.prevState) || void 0 === s ? void 0 : s.kind) === t) {
    return e.prevState.prevState.prevState;
  }
  if ((null === (d = null === (l = null === (c = null === (u = e.prevState) || void 0 === u ? void 0 : u.prevState) || void 0 === c ? void 0 : c.prevState) || void 0 === l ? void 0 : l.prevState) || void 0 === d ? void 0 : d.kind) === t) {
    return e.prevState.prevState.prevState.prevState;
  }
};

function getVariableCompletions(e, t, n) {
  var r = null;
  var i;
  var a = Object.create({});
  runOnlineParser$1(e, (e, o) => {
    if ((null == o ? void 0 : o.kind) === Me.VARIABLE && o.name) {
      r = o.name;
    }
    if ((null == o ? void 0 : o.kind) === Me.NAMED_TYPE && r) {
      var s = getParentDefinition$1(o, Me.TYPE);
      if (null == s ? void 0 : s.type) {
        i = t.getType(null == s ? void 0 : s.type);
      }
    }
    if (r && i && !a[r]) {
      a[r] = {
        detail: i.toString(),
        insertText: "$" === n.string ? r : "$" + r,
        label: r,
        type: i,
        kind: Oe.Variable
      };
      r = null;
      i = null;
    }
  });
  return objectValues(a);
}

function getTokenAtPosition(e, t, n = 0) {
  var r = null;
  var i = null;
  var a = null;
  var o = runOnlineParser$1(e, (e, o, s, u) => {
    if (u !== t.line || e.getCurrentPosition() + n < t.character + 1) {
      return;
    }
    r = s;
    i = Object.assign({}, o);
    a = e.current();
    return "BREAK";
  });
  return {
    start: o.start,
    end: o.end,
    string: a || o.string,
    state: i || o.state,
    style: r || o.style
  };
}

function runOnlineParser$1(t, n) {
  var r = t.split("\n");
  var i = e.onlineParser();
  var a = i.startState();
  var o = "";
  var s = new e.CharacterStream("");
  for (var u = 0; u < r.length; u++) {
    s = new e.CharacterStream(r[u]);
    while (!s.eol()) {
      if ("BREAK" === n(s, a, o = i.token(s, a), u)) {
        break;
      }
    }
    n(s, a, o, u);
    if (!a.kind) {
      a = i.startState();
    }
  }
  return {
    start: s.getStartOfToken(),
    end: s.getCurrentPosition(),
    string: s.current(),
    state: a,
    style: o
  };
}

function getTypeInfo(e, t) {
  var n;
  var r;
  var a;
  var o;
  var s;
  var u;
  var c;
  var l;
  var d;
  var f;
  var v;
  forEachState(t, t => {
    var p;
    switch (t.kind) {
     case Me.QUERY:
     case "ShortQuery":
      f = e.getQueryType();
      break;

     case Me.MUTATION:
      f = e.getMutationType();
      break;

     case Me.SUBSCRIPTION:
      f = e.getSubscriptionType();
      break;

     case Me.INLINE_FRAGMENT:
     case Me.FRAGMENT_DEFINITION:
      if (t.type) {
        f = e.getType(t.type);
      }
      break;

     case Me.FIELD:
     case Me.ALIASED_FIELD:
      if (!f || !t.name) {
        s = null;
      } else {
        s = d ? getFieldDef(e, d, t.name) : null;
        f = s ? s.type : null;
      }
      break;

     case Me.SELECTION_SET:
      d = i.getNamedType(f);
      break;

     case Me.DIRECTIVE:
      a = t.name ? e.getDirective(t.name) : null;
      break;

     case Me.INTERFACE_DEF:
      if (t.name) {
        c = null;
        v = new i.GraphQLInterfaceType({
          name: t.name,
          interfaces: [],
          fields: {}
        });
      }
      break;

     case Me.OBJECT_TYPE_DEF:
      if (t.name) {
        v = null;
        c = new i.GraphQLObjectType({
          name: t.name,
          interfaces: [],
          fields: {}
        });
      }
      break;

     case Me.ARGUMENTS:
      if (t.prevState) {
        switch (t.prevState.kind) {
         case Me.FIELD:
          r = s && s.args;
          break;

         case Me.DIRECTIVE:
          r = a && a.args;
          break;

         case Me.ALIASED_FIELD:
          var g = null === (p = t.prevState) || void 0 === p ? void 0 : p.name;
          if (!g) {
            r = null;
            break;
          }
          var m = d ? getFieldDef(e, d, g) : null;
          if (!m) {
            r = null;
            break;
          }
          r = m.args;
          break;

         default:
          r = null;
        }
      } else {
        r = null;
      }
      break;

     case Me.ARGUMENT:
      if (r) {
        for (var E = 0; E < r.length; E++) {
          if (r[E].name === t.name) {
            n = r[E];
            break;
          }
        }
      }
      u = null == n ? void 0 : n.type;
      break;

     case Me.ENUM_VALUE:
      var T = i.getNamedType(u);
      o = T instanceof i.GraphQLEnumType ? T.getValues().find(e => e.value === t.name) : null;
      break;

     case Me.LIST_VALUE:
      var y = i.getNullableType(u);
      u = y instanceof i.GraphQLList ? y.ofType : null;
      break;

     case Me.OBJECT_VALUE:
      var h = i.getNamedType(u);
      l = h instanceof i.GraphQLInputObjectType ? h.getFields() : null;
      break;

     case Me.OBJECT_FIELD:
      var I = t.name && l ? l[t.name] : null;
      u = null == I ? void 0 : I.type;
      break;

     case Me.NAMED_TYPE:
      if (t.name) {
        f = e.getType(t.name);
      }
    }
  });
  return {
    argDef: n,
    argDefs: r,
    directiveDef: a,
    enumValue: o,
    fieldDef: s,
    inputType: u,
    objectFieldDefs: l,
    parentType: d,
    type: f,
    interfaceDef: v,
    objectTypeDef: c
  };
}

var we;

!function(e) {
  e.TYPE_SYSTEM = "TYPE_SYSTEM";
  e.EXECUTABLE = "EXECUTABLE";
}(we || (we = {}));

function unwrapType(e) {
  if (e.prevState && e.kind && [ Me.NAMED_TYPE, Me.LIST_TYPE, Me.TYPE, Me.NON_NULL_TYPE ].includes(e.kind)) {
    return unwrapType(e.prevState);
  }
  return e;
}

function getHoverInformation(e, t, n, r, i) {
  var a = r || getTokenAtPosition(t, n);
  if (!e || !a || !a.state) {
    return "";
  }
  var {kind: o, step: s} = a.state;
  var u = getTypeInfo(e, a.state);
  var c = Object.assign(Object.assign({}, i), {
    schema: e
  });
  if ("Field" === o && 0 === s && u.fieldDef || "AliasedField" === o && 2 === s && u.fieldDef) {
    var l = [];
    renderMdCodeStart(l, c);
    !function renderField(e, t, n) {
      renderQualifiedField(e, t, n);
      renderTypeAnnotation(e, t, n, t.type);
    }(l, u, c);
    renderMdCodeEnd(l, c);
    renderDescription(l, c, u.fieldDef);
    return l.join("").trim();
  }
  if ("Directive" === o && 1 === s && u.directiveDef) {
    var d = [];
    renderMdCodeStart(d, c);
    renderDirective(d, u);
    renderMdCodeEnd(d, c);
    renderDescription(d, c, u.directiveDef);
    return d.join("").trim();
  }
  if ("Argument" === o && 0 === s && u.argDef) {
    var f = [];
    renderMdCodeStart(f, c);
    !function renderArg(e, t, n) {
      if (t.directiveDef) {
        renderDirective(e, t);
      } else if (t.fieldDef) {
        renderQualifiedField(e, t, n);
      }
      if (!t.argDef) {
        return;
      }
      var {name: r} = t.argDef;
      text(e, "(");
      text(e, r);
      renderTypeAnnotation(e, t, n, t.inputType);
      text(e, ")");
    }(f, u, c);
    renderMdCodeEnd(f, c);
    renderDescription(f, c, u.argDef);
    return f.join("").trim();
  }
  if ("EnumValue" === o && u.enumValue && "description" in u.enumValue) {
    var v = [];
    renderMdCodeStart(v, c);
    !function renderEnumValue(e, t, n) {
      if (!t.enumValue) {
        return;
      }
      var {name: r} = t.enumValue;
      renderType(e, t, n, t.inputType);
      text(e, ".");
      text(e, r);
    }(v, u, c);
    renderMdCodeEnd(v, c);
    renderDescription(v, c, u.enumValue);
    return v.join("").trim();
  }
  if ("NamedType" === o && u.type && "description" in u.type) {
    var p = [];
    renderMdCodeStart(p, c);
    renderType(p, u, c, u.type);
    renderMdCodeEnd(p, c);
    renderDescription(p, c, u.type);
    return p.join("").trim();
  }
  return "";
}

function renderMdCodeStart(e, t) {
  if (t.useMarkdown) {
    text(e, "```graphql\n");
  }
}

function renderMdCodeEnd(e, t) {
  if (t.useMarkdown) {
    text(e, "\n```");
  }
}

function renderQualifiedField(e, t, n) {
  if (!t.fieldDef) {
    return;
  }
  var r = t.fieldDef.name;
  if ("__" !== r.slice(0, 2)) {
    renderType(e, t, n, t.parentType);
    text(e, ".");
  }
  text(e, r);
}

function renderDirective(e, t, n) {
  if (!t.directiveDef) {
    return;
  }
  text(e, "@" + t.directiveDef.name);
}

function renderTypeAnnotation(e, t, n, r) {
  text(e, ": ");
  renderType(e, t, n, r);
}

function renderType(e, t, n, r) {
  if (!r) {
    return;
  }
  if (r instanceof i.GraphQLNonNull) {
    renderType(e, t, n, r.ofType);
    text(e, "!");
  } else if (r instanceof i.GraphQLList) {
    text(e, "[");
    renderType(e, t, n, r.ofType);
    text(e, "]");
  } else {
    text(e, r.name);
  }
}

function renderDescription(e, t, n) {
  if (!n) {
    return;
  }
  var r = "string" == typeof n.description ? n.description : null;
  if (r) {
    text(e, "\n\n");
    text(e, r);
  }
  !function renderDeprecation(e, t, n) {
    if (!n) {
      return;
    }
    var r = n.deprecationReason || null;
    if (!r) {
      return;
    }
    text(e, "\n\n");
    text(e, "Deprecated: ");
    text(e, r);
  }(e, 0, n);
}

function text(e, t) {
  e.push(t);
}

class Cursor {
  constructor(e, t) {
    this.line = e;
    this.character = t;
  }
  setLine(e) {
    this.line = e;
  }
  setCharacter(e) {
    this.character = e;
  }
  lessThanOrEqualTo(e) {
    return this.line < e.line || this.line === e.line && this.character <= e.character;
  }
}

var getToken = (t, n) => {
  if (!e.ts.isTemplateLiteral(t) && !e.ts.isStringLiteralLike(t)) {
    return;
  }
  var r = t.getText().slice(1, -1).split("\n");
  var i = e.onlineParser();
  var a = i.startState();
  var o = t.getStart() + 1;
  var s = void 0;
  var u = void 0;
  for (var c = 0; c < r.length; c++) {
    if (s) {
      continue;
    }
    var l = o - 1;
    var d = new e.CharacterStream(r[c] + "\n");
    while (!d.eol()) {
      var f = i.token(d, a);
      var v = d.current();
      if (l + d.getStartOfToken() + 1 <= n && l + d.getCurrentPosition() >= n) {
        s = u ? u : {
          line: c,
          start: d.getStartOfToken() + 1,
          end: d.getCurrentPosition(),
          string: v,
          state: a,
          tokenKind: f
        };
        break;
      } else if ("on" === v) {
        u = {
          line: c,
          start: d.getStartOfToken() + 1,
          end: d.getCurrentPosition(),
          string: v,
          state: a,
          tokenKind: f
        };
      } else if ("." === v || ".." === v) {
        u = {
          line: c,
          start: d.getStartOfToken() + 1,
          end: d.getCurrentPosition(),
          string: v,
          state: a,
          tokenKind: f
        };
      } else {
        u = void 0;
      }
    }
    o += r[c].length + 1;
  }
  return s;
};

function hintList(e, t) {
  return function filterAndSortList(e, t) {
    if (!t) {
      return filterNonEmpty(e, e => !e.isDeprecated);
    }
    var n = e.map(e => ({
      proximity: getProximity(normalizeText(e.label), t),
      entry: e
    }));
    return filterNonEmpty(filterNonEmpty(n, e => e.proximity <= 2), e => !e.entry.isDeprecated).sort((e, t) => (e.entry.isDeprecated ? 1 : 0) - (t.entry.isDeprecated ? 1 : 0) || e.proximity - t.proximity || e.entry.label.length - t.entry.label.length).map(e => e.entry);
  }(t, normalizeText(e.string));
}

function filterNonEmpty(e, t) {
  var n = e.filter(t);
  return 0 === n.length ? e : n;
}

function normalizeText(e) {
  return e.toLowerCase().replace(/\W/g, "");
}

function getProximity(e, t) {
  var n = function lexicalDistance(e, t) {
    var n;
    var r;
    var i = [];
    var a = e.length;
    var o = t.length;
    for (n = 0; n <= a; n++) {
      i[n] = [ n ];
    }
    for (r = 1; r <= o; r++) {
      i[0][r] = r;
    }
    for (n = 1; n <= a; n++) {
      for (r = 1; r <= o; r++) {
        var s = e[n - 1] === t[r - 1] ? 0 : 1;
        i[n][r] = Math.min(i[n - 1][r] + 1, i[n][r - 1] + 1, i[n - 1][r - 1] + s);
        if (n > 1 && r > 1 && e[n - 1] === t[r - 2] && e[n - 2] === t[r - 1]) {
          i[n][r] = Math.min(i[n][r], i[n - 2][r - 2] + s);
        }
      }
    }
    return i[a][o];
  }(t, e);
  if (e.length > t.length) {
    n -= e.length - t.length - 1;
    n += 0 === e.indexOf(t) ? 0 : .5;
  }
  return n;
}

function getGraphQLCompletions(t, n, r, a) {
  var o = a.config.templateIsCallExpression ?? !0;
  var s = a.languageService.getProgram()?.getTypeChecker();
  var u = e.getSource(a, t);
  if (!u) {
    return;
  }
  var c = e.findNode(u, n);
  if (!c) {
    return;
  }
  c = o ? e.bubbleUpCallExpression(c) : e.bubbleUpTemplate(c);
  var l, d, f;
  if (o && e.isGraphQLCall(c, s)) {
    var v = e.getSchemaName(c, s);
    f = v && r.multi[v] ? r.multi[v]?.schema : r.current?.schema;
    var p = getToken(c.arguments[0], n);
    if (!f || !p || "." === p.string || ".." === p.string) {
      return;
    }
    l = `${c.arguments[0].getText().slice(1, -1)}\n${e.getAllFragments(c, a).map(t => e.print(t)).join("\n")}`;
    d = new Cursor(p.line, p.start - 1);
  } else if (!o && e.isGraphQLTag(c)) {
    var g = getToken(c.template, n);
    if (!g || !r.current || "." === g.string || ".." === g.string) {
      return;
    }
    var {combinedText: m, resolvedSpans: E} = e.resolveTemplate(c, t, a);
    var T = E.filter(e => e.original.start < n && e.original.start + e.original.length < n).reduce((e, t) => e + (t.lines - 1), 0);
    g.line = g.line + T;
    l = m;
    d = new Cursor(g.line, g.start - 1);
    f = r.current.schema;
  } else {
    return;
  }
  var [y, h] = function getSuggestionsInternal(e, t, n) {
    var r = getTokenAtPosition(t, n);
    var a = [];
    try {
      a = i.parse(t, {
        noLocation: !0
      }).definitions.filter(e => e.kind === i.Kind.FRAGMENT_DEFINITION);
    } catch (e) {}
    var o = "on" === r.string && "TypeCondition" === r.state.kind;
    var s = getAutocompleteSuggestions(e, t, n, o ? {
      ...r,
      state: {
        ...r.state,
        step: 1
      },
      type: null
    } : void 0);
    var u = !o ? function getSuggestionsForFragmentSpread(e, t, n, r, a) {
      if (!r) {
        return [];
      }
      var o = n.getTypeMap();
      var s = getDefinitionState(e.state);
      return hintList(e, a.filter(e => o[e.typeCondition.name.value] && !(s && s.kind === Me.FRAGMENT_DEFINITION && s.name === e.name.value) && i.isCompositeType(t.parentType) && i.isCompositeType(o[e.typeCondition.name.value]) && i.doTypesOverlap(n, t.parentType, o[e.typeCondition.name.value])).map(e => ({
        label: e.name.value,
        detail: String(o[e.typeCondition.name.value]),
        documentation: `fragment ${e.name.value} on ${e.typeCondition.name.value}`,
        kind: Oe.Field,
        type: o[e.typeCondition.name.value]
      })));
    }(r, getTypeInfo(e, r.state), e, t, a) : [];
    var c = "Invalid" === r.state.kind ? r.state.prevState : r.state;
    var l = getParentDefinition(r.state, Me.FIELD)?.name;
    if (c && l) {
      var {kind: d} = c;
      if (d === Me.ARGUMENTS || d === Me.ARGUMENT) {
        var f = new Set;
        runOnlineParser(t, (e, t) => {
          if (t.kind === Me.ARGUMENT) {
            var n = getParentDefinition(t, Me.FIELD);
            if (l && t.name && n?.name === l) {
              f.add(t.name);
            }
          }
        });
        s = s.filter(e => !f.has(e.label));
      }
      if (d === Me.SELECTION_SET || d === Me.FIELD || d === Me.ALIASED_FIELD) {
        var v = new Set;
        var p = getUsedFragments(t, l);
        runOnlineParser(t, (e, t) => {
          if (t.kind === Me.FIELD || t.kind === Me.ALIASED_FIELD) {
            var n = getParentDefinition(t, Me.FIELD);
            if (n && n.name === l && t.name) {
              v.add(t.name);
            }
          }
        });
        s = s.filter(e => !v.has(e.label));
        u = u.filter(e => !p.has(e.label));
      }
      if (d === Me.FRAGMENT_SPREAD) {
        var g = getUsedFragments(t, l);
        s = s.filter(e => !g.has(e.label));
        u = u.filter(e => !g.has(e.label));
      }
    }
    return [ s, u ];
  }(f, l, d);
  return {
    isGlobalCompletion: !1,
    isMemberCompletion: !1,
    isNewIdentifierLocation: !1,
    entries: [ ...y.map(t => ({
      ...t,
      kind: e.ts.ScriptElementKind.variableElement,
      name: t.label,
      kindModifiers: "declare",
      sortText: t.sortText || "0",
      labelDetails: {
        detail: t.type ? " " + t.type?.toString() : void 0,
        description: t.documentation
      }
    })), ...h.map(t => ({
      ...t,
      kind: e.ts.ScriptElementKind.variableElement,
      name: t.label,
      insertText: "..." + t.label,
      kindModifiers: "declare",
      sortText: "0",
      labelDetails: {
        description: t.documentation
      }
    })) ]
  };
}

function getUsedFragments(e, t) {
  var n = new Set;
  runOnlineParser(e, (e, r) => {
    if (r.kind === Me.FRAGMENT_SPREAD && r.name) {
      var i = getParentDefinition(r, Me.FIELD);
      if (t && i?.name === t) {
        n.add(r.name);
      }
    }
  });
  return n;
}

function getParentDefinition(e, t) {
  if (e.prevState?.kind === t) {
    return e.prevState;
  }
  if (e.prevState?.prevState?.kind === t) {
    return e.prevState.prevState;
  }
  if (e.prevState?.prevState?.prevState?.kind === t) {
    return e.prevState.prevState.prevState;
  }
  if (e.prevState?.prevState?.prevState?.prevState?.kind === t) {
    return e.prevState.prevState.prevState.prevState;
  }
}

function runOnlineParser(t, n) {
  var r = t.split("\n");
  var i = e.onlineParser();
  var a = i.startState();
  var o = "";
  var s = new e.CharacterStream("");
  for (var u = 0; u < r.length; u++) {
    s = new e.CharacterStream(r[u]);
    while (!s.eol()) {
      if ("BREAK" === n(s, a, o = i.token(s, a), u)) {
        break;
      }
    }
    n(s, a, o, u);
    if (!a.kind) {
      a = i.startState();
    }
  }
  return {
    start: s.getStartOfToken(),
    end: s.getCurrentPosition(),
    string: s.current(),
    state: a,
    style: o
  };
}

function create(t) {
  var logger = e => t.project.projectService.logger.info(`[GraphQLSP] ${e}`);
  var n = t.config;
  logger("config: " + JSON.stringify(n));
  logger("Setting up the GraphQL Plugin");
  if (n.template) {
    e.templates.add(n.template);
  }
  var r = function createBasicDecorator(e) {
    var t = Object.create(null);
    var _loop = function() {
      var r = e.languageService[n];
      t[n] = (...t) => r.apply(e.languageService, t);
    };
    for (var n of Object.keys(e.languageService)) {
      _loop();
    }
    return t;
  }(t);
  var i = loadSchema(t, logger);
  var guard = (e, t, n) => {
    try {
      return n();
    } catch (n) {
      logger(`Unexpected error in ${e}: ${n}`);
      return t;
    }
  };
  r.getSemanticDiagnostics = n => {
    var r = t.languageService.getSemanticDiagnostics(n);
    return guard("getSemanticDiagnostics", r, () => {
      i.checkStale();
      if (r.some(t => e.ALL_DIAGNOSTICS.includes(t.code))) {
        return r;
      }
      var a = e.getGraphQLDiagnostics(n, i, t);
      return a ? [ ...a, ...r ] : r;
    });
  };
  r.getCompletionsAtPosition = (e, n, r) => {
    var a = guard("getCompletionsAtPosition", void 0, () => getGraphQLCompletions(e, n, i, t));
    if (a && a.entries.length) {
      return a;
    } else {
      return t.languageService.getCompletionsAtPosition(e, n, r) || {
        isGlobalCompletion: !1,
        isMemberCompletion: !1,
        isNewIdentifierLocation: !1,
        entries: []
      };
    }
  };
  r.getEditsForRefactor = (n, r, i, a, o, s, u) => {
    var c = t.languageService.getEditsForRefactor(n, r, i, a, o, s, u);
    var l = guard("getEditsForRefactor", void 0, () => e.getPersistedCodeFixAtPosition(n, "number" == typeof i ? i : i.pos, t));
    if (!l) {
      return c;
    }
    return {
      edits: [ {
        fileName: n,
        textChanges: [ {
          newText: l.replacement,
          span: l.span
        } ]
      } ]
    };
  };
  r.getApplicableRefactors = (n, r, i, a, o, s) => {
    var u = t.languageService.getApplicableRefactors(n, r, i, a, o, s);
    if (guard("getApplicableRefactors", void 0, () => e.getPersistedCodeFixAtPosition(n, "number" == typeof r ? r : r.pos, t))) {
      return [ {
        name: "GraphQL",
        description: "Operations specific to gql.tada!",
        actions: [ {
          name: "Insert document-id",
          description: "Generate a document-id for your persisted-operation, by default a SHA256 hash."
        } ],
        inlineable: !0
      }, ...u ];
    } else {
      return u;
    }
  };
  r.getQuickInfoAtPosition = (...n) => {
    var [r, a] = n;
    var o = guard("getQuickInfoAtPosition", void 0, () => function getGraphQLQuickInfo(t, n, r, i) {
      var a = i.config.templateIsCallExpression ?? !0;
      var o = i.languageService.getProgram()?.getTypeChecker();
      var s = e.getSource(i, t);
      if (!s) {
        return;
      }
      var u = e.findNode(s, n);
      if (!u) {
        return;
      }
      u = a ? e.bubbleUpCallExpression(u) : e.bubbleUpTemplate(u);
      var c, l, d;
      if (a && e.isGraphQLCall(u, o)) {
        var f = i.languageService.getProgram()?.getTypeChecker();
        var v = e.getSchemaName(u, f);
        d = v && r.multi[v] ? r.multi[v]?.schema : r.current?.schema;
        var p = getToken(u.arguments[0], n);
        if (!d || !p) {
          return;
        }
        l = u.arguments[0].getText();
        c = new Cursor(p.line, p.start - 1);
      } else if (!a && e.isGraphQLTag(u)) {
        var g = getToken(u.template, n);
        if (!g || !r.current) {
          return;
        }
        var {combinedText: m, resolvedSpans: E} = e.resolveTemplate(u, t, i);
        var T = E.filter(e => e.original.start < n && e.original.start + e.original.length < n).reduce((e, t) => e + (t.lines - 1), 0);
        g.line = g.line + T;
        l = m;
        c = new Cursor(g.line, g.start - 1);
        d = r.current.schema;
      } else {
        return;
      }
      var y = getHoverInformation(d, l, c);
      return {
        kind: e.ts.ScriptElementKind.label,
        textSpan: {
          start: n,
          length: 1
        },
        kindModifiers: "text",
        documentation: Array.isArray(y) ? y.map(e => ({
          kind: "text",
          text: e
        })) : [ {
          kind: "text",
          text: y
        } ]
      };
    }(r, a, i, t));
    if (o) {
      return o;
    }
    return t.languageService.getQuickInfoAtPosition(...n);
  };
  logger("proxy: " + JSON.stringify(r));
  return r;
}

module.exports = t => {
  e.init(t);
  return {
    create
  };
};
