// @0no-co/graphqlsp/dist/chunks/api-chunk.mjs
import { Kind as e, specifiedRules as t, NoUnusedFragmentsRule as i, ExecutableDefinitionsRule as r, KnownFragmentNamesRule as a, validate as n, LoneSchemaDefinitionRule as s, UniqueOperationTypesRule as o, UniqueTypeNamesRule as l, UniqueEnumValueNamesRule as u, UniqueFieldDefinitionNamesRule as c, UniqueDirectiveNamesRule as f, KnownTypeNamesRule as h, KnownDirectivesRule as d, UniqueDirectivesPerLocationRule as v, PossibleTypeExtensionsRule as g, UniqueArgumentNamesRule as m, UniqueInputFieldNamesRule as S, print as y, parse as x, GraphQLError as T, NoDeprecatedCustomRule as E, visit as A } from "graphql";

import { createHash as b } from "crypto";

var k;

function init(e) {
  k = e.typescript;
}

class CharacterStream {
  constructor(e) {
    this._start = 0;
    this._pos = 0;
    this.getStartOfToken = () => this._start;
    this.getCurrentPosition = () => this._pos;
    this.eol = () => this._sourceText.length === this._pos;
    this.sol = () => 0 === this._pos;
    this.peek = () => this._sourceText.charAt(this._pos) || null;
    this.next = () => {
      var e = this._sourceText.charAt(this._pos);
      this._pos++;
      return e;
    };
    this.eat = e => {
      if (this._testNextCharacter(e)) {
        this._start = this._pos;
        this._pos++;
        return this._sourceText.charAt(this._pos - 1);
      }
      return;
    };
    this.eatWhile = e => {
      var t = this._testNextCharacter(e);
      var i = !1;
      if (t) {
        i = t;
        this._start = this._pos;
      }
      while (t) {
        this._pos++;
        t = this._testNextCharacter(e);
        i = !0;
      }
      return i;
    };
    this.eatSpace = () => this.eatWhile(/[\s\u00a0]/);
    this.skipToEnd = () => {
      this._pos = this._sourceText.length;
    };
    this.skipTo = e => {
      this._pos = e;
    };
    this.match = (e, t = !0, i = !1) => {
      var r = null;
      var a = null;
      if ("string" == typeof e) {
        a = new RegExp(e, i ? "i" : "g").test(this._sourceText.slice(this._pos, this._pos + e.length));
        r = e;
      } else if (e instanceof RegExp) {
        r = null == (a = this._sourceText.slice(this._pos).match(e)) ? void 0 : a[0];
      }
      if (null != a && ("string" == typeof e || a instanceof Array && this._sourceText.startsWith(a[0], this._pos))) {
        if (t) {
          this._start = this._pos;
          if (r && r.length) {
            this._pos += r.length;
          }
        }
        return a;
      }
      return !1;
    };
    this.backUp = e => {
      this._pos -= e;
    };
    this.column = () => this._pos;
    this.indentation = () => {
      var e = this._sourceText.match(/\s*/);
      var t = 0;
      if (e && 0 !== e.length) {
        var i = e[0];
        var r = 0;
        while (i.length > r) {
          if (9 === i.charCodeAt(r)) {
            t += 2;
          } else {
            t++;
          }
          r++;
        }
      }
      return t;
    };
    this.current = () => this._sourceText.slice(this._start, this._pos);
    this._sourceText = e;
  }
  _testNextCharacter(e) {
    var t = this._sourceText.charAt(this._pos);
    var i = !1;
    if ("string" == typeof e) {
      i = t === e;
    } else {
      i = e instanceof RegExp ? e.test(t) : e(t);
    }
    return i;
  }
}

function opt(e) {
  return {
    ofRule: e
  };
}

function list(e, t) {
  return {
    ofRule: e,
    isList: !0,
    separator: t
  };
}

function t$1(e, t) {
  return {
    style: t,
    match: t => t.kind === e
  };
}

function p(e, t) {
  return {
    style: t || "punctuation",
    match: t => "Punctuation" === t.kind && t.value === e
  };
}

var isIgnored = e => " " === e || "\t" === e || "," === e || "\n" === e || "\r" === e || "\ufeff" === e || " " === e;

var C = {
  Name: /^[_A-Za-z][_0-9A-Za-z]*/,
  Punctuation: /^(?:!|\$|\(|\)|\.\.\.|:|=|&|@|\[|]|\{|\||\})/,
  Number: /^-?(?:0|(?:[1-9][0-9]*))(?:\.[0-9]*)?(?:[eE][+-]?[0-9]+)?/,
  String: /^(?:"""(?:\\"""|[^"]|"[^"]|""[^"])*(?:""")?|"(?:[^"\\]|\\(?:"|\/|\\|b|f|n|r|t|u[0-9a-fA-F]{4}))*"?)/,
  Comment: /^#.*/
};

var w = {
  Document: [ list("Definition") ],
  Definition(t) {
    switch (t.value) {
     case "{":
      return "ShortQuery";

     case "query":
      return "Query";

     case "mutation":
      return "Mutation";

     case "subscription":
      return "Subscription";

     case "fragment":
      return e.FRAGMENT_DEFINITION;

     case "schema":
      return "SchemaDef";

     case "scalar":
      return "ScalarDef";

     case "type":
      return "ObjectTypeDef";

     case "interface":
      return "InterfaceDef";

     case "union":
      return "UnionDef";

     case "enum":
      return "EnumDef";

     case "input":
      return "InputDef";

     case "extend":
      return "ExtendDef";

     case "directive":
      return "DirectiveDef";
    }
  },
  ShortQuery: [ "SelectionSet" ],
  Query: [ word("query"), opt(name$1("def")), opt("VariableDefinitions"), list("Directive"), "SelectionSet" ],
  Mutation: [ word("mutation"), opt(name$1("def")), opt("VariableDefinitions"), list("Directive"), "SelectionSet" ],
  Subscription: [ word("subscription"), opt(name$1("def")), opt("VariableDefinitions"), list("Directive"), "SelectionSet" ],
  VariableDefinitions: [ p("("), list("VariableDefinition"), p(")") ],
  VariableDefinition: [ "Variable", p(":"), "Type", opt("DefaultValue") ],
  Variable: [ p("$", "variable"), name$1("variable") ],
  DefaultValue: [ p("="), "Value" ],
  SelectionSet: [ p("{"), list("Selection"), p("}") ],
  Selection: (e, t) => "..." === e.value ? t.match(/[\s\u00a0,]*(on\b|@|{)/, !1) ? "InlineFragment" : "FragmentSpread" : t.match(/[\s\u00a0,]*:/, !1) ? "AliasedField" : "Field",
  AliasedField: [ name$1("property"), p(":"), name$1("qualifier"), opt("Arguments"), list("Directive"), opt("SelectionSet") ],
  Field: [ name$1("property"), opt("Arguments"), list("Directive"), opt("SelectionSet") ],
  Arguments: [ p("("), list("Argument"), p(")") ],
  Argument: [ name$1("attribute"), p(":"), "Value" ],
  FragmentSpread: [ p("..."), name$1("def"), list("Directive") ],
  InlineFragment: [ p("..."), opt("TypeCondition"), list("Directive"), "SelectionSet" ],
  FragmentDefinition: [ word("fragment"), opt(function butNot(e, t) {
    var i = e.match;
    e.match = e => {
      var r = !1;
      if (i) {
        r = i(e);
      }
      return r && t.every(t => t.match && !t.match(e));
    };
    return e;
  }(name$1("def"), [ word("on") ])), "TypeCondition", list("Directive"), "SelectionSet" ],
  TypeCondition: [ word("on"), "NamedType" ],
  Value(e) {
    switch (e.kind) {
     case "Number":
      return "NumberValue";

     case "String":
      return "StringValue";

     case "Punctuation":
      switch (e.value) {
       case "[":
        return "ListValue";

       case "{":
        return "ObjectValue";

       case "$":
        return "Variable";

       case "&":
        return "NamedType";
      }
      return null;

     case "Name":
      switch (e.value) {
       case "true":
       case "false":
        return "BooleanValue";
      }
      if ("null" === e.value) {
        return "NullValue";
      }
      return "EnumValue";
    }
  },
  NumberValue: [ t$1("Number", "number") ],
  StringValue: [ {
    style: "string",
    match: e => "String" === e.kind,
    update(e, t) {
      if (t.value.startsWith('"""')) {
        e.inBlockstring = !t.value.slice(3).endsWith('"""');
      }
    }
  } ],
  BooleanValue: [ t$1("Name", "builtin") ],
  NullValue: [ t$1("Name", "keyword") ],
  EnumValue: [ name$1("string-2") ],
  ListValue: [ p("["), list("Value"), p("]") ],
  ObjectValue: [ p("{"), list("ObjectField"), p("}") ],
  ObjectField: [ name$1("attribute"), p(":"), "Value" ],
  Type: e => "[" === e.value ? "ListType" : "NonNullType",
  ListType: [ p("["), "Type", p("]"), opt(p("!")) ],
  NonNullType: [ "NamedType", opt(p("!")) ],
  NamedType: [ function type$1(e) {
    return {
      style: e,
      match: e => "Name" === e.kind,
      update(e, t) {
        var i;
        if (null === (i = e.prevState) || void 0 === i ? void 0 : i.prevState) {
          e.name = t.value;
          e.prevState.prevState.type = t.value;
        }
      }
    };
  }("atom") ],
  Directive: [ p("@", "meta"), name$1("meta"), opt("Arguments") ],
  DirectiveDef: [ word("directive"), p("@", "meta"), name$1("meta"), opt("ArgumentsDef"), word("on"), list("DirectiveLocation", p("|")) ],
  InterfaceDef: [ word("interface"), name$1("atom"), opt("Implements"), list("Directive"), p("{"), list("FieldDef"), p("}") ],
  Implements: [ word("implements"), list("NamedType", p("&")) ],
  DirectiveLocation: [ name$1("string-2") ],
  SchemaDef: [ word("schema"), list("Directive"), p("{"), list("OperationTypeDef"), p("}") ],
  OperationTypeDef: [ name$1("keyword"), p(":"), name$1("atom") ],
  ScalarDef: [ word("scalar"), name$1("atom"), list("Directive") ],
  ObjectTypeDef: [ word("type"), name$1("atom"), opt("Implements"), list("Directive"), p("{"), list("FieldDef"), p("}") ],
  FieldDef: [ name$1("property"), opt("ArgumentsDef"), p(":"), "Type", list("Directive") ],
  ArgumentsDef: [ p("("), list("InputValueDef"), p(")") ],
  InputValueDef: [ name$1("attribute"), p(":"), "Type", opt("DefaultValue"), list("Directive") ],
  UnionDef: [ word("union"), name$1("atom"), list("Directive"), p("="), list("UnionMember", p("|")) ],
  UnionMember: [ "NamedType" ],
  EnumDef: [ word("enum"), name$1("atom"), list("Directive"), p("{"), list("EnumValueDef"), p("}") ],
  EnumValueDef: [ name$1("string-2"), list("Directive") ],
  InputDef: [ word("input"), name$1("atom"), list("Directive"), p("{"), list("InputValueDef"), p("}") ],
  ExtendDef: [ word("extend"), "ExtensionDefinition" ],
  ExtensionDefinition(t) {
    switch (t.value) {
     case "schema":
      return e.SCHEMA_EXTENSION;

     case "scalar":
      return e.SCALAR_TYPE_EXTENSION;

     case "type":
      return e.OBJECT_TYPE_EXTENSION;

     case "interface":
      return e.INTERFACE_TYPE_EXTENSION;

     case "union":
      return e.UNION_TYPE_EXTENSION;

     case "enum":
      return e.ENUM_TYPE_EXTENSION;

     case "input":
      return e.INPUT_OBJECT_TYPE_EXTENSION;
    }
  },
  [e.SCHEMA_EXTENSION]: [ "SchemaDef" ],
  [e.SCALAR_TYPE_EXTENSION]: [ "ScalarDef" ],
  [e.OBJECT_TYPE_EXTENSION]: [ "ObjectTypeDef" ],
  [e.INTERFACE_TYPE_EXTENSION]: [ "InterfaceDef" ],
  [e.UNION_TYPE_EXTENSION]: [ "UnionDef" ],
  [e.ENUM_TYPE_EXTENSION]: [ "EnumDef" ],
  [e.INPUT_OBJECT_TYPE_EXTENSION]: [ "InputDef" ]
};

function word(e) {
  return {
    style: "keyword",
    match: t => "Name" === t.kind && t.value === e
  };
}

function name$1(e) {
  return {
    style: e,
    match: e => "Name" === e.kind,
    update(e, t) {
      e.name = t.value;
    }
  };
}

function onlineParser(t = {
  eatWhitespace: e => e.eatWhile(isIgnored),
  lexRules: C,
  parseRules: w,
  editorConfig: {}
}) {
  return {
    startState() {
      var i = {
        level: 0,
        step: 0,
        name: null,
        kind: null,
        type: null,
        rule: null,
        needsSeparator: !1,
        prevState: null
      };
      pushRule(t.parseRules, i, e.DOCUMENT);
      return i;
    },
    token: (e, i) => function getToken(e, t, i) {
      var r;
      if (t.inBlockstring) {
        if (e.match(/.*"""/)) {
          t.inBlockstring = !1;
          return "string";
        }
        e.skipToEnd();
        return "string";
      }
      var {lexRules: a, parseRules: n, eatWhitespace: s, editorConfig: o} = i;
      if (t.rule && 0 === t.rule.length) {
        popRule(t);
      } else if (t.needsAdvance) {
        t.needsAdvance = !1;
        advanceRule(t, !0);
      }
      if (e.sol()) {
        var l = (null == o ? void 0 : o.tabSize) || 2;
        t.indentLevel = Math.floor(e.indentation() / l);
      }
      if (s(e)) {
        return "ws";
      }
      var u = function lex(e, t) {
        var i = Object.keys(e);
        for (var r = 0; r < i.length; r++) {
          var a = t.match(e[i[r]]);
          if (a && a instanceof Array) {
            return {
              kind: i[r],
              value: a[0]
            };
          }
        }
      }(a, e);
      if (!u) {
        if (!e.match(/\S+/)) {
          e.match(/\s/);
        }
        pushRule(D, t, "Invalid");
        return "invalidchar";
      }
      if ("Comment" === u.kind) {
        pushRule(D, t, "Comment");
        return "comment";
      }
      var c = assign({}, t);
      if ("Punctuation" === u.kind) {
        if (/^[{([]/.test(u.value)) {
          if (void 0 !== t.indentLevel) {
            t.levels = (t.levels || []).concat(t.indentLevel + 1);
          }
        } else if (/^[})\]]/.test(u.value)) {
          var f = t.levels = (t.levels || []).slice(0, -1);
          if (t.indentLevel && f.length > 0 && f.at(-1) < t.indentLevel) {
            t.indentLevel = f.at(-1);
          }
        }
      }
      while (t.rule) {
        var h = "function" == typeof t.rule ? 0 === t.step ? t.rule(u, e) : null : t.rule[t.step];
        if (t.needsSeparator) {
          h = null == h ? void 0 : h.separator;
        }
        if (h) {
          if (h.ofRule) {
            h = h.ofRule;
          }
          if ("string" == typeof h) {
            pushRule(n, t, h);
            continue;
          }
          if (null === (r = h.match) || void 0 === r ? void 0 : r.call(h, u)) {
            if (h.update) {
              h.update(t, u);
            }
            if ("Punctuation" === u.kind) {
              advanceRule(t, !0);
            } else {
              t.needsAdvance = !0;
            }
            return h.style;
          }
        }
        unsuccessful(t);
      }
      assign(t, c);
      pushRule(D, t, "Invalid");
      return "invalidchar";
    }(e, i, t)
  };
}

function assign(e, t) {
  var i = Object.keys(t);
  for (var r = 0; r < i.length; r++) {
    e[i[r]] = t[i[r]];
  }
  return e;
}

var D = {
  Invalid: [],
  Comment: []
};

function pushRule(e, t, i) {
  if (!e[i]) {
    throw new TypeError("Unknown rule: " + i);
  }
  t.prevState = Object.assign({}, t);
  t.kind = i;
  t.name = null;
  t.type = null;
  t.rule = e[i];
  t.step = 0;
  t.needsSeparator = !1;
}

function popRule(e) {
  if (!e.prevState) {
    return;
  }
  e.kind = e.prevState.kind;
  e.name = e.prevState.name;
  e.type = e.prevState.type;
  e.rule = e.prevState.rule;
  e.step = e.prevState.step;
  e.needsSeparator = e.prevState.needsSeparator;
  e.prevState = e.prevState.prevState;
}

function advanceRule(e, t) {
  var i;
  if (isList(e) && e.rule) {
    var r = e.rule[e.step];
    if (r.separator) {
      var {separator: a} = r;
      e.needsSeparator = !e.needsSeparator;
      if (!e.needsSeparator && a.ofRule) {
        return;
      }
    }
    if (t) {
      return;
    }
  }
  e.needsSeparator = !1;
  e.step++;
  while (e.rule && !(Array.isArray(e.rule) && e.step < e.rule.length)) {
    popRule(e);
    if (e.rule) {
      if (isList(e)) {
        if (null === (i = e.rule) || void 0 === i ? void 0 : i[e.step].separator) {
          e.needsSeparator = !e.needsSeparator;
        }
      } else {
        e.needsSeparator = !1;
        e.step++;
      }
    }
  }
}

function isList(e) {
  var t = Array.isArray(e.rule) && "string" != typeof e.rule[e.step] && e.rule[e.step];
  return t && t.isList;
}

function unsuccessful(e) {
  while (e.rule && (!Array.isArray(e.rule) || !e.rule[e.step].ofRule)) {
    popRule(e);
  }
  if (e.rule) {
    advanceRule(e, !1);
  }
}

function getDefaultExportFromCjs(e) {
  return e && e.__esModule && Object.prototype.hasOwnProperty.call(e, "default") ? e.default : e;
}

class Range {
  constructor(e, t) {
    this.containsPosition = e => {
      if (this.start.line === e.line) {
        return this.start.character <= e.character;
      }
      if (this.end.line === e.line) {
        return this.end.character >= e.character;
      }
      return this.start.line <= e.line && this.end.line >= e.line;
    };
    this.start = e;
    this.end = t;
  }
  setStart(e, t) {
    this.start = new Position(e, t);
  }
  setEnd(e, t) {
    this.end = new Position(e, t);
  }
}

class Position {
  constructor(e, t) {
    this.lessThanOrEqualTo = e => this.line < e.line || this.line === e.line && this.character <= e.character;
    this.line = e;
    this.character = t;
  }
  setLine(e) {
    this.line = e;
  }
  setCharacter(e) {
    this.character = e;
  }
}

var L = [ s, o, l, u, c, f, h, d, v, g, m, S ];

var F = "Error", N = "Warning", _ = "Information", I = "Hint";

var O = {
  [F]: 1,
  [N]: 2,
  [_]: 3,
  [I]: 4
};

var invariant = (e, t) => {
  if (!e) {
    throw new Error(t);
  }
};

function getDiagnostics(s, o = null, l, u, c) {
  var f, h;
  var d = null;
  var v = "";
  if (c) {
    v = "string" == typeof c ? c : c.reduce((e, t) => e + y(t) + "\n\n", "");
  }
  var g = v ? `${s}\n\n${v}` : s;
  try {
    d = x(g);
  } catch (e) {
    if (e instanceof T) {
      var m = function getRange(e, t) {
        var i = onlineParser();
        var r = i.startState();
        var a = t.split("\n");
        invariant(a.length >= e.line, "Query text must have more lines than where the error happened");
        var n = null;
        for (var s = 0; s < e.line; s++) {
          n = new CharacterStream(a[s]);
          while (!n.eol()) {
            if ("invalidchar" === i.token(n, r)) {
              break;
            }
          }
        }
        invariant(n, "Expected Parser stream to be available.");
        var o = e.line - 1;
        var l = n.getStartOfToken();
        var u = n.getCurrentPosition();
        return new Range(new Position(o, l), new Position(o, u));
      }(null !== (h = null === (f = e.locations) || void 0 === f ? void 0 : f[0]) && void 0 !== h ? h : {
        line: 0,
        column: 0
      }, g);
      return [ {
        severity: O.Error,
        message: e.message,
        source: "GraphQL: Syntax",
        range: m
      } ];
    }
    throw e;
  }
  return function validateQuery(s, o = null, l, u) {
    if (!o) {
      return [];
    }
    var c = function validateWithCustomRules(s, o, l, u, c) {
      var f = t.filter(e => {
        if (e === i || e === r) {
          return !1;
        }
        if (u && e === a) {
          return !1;
        }
        return !0;
      });
      if (l) {
        Array.prototype.push.apply(f, l);
      }
      if (c) {
        Array.prototype.push.apply(f, L);
      }
      return n(s, o, f).filter(t => {
        if (t.message.includes("Unknown directive") && t.nodes) {
          var i = t.nodes[0];
          if (i && i.kind === e.DIRECTIVE) {
            var r = i.name.value;
            if ("arguments" === r || "argumentDefinitions" === r) {
              return !1;
            }
          }
        }
        return !0;
      });
    }(o, s, l, u).flatMap(e => annotations(e, O.Error, "Validation"));
    var f = n(o, s, [ E ]).flatMap(e => annotations(e, O.Warning, "Deprecation"));
    return c.concat(f);
  }(d, o, l, u);
}

function annotations(e, t, i) {
  if (!e.nodes) {
    return [];
  }
  var r = [];
  for (var [a, n] of e.nodes.entries()) {
    var s = "Variable" !== n.kind && "name" in n && void 0 !== n.name ? n.name : "variable" in n && void 0 !== n.variable ? n.variable : n;
    if (s) {
      invariant(e.locations, "GraphQL validation error requires locations.");
      var o = e.locations[a];
      var l = getLocation(s);
      var u = o.column + (l.end - l.start);
      r.push({
        source: `GraphQL: ${i}`,
        message: e.message,
        severity: t,
        range: new Range(new Position(o.line - 1, o.column - 1), new Position(o.line - 1, u))
      });
    }
  }
  return r;
}

function getLocation(e) {
  var t = e.loc;
  invariant(t, "Expected ASTNode to have a location.");
  return t;
}

var z = "FragmentDefinition";

class GraphQLError extends Error {
  constructor(e, t, i, r, a, n, s) {
    if (super(e), this.name = "GraphQLError", this.message = e, a) {
      this.path = a;
    }
    if (t) {
      this.nodes = Array.isArray(t) ? t : [ t ];
    }
    if (i) {
      this.source = i;
    }
    if (r) {
      this.positions = r;
    }
    if (n) {
      this.originalError = n;
    }
    var o = s;
    if (!o && n) {
      var l = n.extensions;
      if (l && "object" == typeof l) {
        o = l;
      }
    }
    this.extensions = o || {};
  }
  toJSON() {
    return {
      ...this,
      message: this.message
    };
  }
  toString() {
    return this.message;
  }
  get [Symbol.toStringTag]() {
    return "GraphQLError";
  }
}

var P;

var V;

function error(e) {
  return new GraphQLError(`Syntax Error: Unexpected token at ${V} in ${e}`);
}

function advance(e) {
  if (e.lastIndex = V, e.test(P)) {
    return P.slice(V, V = e.lastIndex);
  }
}

var R = / +(?=[^\s])/y;

function blockString(e) {
  var t = e.split("\n");
  var i = "";
  var r = 0;
  var a = 0;
  var n = t.length - 1;
  for (var s = 0; s < t.length; s++) {
    if (R.lastIndex = 0, R.test(t[s])) {
      if (s && (!r || R.lastIndex < r)) {
        r = R.lastIndex;
      }
      a = a || s, n = s;
    }
  }
  for (var o = a; o <= n; o++) {
    if (o !== a) {
      i += "\n";
    }
    i += t[o].slice(r).replace(/\\"""/g, '"""');
  }
  return i;
}

function ignored() {
  for (var e = 0 | P.charCodeAt(V++); 9 === e || 10 === e || 13 === e || 32 === e || 35 === e || 44 === e || 65279 === e; e = 0 | P.charCodeAt(V++)) {
    if (35 === e) {
      for (;(e = 0 | P.charCodeAt(V++)) && 10 !== e && 13 !== e; ) {}
    }
  }
  V--;
}

function name() {
  var e = V;
  for (var t = 0 | P.charCodeAt(V++); t >= 48 && t <= 57 || t >= 65 && t <= 90 || 95 === t || t >= 97 && t <= 122; t = 0 | P.charCodeAt(V++)) {}
  if (e === V - 1) {
    throw error("Name");
  }
  var i = P.slice(e, --V);
  return ignored(), i;
}

function nameNode() {
  return {
    kind: "Name",
    value: name()
  };
}

var M = /(?:"""|(?:[\s\S]*?[^\\])""")/y;

var B = /(?:(?:\.\d+)?[eE][+-]?\d+|\.\d+)/y;

function value(e) {
  var t;
  switch (P.charCodeAt(V)) {
   case 91:
    V++, ignored();
    var i = [];
    for (;93 !== P.charCodeAt(V); ) {
      i.push(value(e));
    }
    return V++, ignored(), {
      kind: "ListValue",
      values: i
    };

   case 123:
    V++, ignored();
    var r = [];
    for (;125 !== P.charCodeAt(V); ) {
      var a = nameNode();
      if (58 !== P.charCodeAt(V++)) {
        throw error("ObjectField");
      }
      ignored(), r.push({
        kind: "ObjectField",
        name: a,
        value: value(e)
      });
    }
    return V++, ignored(), {
      kind: "ObjectValue",
      fields: r
    };

   case 36:
    if (e) {
      throw error("Variable");
    }
    return V++, {
      kind: "Variable",
      name: nameNode()
    };

   case 34:
    if (34 === P.charCodeAt(V + 1) && 34 === P.charCodeAt(V + 2)) {
      if (V += 3, null == (t = advance(M))) {
        throw error("StringValue");
      }
      return ignored(), {
        kind: "StringValue",
        value: blockString(t.slice(0, -3)),
        block: !0
      };
    } else {
      var n = V;
      var s;
      V++;
      var o = !1;
      for (s = 0 | P.charCodeAt(V++); 92 === s && (V++, o = !0) || 10 !== s && 13 !== s && 34 !== s && s; s = 0 | P.charCodeAt(V++)) {}
      if (34 !== s) {
        throw error("StringValue");
      }
      return t = P.slice(n, V), ignored(), {
        kind: "StringValue",
        value: o ? JSON.parse(t) : t.slice(1, -1),
        block: !1
      };
    }

   case 45:
   case 48:
   case 49:
   case 50:
   case 51:
   case 52:
   case 53:
   case 54:
   case 55:
   case 56:
   case 57:
    var l = V++;
    var u;
    for (;(u = 0 | P.charCodeAt(V++)) >= 48 && u <= 57; ) {}
    var c = P.slice(l, --V);
    if (46 === (u = P.charCodeAt(V)) || 69 === u || 101 === u) {
      if (null == (t = advance(B))) {
        throw error("FloatValue");
      }
      return ignored(), {
        kind: "FloatValue",
        value: c + t
      };
    } else {
      return ignored(), {
        kind: "IntValue",
        value: c
      };
    }

   case 110:
    if (117 === P.charCodeAt(V + 1) && 108 === P.charCodeAt(V + 2) && 108 === P.charCodeAt(V + 3)) {
      return V += 4, ignored(), {
        kind: "NullValue"
      };
    } else {
      break;
    }

   case 116:
    if (114 === P.charCodeAt(V + 1) && 117 === P.charCodeAt(V + 2) && 101 === P.charCodeAt(V + 3)) {
      return V += 4, ignored(), {
        kind: "BooleanValue",
        value: !0
      };
    } else {
      break;
    }

   case 102:
    if (97 === P.charCodeAt(V + 1) && 108 === P.charCodeAt(V + 2) && 115 === P.charCodeAt(V + 3) && 101 === P.charCodeAt(V + 4)) {
      return V += 5, ignored(), {
        kind: "BooleanValue",
        value: !1
      };
    } else {
      break;
    }
  }
  return {
    kind: "EnumValue",
    value: name()
  };
}

function arguments_(e) {
  if (40 === P.charCodeAt(V)) {
    var t = [];
    V++, ignored();
    do {
      var i = nameNode();
      if (58 !== P.charCodeAt(V++)) {
        throw error("Argument");
      }
      ignored(), t.push({
        kind: "Argument",
        name: i,
        value: value(e)
      });
    } while (41 !== P.charCodeAt(V));
    return V++, ignored(), t;
  }
}

function directives(e) {
  if (64 === P.charCodeAt(V)) {
    var t = [];
    do {
      V++, t.push({
        kind: "Directive",
        name: nameNode(),
        arguments: arguments_(e)
      });
    } while (64 === P.charCodeAt(V));
    return t;
  }
}

function type() {
  var e = 0;
  for (;91 === P.charCodeAt(V); ) {
    e++, V++, ignored();
  }
  var t = {
    kind: "NamedType",
    name: nameNode()
  };
  do {
    if (33 === P.charCodeAt(V)) {
      V++, ignored(), t = {
        kind: "NonNullType",
        type: t
      };
    }
    if (e) {
      if (93 !== P.charCodeAt(V++)) {
        throw error("NamedType");
      }
      ignored(), t = {
        kind: "ListType",
        type: t
      };
    }
  } while (e--);
  return t;
}

function selectionSetStart() {
  if (123 !== P.charCodeAt(V++)) {
    throw error("SelectionSet");
  }
  return ignored(), selectionSet();
}

function selectionSet() {
  var e = [];
  do {
    if (46 === P.charCodeAt(V)) {
      if (46 !== P.charCodeAt(++V) || 46 !== P.charCodeAt(++V)) {
        throw error("SelectionSet");
      }
      switch (V++, ignored(), P.charCodeAt(V)) {
       case 64:
        e.push({
          kind: "InlineFragment",
          typeCondition: void 0,
          directives: directives(!1),
          selectionSet: selectionSetStart()
        });
        break;

       case 111:
        if (110 === P.charCodeAt(V + 1)) {
          V += 2, ignored(), e.push({
            kind: "InlineFragment",
            typeCondition: {
              kind: "NamedType",
              name: nameNode()
            },
            directives: directives(!1),
            selectionSet: selectionSetStart()
          });
        } else {
          e.push({
            kind: "FragmentSpread",
            name: nameNode(),
            directives: directives(!1)
          });
        }
        break;

       case 123:
        V++, ignored(), e.push({
          kind: "InlineFragment",
          typeCondition: void 0,
          directives: void 0,
          selectionSet: selectionSet()
        });
        break;

       default:
        e.push({
          kind: "FragmentSpread",
          name: nameNode(),
          directives: directives(!1)
        });
      }
    } else {
      var t = nameNode();
      var i = void 0;
      if (58 === P.charCodeAt(V)) {
        V++, ignored(), i = t, t = nameNode();
      }
      var r = arguments_(!1);
      var a = directives(!1);
      var n = void 0;
      if (123 === P.charCodeAt(V)) {
        V++, ignored(), n = selectionSet();
      }
      e.push({
        kind: "Field",
        alias: i,
        name: t,
        arguments: r,
        directives: a,
        selectionSet: n
      });
    }
  } while (125 !== P.charCodeAt(V));
  return V++, ignored(), {
    kind: "SelectionSet",
    selections: e
  };
}

function variableDefinitions() {
  if (ignored(), 40 === P.charCodeAt(V)) {
    var e = [];
    V++, ignored();
    do {
      var t = void 0;
      if (34 === P.charCodeAt(V)) {
        t = value(!0);
      }
      if (36 !== P.charCodeAt(V++)) {
        throw error("Variable");
      }
      var i = nameNode();
      if (58 !== P.charCodeAt(V++)) {
        throw error("VariableDefinition");
      }
      ignored();
      var r = type();
      var a = void 0;
      if (61 === P.charCodeAt(V)) {
        V++, ignored(), a = value(!0);
      }
      ignored();
      var n = {
        kind: "VariableDefinition",
        variable: {
          kind: "Variable",
          name: i
        },
        type: r,
        defaultValue: a,
        directives: directives(!0)
      };
      if (t) {
        n.description = t;
      }
      e.push(n);
    } while (41 !== P.charCodeAt(V));
    return V++, ignored(), e;
  }
}

function fragmentDefinition(e) {
  var t = nameNode();
  if (111 !== P.charCodeAt(V++) || 110 !== P.charCodeAt(V++)) {
    throw error("FragmentDefinition");
  }
  ignored();
  var i = {
    kind: "FragmentDefinition",
    name: t,
    typeCondition: {
      kind: "NamedType",
      name: nameNode()
    },
    directives: directives(!1),
    selectionSet: selectionSetStart()
  };
  if (e) {
    i.description = e;
  }
  return i;
}

function definitions() {
  var e = [];
  do {
    var t = void 0;
    if (34 === P.charCodeAt(V)) {
      t = value(!0);
    }
    if (123 === P.charCodeAt(V)) {
      if (t) {
        throw error("Document");
      }
      V++, ignored(), e.push({
        kind: "OperationDefinition",
        operation: "query",
        name: void 0,
        variableDefinitions: void 0,
        directives: void 0,
        selectionSet: selectionSet()
      });
    } else {
      var i = name();
      switch (i) {
       case "fragment":
        e.push(fragmentDefinition(t));
        break;

       case "query":
       case "mutation":
       case "subscription":
        var r;
        var a = void 0;
        if (40 !== (r = P.charCodeAt(V)) && 64 !== r && 123 !== r) {
          a = nameNode();
        }
        var n = {
          kind: "OperationDefinition",
          operation: i,
          name: a,
          variableDefinitions: variableDefinitions(),
          directives: directives(!1),
          selectionSet: selectionSetStart()
        };
        if (t) {
          n.description = t;
        }
        e.push(n);
        break;

       default:
        throw error("Document");
      }
    }
  } while (V < P.length);
  return e;
}

function parse(e, t) {
  if (P = e.body ? e.body : e, V = 0, ignored(), t && t.noLocation) {
    return {
      kind: "Document",
      definitions: definitions()
    };
  } else {
    return {
      kind: "Document",
      definitions: definitions(),
      loc: {
        start: 0,
        end: P.length,
        startToken: void 0,
        endToken: void 0,
        source: {
          body: P,
          name: "graphql.web",
          locationOffset: {
            line: 1,
            column: 1
          }
        }
      }
    };
  }
}

var $ = {};

function visit(e, t) {
  var i = [];
  var r = [];
  try {
    var a = function traverse(e, a, n) {
      var s = !1;
      var o = t[e.kind] && t[e.kind].enter || t[e.kind] || t.enter;
      var l = o && o.call(t, e, a, n, r, i);
      if (!1 === l) {
        return e;
      } else if (null === l) {
        return null;
      } else if (l === $) {
        throw $;
      } else if (l && "string" == typeof l.kind) {
        s = l !== e, e = l;
      }
      if (n) {
        i.push(n);
      }
      var u;
      var c = {
        ...e
      };
      for (var f in e) {
        r.push(f);
        var h = e[f];
        if (Array.isArray(h)) {
          var d = [];
          for (var v = 0; v < h.length; v++) {
            if (null != h[v] && "string" == typeof h[v].kind) {
              if (i.push(e), r.push(v), u = traverse(h[v], v, h), r.pop(), i.pop(), null == u) {
                s = !0;
              } else {
                s = s || u !== h[v], d.push(u);
              }
            }
          }
          h = d;
        } else if (null != h && "string" == typeof h.kind) {
          if (void 0 !== (u = traverse(h, f, e))) {
            s = s || h !== u, h = u;
          }
        }
        if (r.pop(), s) {
          c[f] = h;
        }
      }
      if (n) {
        i.pop();
      }
      var g = t[e.kind] && t[e.kind].leave || t.leave;
      var m = g && g.call(t, e, a, n, r, i);
      if (m === $) {
        throw $;
      } else if (void 0 !== m) {
        return m;
      } else if (void 0 !== l) {
        return s ? c : l;
      } else {
        return s ? c : e;
      }
    }(e);
    return void 0 !== a && !1 !== a ? a : e;
  } catch (t) {
    if (t !== $) {
      throw t;
    }
    return e;
  }
}

function mapJoin(e, t, i) {
  var r = "";
  for (var a = 0; a < e.length; a++) {
    if (a) {
      r += t;
    }
    r += i(e[a]);
  }
  return r;
}

var j = "\n";

var W = {
  OperationDefinition(e) {
    var t = "";
    if (e.description) {
      t += W.StringValue(e.description) + "\n";
    }
    if (t += e.operation, e.name) {
      t += " " + e.name.value;
    }
    if (e.variableDefinitions && e.variableDefinitions.length) {
      if (!e.name) {
        t += " ";
      }
      t += "(" + mapJoin(e.variableDefinitions, ", ", W.VariableDefinition) + ")";
    }
    if (e.directives && e.directives.length) {
      t += " " + mapJoin(e.directives, " ", W.Directive);
    }
    var i = W.SelectionSet(e.selectionSet);
    return "query" !== t ? t + " " + i : i;
  },
  VariableDefinition(e) {
    var t = "";
    if (e.description) {
      t += W.StringValue(e.description) + " ";
    }
    if (t += W.Variable(e.variable) + ": " + _print(e.type), e.defaultValue) {
      t += " = " + _print(e.defaultValue);
    }
    if (e.directives && e.directives.length) {
      t += " " + mapJoin(e.directives, " ", W.Directive);
    }
    return t;
  },
  Field(e) {
    var t = e.alias ? e.alias.value + ": " + e.name.value : e.name.value;
    if (e.arguments && e.arguments.length) {
      var i = mapJoin(e.arguments, ", ", W.Argument);
      if (t.length + i.length + 2 > 80) {
        t += "(" + (j += "  ") + mapJoin(e.arguments, j, W.Argument) + (j = j.slice(0, -2)) + ")";
      } else {
        t += "(" + i + ")";
      }
    }
    if (e.directives && e.directives.length) {
      t += " " + mapJoin(e.directives, " ", W.Directive);
    }
    if (e.selectionSet && e.selectionSet.selections.length) {
      t += " " + W.SelectionSet(e.selectionSet);
    }
    return t;
  },
  StringValue(e) {
    if (e.block) {
      return function printBlockString(e) {
        return '"""\n' + e.replace(/"""/g, '\\"""') + '\n"""';
      }(e.value).replace(/\n/g, j);
    } else {
      return function printString(e) {
        return JSON.stringify(e);
      }(e.value);
    }
  },
  BooleanValue: e => "" + e.value,
  NullValue: e => "null",
  IntValue: e => e.value,
  FloatValue: e => e.value,
  EnumValue: e => e.value,
  Name: e => e.value,
  Variable: e => "$" + e.name.value,
  ListValue: e => "[" + mapJoin(e.values, ", ", _print) + "]",
  ObjectValue: e => "{" + mapJoin(e.fields, ", ", W.ObjectField) + "}",
  ObjectField: e => e.name.value + ": " + _print(e.value),
  Document(e) {
    if (!e.definitions || !e.definitions.length) {
      return "";
    } else {
      return mapJoin(e.definitions, "\n\n", _print);
    }
  },
  SelectionSet: e => "{" + (j += "  ") + mapJoin(e.selections, j, _print) + (j = j.slice(0, -2)) + "}",
  Argument: e => e.name.value + ": " + _print(e.value),
  FragmentSpread(e) {
    var t = "..." + e.name.value;
    if (e.directives && e.directives.length) {
      t += " " + mapJoin(e.directives, " ", W.Directive);
    }
    return t;
  },
  InlineFragment(e) {
    var t = "...";
    if (e.typeCondition) {
      t += " on " + e.typeCondition.name.value;
    }
    if (e.directives && e.directives.length) {
      t += " " + mapJoin(e.directives, " ", W.Directive);
    }
    return t + " " + W.SelectionSet(e.selectionSet);
  },
  FragmentDefinition(e) {
    var t = "";
    if (e.description) {
      t += W.StringValue(e.description) + "\n";
    }
    if (t += "fragment " + e.name.value, t += " on " + e.typeCondition.name.value, e.directives && e.directives.length) {
      t += " " + mapJoin(e.directives, " ", W.Directive);
    }
    return t + " " + W.SelectionSet(e.selectionSet);
  },
  Directive(e) {
    var t = "@" + e.name.value;
    if (e.arguments && e.arguments.length) {
      t += "(" + mapJoin(e.arguments, ", ", W.Argument) + ")";
    }
    return t;
  },
  NamedType: e => e.name.value,
  ListType: e => "[" + _print(e.type) + "]",
  NonNullType: e => _print(e.type) + "!"
};

var _print = e => W[e.kind](e);

function print(e) {
  return j = "\n", W[e.kind] ? W[e.kind](e) : "";
}

var U = new Set([ "gql", "graphql" ]);

var isIIFE = e => k.isCallExpression(e) && 0 === e.arguments.length && (k.isFunctionExpression(e.expression) || k.isArrowFunction(e.expression)) && !e.expression.asteriskToken && !e.expression.modifiers?.length;

var isGraphQLFunctionIdentifier = e => k.isIdentifier(e) && U.has(e.escapedText);

var K = new WeakMap;

var G = new WeakMap;

var getCached = (e, t, i, r) => {
  var a = t.getSymbolAtLocation(i);
  if (!a) {
    return r();
  }
  var n = e.get(t);
  if (!n) {
    e.set(t, n = new WeakMap);
  }
  var s = n.get(a);
  if (void 0 === s) {
    n.set(a, s = r());
  }
  return s;
};

var isTadaGraphQLFunction = (e, t) => {
  if (!k.isLeftHandSideExpression(e) || !t) {
    return !1;
  }
  return getCached(K, t, e, () => {
    var i = t.getTypeAtLocation(e);
    return null != i && null != i.getProperty("scalar") && null != i.getProperty("persisted");
  });
};

var q = /^(?:[\s,]+|#[^\n\r]*)*(?:\{|(?:query|mutation|subscription|fragment)(?=[\s,{(@#]|$))/;

var isTadaGraphQLCall = (e, t) => {
  if (!k.isCallExpression(e)) {
    return !1;
  } else if (e.arguments.length < 1 || e.arguments.length > 2) {
    return !1;
  } else if (!k.isStringLiteralLike(e.arguments[0])) {
    return !1;
  } else if (!q.test(e.arguments[0].text)) {
    return !1;
  }
  return t ? isTadaGraphQLFunction(e.expression, t) : !1;
};

var isTadaPersistedCall = (e, t) => {
  if (!e) {
    return !1;
  } else if (!k.isCallExpression(e)) {
    return !1;
  } else if (!k.isPropertyAccessExpression(e.expression)) {
    return !1;
  } else if (!k.isIdentifier(e.expression.name) || "persisted" !== e.expression.name.escapedText) {
    return !1;
  } else if (isGraphQLFunctionIdentifier(e.expression.expression)) {
    return !0;
  } else {
    return isTadaGraphQLFunction(e.expression.expression, t);
  }
};

var isGraphQLCall = (e, t) => k.isCallExpression(e) && e.arguments.length >= 1 && e.arguments.length <= 2 && (isGraphQLFunctionIdentifier(e.expression) || isTadaGraphQLCall(e, t));

var isGraphQLTag = e => k.isTaggedTemplateExpression(e) && isGraphQLFunctionIdentifier(e.tag);

var getSchemaName = (e, t, i = !1) => {
  if (!t) {
    return null;
  }
  var r = i ? e.getChildAt(0).getChildAt(0) : e.expression;
  return getCached(G, t, r, () => {
    var e = t.getTypeAtLocation(r);
    if (e) {
      var i = e.getProperty("__name");
      if (i) {
        var a = t.getTypeOfSymbol(i);
        if (a.isUnionOrIntersection()) {
          var n = a.types.find(e => e.isStringLiteral());
          return n && n.isStringLiteral() ? n.value : null;
        } else if (a.isStringLiteral()) {
          return a.value;
        }
      }
    }
    return null;
  });
};

function isValueDeclaration(e) {
  switch (e.kind) {
   case k.SyntaxKind.BinaryExpression:
   case k.SyntaxKind.ArrowFunction:
   case k.SyntaxKind.BindingElement:
   case k.SyntaxKind.ClassDeclaration:
   case k.SyntaxKind.ClassExpression:
   case k.SyntaxKind.ClassStaticBlockDeclaration:
   case k.SyntaxKind.Constructor:
   case k.SyntaxKind.EnumDeclaration:
   case k.SyntaxKind.EnumMember:
   case k.SyntaxKind.ExportAssignment:
   case k.SyntaxKind.FunctionDeclaration:
   case k.SyntaxKind.FunctionExpression:
   case k.SyntaxKind.GetAccessor:
   case k.SyntaxKind.JsxAttribute:
   case k.SyntaxKind.MethodDeclaration:
   case k.SyntaxKind.Parameter:
   case k.SyntaxKind.PropertyAssignment:
   case k.SyntaxKind.PropertyDeclaration:
   case k.SyntaxKind.SetAccessor:
   case k.SyntaxKind.ShorthandPropertyAssignment:
   case k.SyntaxKind.VariableDeclaration:
    return !0;

   default:
    return !1;
  }
}

function getValueOfValueDeclaration(e) {
  switch (e.kind) {
   case k.SyntaxKind.ClassExpression:
   case k.SyntaxKind.ClassDeclaration:
   case k.SyntaxKind.ArrowFunction:
   case k.SyntaxKind.ClassStaticBlockDeclaration:
   case k.SyntaxKind.Constructor:
   case k.SyntaxKind.EnumDeclaration:
   case k.SyntaxKind.FunctionDeclaration:
   case k.SyntaxKind.FunctionExpression:
   case k.SyntaxKind.GetAccessor:
   case k.SyntaxKind.SetAccessor:
   case k.SyntaxKind.MethodDeclaration:
    return e;

   case k.SyntaxKind.BindingElement:
   case k.SyntaxKind.EnumMember:
   case k.SyntaxKind.JsxAttribute:
   case k.SyntaxKind.Parameter:
   case k.SyntaxKind.PropertyAssignment:
   case k.SyntaxKind.PropertyDeclaration:
   case k.SyntaxKind.VariableDeclaration:
    return e.initializer;

   case k.SyntaxKind.ExportAssignment:
    return e.expression;

   case k.SyntaxKind.BinaryExpression:
    return function isAssignmentOperator(e) {
      switch (e.kind) {
       case k.SyntaxKind.EqualsToken:
       case k.SyntaxKind.BarBarEqualsToken:
       case k.SyntaxKind.AmpersandAmpersandEqualsToken:
       case k.SyntaxKind.QuestionQuestionEqualsToken:
        return !0;

       default:
        return !1;
      }
    }(e.operatorToken) ? e.right : void 0;

   case k.SyntaxKind.ShorthandPropertyAssignment:
    return e.objectAssignmentInitializer;

   default:
    return;
  }
}

function climbPastPropertyOrElementAccess(e) {
  if (e.parent && k.isPropertyAccessExpression(e.parent) && e.parent.name === e) {
    return e.parent;
  } else if (e.parent && k.isElementAccessExpression(e.parent) && e.parent.argumentExpression === e) {
    return e.parent;
  } else {
    return e;
  }
}

function getNameFromPropertyName(e) {
  if (k.isComputedPropertyName(e)) {
    return k.isStringLiteralLike(e.expression) || k.isNumericLiteral(e.expression) ? e.expression.text : void 0;
  } else if (k.isPrivateIdentifier(e) || k.isMemberName(e)) {
    return k.idText(e);
  } else {
    return e.text;
  }
}

function getDeclarationOfIdentifier(e, t) {
  var i = t.getSymbolAtLocation(e);
  if (i?.declarations?.[0] && i.flags & k.SymbolFlags.Alias && (e.parent === i?.declarations?.[0] || !k.isNamespaceImport(i.declarations[0]))) {
    var r = t.getAliasedSymbol(i);
    if (r.declarations) {
      i = r;
    }
  }
  if (i && k.isShorthandPropertyAssignment(e.parent)) {
    var a = t.getShorthandAssignmentValueSymbol(i.valueDeclaration);
    if (a) {
      i = a;
    }
  } else if (k.isBindingElement(e.parent) && k.isObjectBindingPattern(e.parent.parent) && e === (e.parent.propertyName || e.parent.name)) {
    var n = getNameFromPropertyName(e);
    var s = n ? t.getTypeAtLocation(e.parent.parent).getProperty(n) : void 0;
    if (s) {
      i = s;
    }
  } else if (k.isObjectLiteralElement(e.parent) && (k.isObjectLiteralExpression(e.parent.parent) || k.isJsxAttributes(e.parent.parent)) && e.parent.name === e) {
    var o = getNameFromPropertyName(e);
    var l = o ? t.getContextualType(e.parent.parent)?.getProperty(o) : void 0;
    if (l) {
      i = l;
    }
  }
  if (i && i.declarations?.length) {
    if (i.flags & k.SymbolFlags.Class && !(i.flags & (k.SymbolFlags.Function | k.SymbolFlags.Variable)) && function isNewExpressionTarget(e) {
      var t = climbPastPropertyOrElementAccess(e).parent;
      return k.isNewExpression(t) && t.expression === e;
    }(e)) {
      for (var u of i.declarations) {
        if (k.isClassLike(u)) {
          return u;
        }
      }
    } else if (function isCallOrNewExpressionTarget(e) {
      var t = climbPastPropertyOrElementAccess(e).parent;
      return k.isCallOrNewExpression(t) && t.expression === e;
    }(e) || function isNameOfFunctionDeclaration(e) {
      return k.isIdentifier(e) && e.parent && k.isFunctionLike(e.parent) && e.parent.name === e;
    }(e)) {
      for (var c of i.declarations) {
        if (k.isFunctionLike(c) && c.body && isValueDeclaration(c)) {
          return c;
        }
      }
    }
    if (i.valueDeclaration && k.isPropertyAccessExpression(i.valueDeclaration)) {
      var f = i.valueDeclaration.parent;
      if (f && k.isBinaryExpression(f) && f.left === i.valueDeclaration) {
        return f;
      }
    }
    if (i.valueDeclaration && isValueDeclaration(i.valueDeclaration)) {
      return i.valueDeclaration;
    }
    for (var h of i.declarations) {
      if (isValueDeclaration(h)) {
        return h;
      }
    }
  }
  return;
}

function getValueOfIdentifier(e, t) {
  while (k.isIdentifier(e)) {
    var i = getDeclarationOfIdentifier(e, t);
    if (!i) {
      return;
    } else {
      var r = getValueOfValueDeclaration(i);
      if (r && k.isIdentifier(r) && r !== e) {
        e = r;
      } else {
        return r;
      }
    }
  }
}

function resolveTemplate(e, t, i) {
  if (k.isStringLiteralLike(e)) {
    return {
      combinedText: e.getText().slice(1, -1),
      resolvedSpans: []
    };
  }
  var r = e.template.getText().slice(1, -1);
  if (k.isNoSubstitutionTemplateLiteral(e.template) || 0 === e.template.templateSpans.length) {
    return {
      combinedText: r,
      resolvedSpans: []
    };
  }
  var a = 0;
  var n = e.template.templateSpans.map(e => {
    if (k.isIdentifier(e.expression)) {
      var t = i.languageService.getProgram()?.getTypeChecker();
      if (!t) {
        return;
      }
      var n = getDeclarationOfIdentifier(e.expression, t);
      if (!n) {
        return;
      }
      var s = n;
      if (k.isVariableDeclaration(s)) {
        var o = e.expression.escapedText;
        var l = getValueOfValueDeclaration(s);
        if (!l) {
          return;
        }
        var u = e.expression.getStart() - 2;
        var c = {
          start: u,
          length: e.expression.end - u + 1
        };
        if (k.isTaggedTemplateExpression(l)) {
          var f = resolveTemplate(l, s.getSourceFile().fileName, i);
          r = r.replace("${" + e.expression.escapedText + "}", f.combinedText);
          var h = {
            lines: f.combinedText.split("\n").length,
            identifier: o,
            original: c,
            new: {
              start: c.start + a,
              length: f.combinedText.length
            }
          };
          a += f.combinedText.length - c.length;
          return h;
        } else if (k.isAsExpression(l) && k.isTaggedTemplateExpression(l.expression)) {
          var d = resolveTemplate(l.expression, s.getSourceFile().fileName, i);
          r = r.replace("${" + e.expression.escapedText + "}", d.combinedText);
          var v = {
            lines: d.combinedText.split("\n").length,
            identifier: o,
            original: c,
            new: {
              start: c.start + a,
              length: d.combinedText.length
            }
          };
          a += d.combinedText.length - c.length;
          return v;
        } else if (k.isAsExpression(l) && k.isAsExpression(l.expression) && k.isObjectLiteralExpression(l.expression.expression)) {
          var g = print(JSON.parse(l.expression.expression.getText()));
          r = r.replace("${" + e.expression.escapedText + "}", g);
          var m = {
            lines: g.split("\n").length,
            identifier: o,
            original: c,
            new: {
              start: c.start + a,
              length: g.length
            }
          };
          a += g.length - c.length;
          return m;
        }
        return;
      }
    }
    return;
  }).filter(Boolean);
  return {
    combinedText: r,
    resolvedSpans: n
  };
}

var resolveTadaFragmentArray = e => {
  if (!e) {
    return;
  }
  while (k.isAsExpression(e)) {
    e = e.expression;
  }
  if (!k.isArrayLiteralExpression(e)) {
    return;
  }
  if (e.elements.every(k.isIdentifier)) {
    return e.elements;
  }
  var t = [];
  for (var i of e.elements) {
    while (k.isPropertyAccessExpression(i)) {
      i = i.name;
    }
    if (k.isIdentifier(i)) {
      t.push(i);
    }
  }
  return t;
};

function getSource(e, t) {
  var i = e.languageService.getProgram();
  if (!i) {
    return;
  }
  var r = i.getSourceFile(t);
  if (!r) {
    return;
  }
  return r;
}

function findNode(e, t) {
  return function find(e) {
    if (t >= e.getStart() && t < e.getEnd()) {
      return k.forEachChild(e, find) || e;
    }
  }(e);
}

function findAllTaggedTemplateNodes(e) {
  var t = [];
  !function find(e) {
    if (isGraphQLTag(e) || k.isNoSubstitutionTemplateLiteral(e) && isGraphQLTag(e.parent)) {
      t.push(e);
      return;
    } else {
      k.forEachChild(e, find);
    }
  }(e);
  return t;
}

function unrollFragment(e, t, i) {
  var r = [];
  var a = [ e ];
  var n = new WeakSet;
  var _unrollElement = e => {
    if (n.has(e)) {
      return;
    }
    n.add(e);
    var t = function resolveIdentifierToGraphQLCall(e, t, i) {
      if (!i) {
        return null;
      }
      var r = getValueOfIdentifier(e, i);
      if (!r) {
        return null;
      }
      return isGraphQLCall(r, i) ? r : null;
    }(e, 0, i);
    if (!t) {
      return;
    }
    var s = resolveTadaFragmentArray(t.arguments[1]);
    if (s) {
      a.push(...s);
    }
    try {
      x(t.arguments[0].getText().slice(1, -1), {
        noLocation: !0
      }).definitions.forEach(e => {
        if ("FragmentDefinition" === e.kind) {
          r.push(e);
        }
      });
    } catch (e) {}
  };
  var s;
  while (void 0 !== (s = a.shift())) {
    _unrollElement(s);
  }
  return r;
}

function unrollTadaFragments(e, t, i) {
  var r = i.languageService.getProgram()?.getTypeChecker();
  e.elements.forEach(e => {
    if (k.isIdentifier(e)) {
      t.push(...unrollFragment(e, 0, r));
    } else if (k.isPropertyAccessExpression(e)) {
      if (k.isIdentifier(e.name)) {
        t.push(...unrollFragment(e.name, 0, r));
      }
    }
  });
  return t;
}

function findAllCallExpressions(e, t, i = !0) {
  var {searchExternal: r = !0, collectFragments: a = !0} = "boolean" == typeof i ? {
    searchExternal: i
  } : i;
  var n = t.languageService.getProgram()?.getTypeChecker();
  var s = [];
  var o = [];
  var l = r ? !1 : !0;
  var u = new Map;
  var unrollFragmentMemoized = e => {
    var t = n && n.getSymbolAtLocation(e);
    if (!t) {
      return unrollFragment(e, 0, n);
    }
    var i = u.get(t);
    if (!i) {
      u.set(t, i = unrollFragment(e, 0, n));
    }
    return i;
  };
  !function find(e) {
    if (!k.isCallExpression(e) || isIIFE(e)) {
      return k.forEachChild(e, find);
    }
    if (!isGraphQLCall(e, n)) {
      return k.forEachChild(e, find);
    }
    var i = getSchemaName(e, n);
    var r = e.arguments[0];
    var u = resolveTadaFragmentArray(e.arguments[1]);
    var c = isTadaGraphQLCall(e, n);
    if (!a) {} else if (!l && !u) {
      if (!c) {
        l = !0;
        o.push(...getAllFragments(e, t));
      }
    } else if (u) {
      for (var f of u) {
        o.push(...unrollFragmentMemoized(f));
      }
    }
    if (r && k.isStringLiteralLike(r)) {
      s.push({
        node: r,
        schema: i,
        tadaFragmentRefs: c ? void 0 === u ? [] : u : void 0
      });
    }
  }(e);
  return {
    nodes: s,
    fragments: o
  };
}

function findAllPersistedCallExpressions(e, t) {
  var i = [];
  var r = t?.languageService.getProgram()?.getTypeChecker();
  !function find(e) {
    if (!k.isCallExpression(e) || isIIFE(e)) {
      return k.forEachChild(e, find);
    }
    if (!isTadaPersistedCall(e, r)) {
      return;
    } else if (t) {
      var a = getSchemaName(e, r, !0);
      i.push({
        node: e,
        schema: a
      });
    } else {
      i.push(e);
    }
  }(e);
  return i;
}

function getAllFragments(e, t) {
  var i = [];
  var r = t.languageService.getProgram()?.getTypeChecker();
  if (!k.isCallExpression(e)) {
    return i;
  }
  var a = resolveTadaFragmentArray(e.arguments[1]);
  if (a) {
    var n = t.languageService.getProgram()?.getTypeChecker();
    for (var s of a) {
      i.push(...unrollFragment(s, 0, n));
    }
    return i;
  } else if (isTadaGraphQLCall(e, r)) {
    return i;
  }
  if (!r) {
    return i;
  }
  var o = function getIdentifierOfChainExpression(e) {
    var t = e;
    while (t) {
      if (k.isPropertyAccessExpression(t)) {
        t = t.name;
      } else if (k.isAsExpression(t) || k.isSatisfiesExpression(t) || k.isNonNullExpression(t) || k.isParenthesizedExpression(t) || k.isExpressionWithTypeArguments(t)) {
        t = t.expression;
      } else if (k.isCommaListExpression(t)) {
        t = t.elements[t.elements.length - 1];
      } else if (k.isIdentifier(t)) {
        return t;
      } else {
        return;
      }
    }
  }(e.expression);
  if (!o) {
    return i;
  }
  var l = getDeclarationOfIdentifier(o, r);
  if (!l) {
    return i;
  }
  var u = l.getSourceFile();
  if (!u) {
    return i;
  }
  var c = [ {
    fileName: u.fileName,
    textSpan: {
      start: l.getStart(),
      length: l.getWidth()
    }
  } ];
  if (!c || !c.length) {
    return i;
  }
  var f = c[0];
  if (!f) {
    return i;
  }
  var h = getSource(t, f.fileName);
  if (!h) {
    return i;
  }
  k.forEachChild(h, e => {
    if (k.isVariableStatement(e) && e.declarationList && e.declarationList.declarations[0] && "documents" === e.declarationList.declarations[0].name.getText()) {
      var [t] = e.declarationList.declarations;
      if (t.initializer && k.isObjectLiteralExpression(t.initializer)) {
        t.initializer.properties.forEach(e => {
          if (k.isPropertyAssignment(e) && k.isStringLiteral(e.name)) {
            try {
              var t = JSON.parse(`${e.name.getText().replace(/'/g, '"')}`);
              if (t.includes("fragment ") && t.includes(" on ")) {
                x(t, {
                  noLocation: !0
                }).definitions.forEach(e => {
                  if ("FragmentDefinition" === e.kind) {
                    i.push(e);
                  }
                });
              }
            } catch (e) {}
          }
        });
      }
    }
  });
  return i;
}

function findAllImports(e) {
  return e.statements.filter(k.isImportDeclaration);
}

function bubbleUpTemplate(e) {
  while (k.isNoSubstitutionTemplateLiteral(e) || k.isToken(e) || k.isTemplateExpression(e) || k.isTemplateSpan(e)) {
    e = e.parent;
  }
  return e;
}

function bubbleUpCallExpression(e) {
  while (k.isStringLiteralLike(e) || k.isToken(e) || k.isTemplateExpression(e) || k.isTemplateSpan(e)) {
    e = e.parent;
  }
  return e;
}

var Q = "object" == typeof performance && performance && "function" == typeof performance.now ? performance : Date;

var J = new Set;

var X = "object" == typeof process && process ? process : {};

var emitWarning = (e, t, i, r) => {
  "function" == typeof X.emitWarning ? X.emitWarning(e, t, i, r) : console.error(`[${i}] ${t}: ${e}`);
};

var Y = globalThis.AbortController;

var H = globalThis.AbortSignal;

if (void 0 === Y) {
  H = class AbortSignal {
    _onabort=[];
    aborted=!1;
    addEventListener(e, t) {
      this._onabort.push(t);
    }
  };
  Y = class AbortController {
    constructor() {
      warnACPolyfill();
    }
    signal=new H;
    abort(e) {
      if (this.signal.aborted) {
        return;
      }
      this.signal.reason = e;
      this.signal.aborted = !0;
      for (var t of this.signal._onabort) {
        t(e);
      }
      this.signal.onabort?.(e);
    }
  };
  var Z = "1" !== X.env?.LRU_CACHE_IGNORE_AC_WARNING;
  var warnACPolyfill = () => {
    if (!Z) {
      return;
    }
    Z = !1;
    emitWarning("AbortController is not defined. If using lru-cache in node 14, load an AbortController polyfill from the `node-abort-controller` package. A minimal polyfill is provided for use by LRUCache.fetch(), but it should not be relied upon in other contexts (eg, passing it to other APIs that use AbortController/AbortSignal might have undesirable effects). You may disable this with LRU_CACHE_IGNORE_AC_WARNING=1 in the env.", "NO_ABORT_CONTROLLER", "ENOTSUP", warnACPolyfill);
  };
}

var isPosInt = e => e && e === Math.floor(e) && e > 0 && isFinite(e);

var getUintArray = e => !isPosInt(e) ? null : e <= Math.pow(2, 8) ? Uint8Array : e <= Math.pow(2, 16) ? Uint16Array : e <= Math.pow(2, 32) ? Uint32Array : e <= Number.MAX_SAFE_INTEGER ? ZeroArray : null;

class ZeroArray extends Array {
  constructor(e) {
    super(e);
    this.fill(0);
  }
}

class Stack {
  static #e=!1;
  static create(e) {
    var t = getUintArray(e);
    if (!t) {
      return [];
    }
    Stack.#e = !0;
    var i = new Stack(e, t);
    Stack.#e = !1;
    return i;
  }
  constructor(e, t) {
    if (!Stack.#e) {
      throw new TypeError("instantiate Stack using Stack.create(n)");
    }
    this.heap = new t(e);
    this.length = 0;
  }
  push(e) {
    this.heap[this.length++] = e;
  }
  pop() {
    return this.heap[--this.length];
  }
}

class LRUCache {
  #t;
  #i;
  #r;
  #a;
  #n;
  #s;
  #o;
  #l;
  #u;
  #c;
  #f;
  #h;
  #d;
  #p;
  #v;
  #g;
  #m;
  #S;
  #y;
  #x;
  #T;
  #E;
  static unsafeExposeInternals(e) {
    return {
      starts: e.#S,
      ttls: e.#y,
      sizes: e.#m,
      keyMap: e.#l,
      keyList: e.#u,
      valList: e.#c,
      next: e.#f,
      prev: e.#h,
      get head() {
        return e.#d;
      },
      get tail() {
        return e.#p;
      },
      free: e.#v,
      isBackgroundFetch: t => e.#A(t),
      backgroundFetch: (t, i, r, a) => e.#b(t, i, r, a),
      moveToTail: t => e.#k(t),
      indexes: t => e.#C(t),
      rindexes: t => e.#w(t),
      isStale: t => e.#D(t)
    };
  }
  get max() {
    return this.#t;
  }
  get maxSize() {
    return this.#i;
  }
  get calculatedSize() {
    return this.#o;
  }
  get size() {
    return this.#s;
  }
  get fetchMethod() {
    return this.#n;
  }
  get dispose() {
    return this.#r;
  }
  get disposeAfter() {
    return this.#a;
  }
  constructor(e) {
    var {max: t = 0, ttl: i, ttlResolution: r = 1, ttlAutopurge: a, updateAgeOnGet: n, updateAgeOnHas: s, allowStale: o, dispose: l, disposeAfter: u, noDisposeOnSet: c, noUpdateTTL: f, maxSize: h = 0, maxEntrySize: d = 0, sizeCalculation: v, fetchMethod: g, noDeleteOnFetchRejection: m, noDeleteOnStaleGet: S, allowStaleOnFetchRejection: y, allowStaleOnFetchAbort: x, ignoreFetchAbort: T} = e;
    if (0 !== t && !isPosInt(t)) {
      throw new TypeError("max option must be a nonnegative integer");
    }
    var E = t ? getUintArray(t) : Array;
    if (!E) {
      throw new Error("invalid max value: " + t);
    }
    this.#t = t;
    this.#i = h;
    this.maxEntrySize = d || this.#i;
    this.sizeCalculation = v;
    if (this.sizeCalculation) {
      if (!this.#i && !this.maxEntrySize) {
        throw new TypeError("cannot set sizeCalculation without setting maxSize or maxEntrySize");
      }
      if ("function" != typeof this.sizeCalculation) {
        throw new TypeError("sizeCalculation set to non-function");
      }
    }
    if (void 0 !== g && "function" != typeof g) {
      throw new TypeError("fetchMethod must be a function if specified");
    }
    this.#n = g;
    this.#T = !!g;
    this.#l = new Map;
    this.#u = new Array(t).fill(void 0);
    this.#c = new Array(t).fill(void 0);
    this.#f = new E(t);
    this.#h = new E(t);
    this.#d = 0;
    this.#p = 0;
    this.#v = Stack.create(t);
    this.#s = 0;
    this.#o = 0;
    if ("function" == typeof l) {
      this.#r = l;
    }
    if ("function" == typeof u) {
      this.#a = u;
      this.#g = [];
    } else {
      this.#a = void 0;
      this.#g = void 0;
    }
    this.#x = !!this.#r;
    this.#E = !!this.#a;
    this.noDisposeOnSet = !!c;
    this.noUpdateTTL = !!f;
    this.noDeleteOnFetchRejection = !!m;
    this.allowStaleOnFetchRejection = !!y;
    this.allowStaleOnFetchAbort = !!x;
    this.ignoreFetchAbort = !!T;
    if (0 !== this.maxEntrySize) {
      if (0 !== this.#i) {
        if (!isPosInt(this.#i)) {
          throw new TypeError("maxSize must be a positive integer if specified");
        }
      }
      if (!isPosInt(this.maxEntrySize)) {
        throw new TypeError("maxEntrySize must be a positive integer if specified");
      }
      this.#L();
    }
    this.allowStale = !!o;
    this.noDeleteOnStaleGet = !!S;
    this.updateAgeOnGet = !!n;
    this.updateAgeOnHas = !!s;
    this.ttlResolution = isPosInt(r) || 0 === r ? r : 1;
    this.ttlAutopurge = !!a;
    this.ttl = i || 0;
    if (this.ttl) {
      if (!isPosInt(this.ttl)) {
        throw new TypeError("ttl must be a positive integer if specified");
      }
      this.#F();
    }
    if (0 === this.#t && 0 === this.ttl && 0 === this.#i) {
      throw new TypeError("At least one of max, maxSize, or ttl is required");
    }
    if (!this.ttlAutopurge && !this.#t && !this.#i) {
      var A = "LRU_CACHE_UNBOUNDED";
      if ((e => !J.has(e))(A)) {
        J.add(A);
        emitWarning("TTL caching without ttlAutopurge, max, or maxSize can result in unbounded memory consumption.", "UnboundedCacheWarning", A, LRUCache);
      }
    }
  }
  getRemainingTTL(e) {
    return this.#l.has(e) ? 1 / 0 : 0;
  }
  #F() {
    var e = new ZeroArray(this.#t);
    var t = new ZeroArray(this.#t);
    this.#y = e;
    this.#S = t;
    this.#N = (i, r, a = Q.now()) => {
      t[i] = 0 !== r ? a : 0;
      e[i] = r;
      if (0 !== r && this.ttlAutopurge) {
        var n = setTimeout(() => {
          if (this.#D(i)) {
            this.delete(this.#u[i]);
          }
        }, r + 1);
        if (n.unref) {
          n.unref();
        }
      }
    };
    this.#_ = i => {
      t[i] = 0 !== e[i] ? Q.now() : 0;
    };
    this.#I = (r, a) => {
      if (e[a]) {
        var n = e[a];
        var s = t[a];
        r.ttl = n;
        r.start = s;
        r.now = i || getNow();
        r.remainingTTL = n - (r.now - s);
      }
    };
    var i = 0;
    var getNow = () => {
      var e = Q.now();
      if (this.ttlResolution > 0) {
        i = e;
        var t = setTimeout(() => i = 0, this.ttlResolution);
        if (t.unref) {
          t.unref();
        }
      }
      return e;
    };
    this.getRemainingTTL = r => {
      var a = this.#l.get(r);
      if (void 0 === a) {
        return 0;
      }
      var n = e[a];
      var s = t[a];
      if (0 === n || 0 === s) {
        return 1 / 0;
      }
      return n - ((i || getNow()) - s);
    };
    this.#D = r => 0 !== e[r] && 0 !== t[r] && (i || getNow()) - t[r] > e[r];
  }
  #_=() => {};
  #I=() => {};
  #N=() => {};
  #D=() => !1;
  #L() {
    var e = new ZeroArray(this.#t);
    this.#o = 0;
    this.#m = e;
    this.#O = t => {
      this.#o -= e[t];
      e[t] = 0;
    };
    this.#z = (e, t, i, r) => {
      if (this.#A(t)) {
        return 0;
      }
      if (!isPosInt(i)) {
        if (r) {
          if ("function" != typeof r) {
            throw new TypeError("sizeCalculation must be a function");
          }
          i = r(t, e);
          if (!isPosInt(i)) {
            throw new TypeError("sizeCalculation return invalid (expect positive integer)");
          }
        } else {
          throw new TypeError("invalid size value (must be positive integer). When maxSize or maxEntrySize is used, sizeCalculation or size must be set.");
        }
      }
      return i;
    };
    this.#P = (t, i, r) => {
      e[t] = i;
      if (this.#i) {
        var a = this.#i - e[t];
        while (this.#o > a) {
          this.#V(!0);
        }
      }
      this.#o += e[t];
      if (r) {
        r.entrySize = i;
        r.totalCalculatedSize = this.#o;
      }
    };
  }
  #O=e => {};
  #P=(e, t, i) => {};
  #z=(e, t, i, r) => {
    if (i || r) {
      throw new TypeError("cannot set size without setting maxSize or maxEntrySize on cache");
    }
    return 0;
  };
  * #C({allowStale: e = this.allowStale} = {}) {
    if (this.#s) {
      for (var t = this.#p; 1; ) {
        if (!this.#R(t)) {
          break;
        }
        if (e || !this.#D(t)) {
          yield t;
        }
        if (t === this.#d) {
          break;
        } else {
          t = this.#h[t];
        }
      }
    }
  }
  * #w({allowStale: e = this.allowStale} = {}) {
    if (this.#s) {
      for (var t = this.#d; 1; ) {
        if (!this.#R(t)) {
          break;
        }
        if (e || !this.#D(t)) {
          yield t;
        }
        if (t === this.#p) {
          break;
        } else {
          t = this.#f[t];
        }
      }
    }
  }
  #R(e) {
    return void 0 !== e && this.#l.get(this.#u[e]) === e;
  }
  * entries() {
    for (var e of this.#C()) {
      if (void 0 !== this.#c[e] && void 0 !== this.#u[e] && !this.#A(this.#c[e])) {
        yield [ this.#u[e], this.#c[e] ];
      }
    }
  }
  * rentries() {
    for (var e of this.#w()) {
      if (void 0 !== this.#c[e] && void 0 !== this.#u[e] && !this.#A(this.#c[e])) {
        yield [ this.#u[e], this.#c[e] ];
      }
    }
  }
  * keys() {
    for (var e of this.#C()) {
      var t = this.#u[e];
      if (void 0 !== t && !this.#A(this.#c[e])) {
        yield t;
      }
    }
  }
  * rkeys() {
    for (var e of this.#w()) {
      var t = this.#u[e];
      if (void 0 !== t && !this.#A(this.#c[e])) {
        yield t;
      }
    }
  }
  * values() {
    for (var e of this.#C()) {
      if (void 0 !== this.#c[e] && !this.#A(this.#c[e])) {
        yield this.#c[e];
      }
    }
  }
  * rvalues() {
    for (var e of this.#w()) {
      if (void 0 !== this.#c[e] && !this.#A(this.#c[e])) {
        yield this.#c[e];
      }
    }
  }
  [Symbol.iterator]() {
    return this.entries();
  }
  find(e, t = {}) {
    for (var i of this.#C()) {
      var r = this.#c[i];
      var a = this.#A(r) ? r.__staleWhileFetching : r;
      if (void 0 === a) {
        continue;
      }
      if (e(a, this.#u[i], this)) {
        return this.get(this.#u[i], t);
      }
    }
  }
  forEach(e, t = this) {
    for (var i of this.#C()) {
      var r = this.#c[i];
      var a = this.#A(r) ? r.__staleWhileFetching : r;
      if (void 0 === a) {
        continue;
      }
      e.call(t, a, this.#u[i], this);
    }
  }
  rforEach(e, t = this) {
    for (var i of this.#w()) {
      var r = this.#c[i];
      var a = this.#A(r) ? r.__staleWhileFetching : r;
      if (void 0 === a) {
        continue;
      }
      e.call(t, a, this.#u[i], this);
    }
  }
  purgeStale() {
    var e = !1;
    for (var t of this.#w({
      allowStale: !0
    })) {
      if (this.#D(t)) {
        this.delete(this.#u[t]);
        e = !0;
      }
    }
    return e;
  }
  dump() {
    var e = [];
    for (var t of this.#C({
      allowStale: !0
    })) {
      var i = this.#u[t];
      var r = this.#c[t];
      var a = this.#A(r) ? r.__staleWhileFetching : r;
      if (void 0 === a || void 0 === i) {
        continue;
      }
      var n = {
        value: a
      };
      if (this.#y && this.#S) {
        n.ttl = this.#y[t];
        var s = Q.now() - this.#S[t];
        n.start = Math.floor(Date.now() - s);
      }
      if (this.#m) {
        n.size = this.#m[t];
      }
      e.unshift([ i, n ]);
    }
    return e;
  }
  load(e) {
    this.clear();
    for (var [t, i] of e) {
      if (i.start) {
        var r = Date.now() - i.start;
        i.start = Q.now() - r;
      }
      this.set(t, i.value, i);
    }
  }
  set(e, t, i = {}) {
    if (void 0 === t) {
      this.delete(e);
      return this;
    }
    var {ttl: r = this.ttl, start: a, noDisposeOnSet: n = this.noDisposeOnSet, sizeCalculation: s = this.sizeCalculation, status: o} = i;
    var {noUpdateTTL: l = this.noUpdateTTL} = i;
    var u = this.#z(e, t, i.size || 0, s);
    if (this.maxEntrySize && u > this.maxEntrySize) {
      if (o) {
        o.set = "miss";
        o.maxEntrySizeExceeded = !0;
      }
      this.delete(e);
      return this;
    }
    var c = 0 === this.#s ? void 0 : this.#l.get(e);
    if (void 0 === c) {
      c = 0 === this.#s ? this.#p : 0 !== this.#v.length ? this.#v.pop() : this.#s === this.#t ? this.#V(!1) : this.#s;
      this.#u[c] = e;
      this.#c[c] = t;
      this.#l.set(e, c);
      this.#f[this.#p] = c;
      this.#h[c] = this.#p;
      this.#p = c;
      this.#s++;
      this.#P(c, u, o);
      if (o) {
        o.set = "add";
      }
      l = !1;
    } else {
      this.#k(c);
      var f = this.#c[c];
      if (t !== f) {
        if (this.#T && this.#A(f)) {
          f.__abortController.abort(new Error("replaced"));
          var {__staleWhileFetching: h} = f;
          if (void 0 !== h && !n) {
            if (this.#x) {
              this.#r?.(h, e, "set");
            }
            if (this.#E) {
              this.#g?.push([ h, e, "set" ]);
            }
          }
        } else if (!n) {
          if (this.#x) {
            this.#r?.(f, e, "set");
          }
          if (this.#E) {
            this.#g?.push([ f, e, "set" ]);
          }
        }
        this.#O(c);
        this.#P(c, u, o);
        this.#c[c] = t;
        if (o) {
          o.set = "replace";
          var d = f && this.#A(f) ? f.__staleWhileFetching : f;
          if (void 0 !== d) {
            o.oldValue = d;
          }
        }
      } else if (o) {
        o.set = "update";
      }
    }
    if (0 !== r && !this.#y) {
      this.#F();
    }
    if (this.#y) {
      if (!l) {
        this.#N(c, r, a);
      }
      if (o) {
        this.#I(o, c);
      }
    }
    if (!n && this.#E && this.#g) {
      var v = this.#g;
      var g;
      while (g = v?.shift()) {
        this.#a?.(...g);
      }
    }
    return this;
  }
  pop() {
    try {
      while (this.#s) {
        var e = this.#c[this.#d];
        this.#V(!0);
        if (this.#A(e)) {
          if (e.__staleWhileFetching) {
            return e.__staleWhileFetching;
          }
        } else if (void 0 !== e) {
          return e;
        }
      }
    } finally {
      if (this.#E && this.#g) {
        var t = this.#g;
        var i;
        while (i = t?.shift()) {
          this.#a?.(...i);
        }
      }
    }
  }
  #V(e) {
    var t = this.#d;
    var i = this.#u[t];
    var r = this.#c[t];
    if (this.#T && this.#A(r)) {
      r.__abortController.abort(new Error("evicted"));
    } else if (this.#x || this.#E) {
      if (this.#x) {
        this.#r?.(r, i, "evict");
      }
      if (this.#E) {
        this.#g?.push([ r, i, "evict" ]);
      }
    }
    this.#O(t);
    if (e) {
      this.#u[t] = void 0;
      this.#c[t] = void 0;
      this.#v.push(t);
    }
    if (1 === this.#s) {
      this.#d = this.#p = 0;
      this.#v.length = 0;
    } else {
      this.#d = this.#f[t];
    }
    this.#l.delete(i);
    this.#s--;
    return t;
  }
  has(e, t = {}) {
    var {updateAgeOnHas: i = this.updateAgeOnHas, status: r} = t;
    var a = this.#l.get(e);
    if (void 0 !== a) {
      var n = this.#c[a];
      if (this.#A(n) && void 0 === n.__staleWhileFetching) {
        return !1;
      }
      if (!this.#D(a)) {
        if (i) {
          this.#_(a);
        }
        if (r) {
          r.has = "hit";
          this.#I(r, a);
        }
        return !0;
      } else if (r) {
        r.has = "stale";
        this.#I(r, a);
      }
    } else if (r) {
      r.has = "miss";
    }
    return !1;
  }
  peek(e, t = {}) {
    var {allowStale: i = this.allowStale} = t;
    var r = this.#l.get(e);
    if (void 0 !== r && (i || !this.#D(r))) {
      var a = this.#c[r];
      return this.#A(a) ? a.__staleWhileFetching : a;
    }
  }
  #b(e, t, i, r) {
    var a = void 0 === t ? void 0 : this.#c[t];
    if (this.#A(a)) {
      return a;
    }
    var n = new Y;
    var {signal: s} = i;
    s?.addEventListener("abort", () => n.abort(s.reason), {
      signal: n.signal
    });
    var o = {
      signal: n.signal,
      options: i,
      context: r
    };
    var cb = (r, a = !1) => {
      var {aborted: s} = n.signal;
      var u = i.ignoreFetchAbort && void 0 !== r;
      if (i.status) {
        if (s && !a) {
          i.status.fetchAborted = !0;
          i.status.fetchError = n.signal.reason;
          if (u) {
            i.status.fetchAbortIgnored = !0;
          }
        } else {
          i.status.fetchResolved = !0;
        }
      }
      if (s && !u && !a) {
        return fetchFail(n.signal.reason);
      }
      if (this.#c[t] === l) {
        if (void 0 === r) {
          if (l.__staleWhileFetching) {
            this.#c[t] = l.__staleWhileFetching;
          } else {
            this.delete(e);
          }
        } else {
          if (i.status) {
            i.status.fetchUpdated = !0;
          }
          this.set(e, r, o.options);
        }
      }
      return r;
    };
    var fetchFail = r => {
      var {aborted: a} = n.signal;
      var s = a && i.allowStaleOnFetchAbort;
      var o = s || i.allowStaleOnFetchRejection;
      var u = l;
      if (this.#c[t] === l) {
        if (!(o || i.noDeleteOnFetchRejection) || void 0 === u.__staleWhileFetching) {
          this.delete(e);
        } else if (!s) {
          this.#c[t] = u.__staleWhileFetching;
        }
      }
      if (o) {
        if (i.status && void 0 !== u.__staleWhileFetching) {
          i.status.returnedStale = !0;
        }
        return u.__staleWhileFetching;
      } else if (u.__returned === u) {
        throw r;
      }
    };
    if (i.status) {
      i.status.fetchDispatched = !0;
    }
    var l = new Promise((t, r) => {
      var s = this.#n?.(e, a, o);
      if (s && s instanceof Promise) {
        s.then(e => t(void 0 === e ? void 0 : e), r);
      }
      n.signal.addEventListener("abort", () => {
        if (!i.ignoreFetchAbort || i.allowStaleOnFetchAbort) {
          t(void 0);
          if (i.allowStaleOnFetchAbort) {
            t = e => cb(e, !0);
          }
        }
      });
    }).then(cb, e => {
      if (i.status) {
        i.status.fetchRejected = !0;
        i.status.fetchError = e;
      }
      return fetchFail(e);
    });
    var u = Object.assign(l, {
      __abortController: n,
      __staleWhileFetching: a,
      __returned: void 0
    });
    if (void 0 === t) {
      this.set(e, u, {
        ...o.options,
        status: void 0
      });
      t = this.#l.get(e);
    } else {
      this.#c[t] = u;
    }
    return u;
  }
  #A(e) {
    if (!this.#T) {
      return !1;
    }
    var t = e;
    return !!t && t instanceof Promise && t.hasOwnProperty("__staleWhileFetching") && t.__abortController instanceof Y;
  }
  async fetch(e, t = {}) {
    var {allowStale: i = this.allowStale, updateAgeOnGet: r = this.updateAgeOnGet, noDeleteOnStaleGet: a = this.noDeleteOnStaleGet, ttl: n = this.ttl, noDisposeOnSet: s = this.noDisposeOnSet, size: o = 0, sizeCalculation: l = this.sizeCalculation, noUpdateTTL: u = this.noUpdateTTL, noDeleteOnFetchRejection: c = this.noDeleteOnFetchRejection, allowStaleOnFetchRejection: f = this.allowStaleOnFetchRejection, ignoreFetchAbort: h = this.ignoreFetchAbort, allowStaleOnFetchAbort: d = this.allowStaleOnFetchAbort, context: v, forceRefresh: g = !1, status: m, signal: S} = t;
    if (!this.#T) {
      if (m) {
        m.fetch = "get";
      }
      return this.get(e, {
        allowStale: i,
        updateAgeOnGet: r,
        noDeleteOnStaleGet: a,
        status: m
      });
    }
    var y = {
      allowStale: i,
      updateAgeOnGet: r,
      noDeleteOnStaleGet: a,
      ttl: n,
      noDisposeOnSet: s,
      size: o,
      sizeCalculation: l,
      noUpdateTTL: u,
      noDeleteOnFetchRejection: c,
      allowStaleOnFetchRejection: f,
      allowStaleOnFetchAbort: d,
      ignoreFetchAbort: h,
      status: m,
      signal: S
    };
    var x = this.#l.get(e);
    if (void 0 === x) {
      if (m) {
        m.fetch = "miss";
      }
      var T = this.#b(e, x, y, v);
      return T.__returned = T;
    } else {
      var E = this.#c[x];
      if (this.#A(E)) {
        var A = i && void 0 !== E.__staleWhileFetching;
        if (m) {
          m.fetch = "inflight";
          if (A) {
            m.returnedStale = !0;
          }
        }
        return A ? E.__staleWhileFetching : E.__returned = E;
      }
      var b = this.#D(x);
      if (!g && !b) {
        if (m) {
          m.fetch = "hit";
        }
        this.#k(x);
        if (r) {
          this.#_(x);
        }
        if (m) {
          this.#I(m, x);
        }
        return E;
      }
      var k = this.#b(e, x, y, v);
      var C = void 0 !== k.__staleWhileFetching && i;
      if (m) {
        m.fetch = b ? "stale" : "refresh";
        if (C && b) {
          m.returnedStale = !0;
        }
      }
      return C ? k.__staleWhileFetching : k.__returned = k;
    }
  }
  get(e, t = {}) {
    var {allowStale: i = this.allowStale, updateAgeOnGet: r = this.updateAgeOnGet, noDeleteOnStaleGet: a = this.noDeleteOnStaleGet, status: n} = t;
    var s = this.#l.get(e);
    if (void 0 !== s) {
      var o = this.#c[s];
      var l = this.#A(o);
      if (n) {
        this.#I(n, s);
      }
      if (this.#D(s)) {
        if (n) {
          n.get = "stale";
        }
        if (!l) {
          if (!a) {
            this.delete(e);
          }
          if (n && i) {
            n.returnedStale = !0;
          }
          return i ? o : void 0;
        } else {
          if (n && i && void 0 !== o.__staleWhileFetching) {
            n.returnedStale = !0;
          }
          return i ? o.__staleWhileFetching : void 0;
        }
      } else {
        if (n) {
          n.get = "hit";
        }
        if (l) {
          return o.__staleWhileFetching;
        }
        this.#k(s);
        if (r) {
          this.#_(s);
        }
        return o;
      }
    } else if (n) {
      n.get = "miss";
    }
  }
  #M(e, t) {
    this.#h[t] = e;
    this.#f[e] = t;
  }
  #k(e) {
    if (e !== this.#p) {
      if (e === this.#d) {
        this.#d = this.#f[e];
      } else {
        this.#M(this.#h[e], this.#f[e]);
      }
      this.#M(this.#p, e);
      this.#p = e;
    }
  }
  delete(e) {
    var t = !1;
    if (0 !== this.#s) {
      var i = this.#l.get(e);
      if (void 0 !== i) {
        t = !0;
        if (1 === this.#s) {
          this.clear();
        } else {
          this.#O(i);
          var r = this.#c[i];
          if (this.#A(r)) {
            r.__abortController.abort(new Error("deleted"));
          } else if (this.#x || this.#E) {
            if (this.#x) {
              this.#r?.(r, e, "delete");
            }
            if (this.#E) {
              this.#g?.push([ r, e, "delete" ]);
            }
          }
          this.#l.delete(e);
          this.#u[i] = void 0;
          this.#c[i] = void 0;
          if (i === this.#p) {
            this.#p = this.#h[i];
          } else if (i === this.#d) {
            this.#d = this.#f[i];
          } else {
            this.#f[this.#h[i]] = this.#f[i];
            this.#h[this.#f[i]] = this.#h[i];
          }
          this.#s--;
          this.#v.push(i);
        }
      }
    }
    if (this.#E && this.#g?.length) {
      var a = this.#g;
      var n;
      while (n = a?.shift()) {
        this.#a?.(...n);
      }
    }
    return t;
  }
  clear() {
    for (var e of this.#w({
      allowStale: !0
    })) {
      var t = this.#c[e];
      if (this.#A(t)) {
        t.__abortController.abort(new Error("deleted"));
      } else {
        var i = this.#u[e];
        if (this.#x) {
          this.#r?.(t, i, "delete");
        }
        if (this.#E) {
          this.#g?.push([ t, i, "delete" ]);
        }
      }
    }
    this.#l.clear();
    this.#c.fill(void 0);
    this.#u.fill(void 0);
    if (this.#y && this.#S) {
      this.#y.fill(0);
      this.#S.fill(0);
    }
    if (this.#m) {
      this.#m.fill(0);
    }
    this.#d = 0;
    this.#p = 0;
    this.#v.length = 0;
    this.#o = 0;
    this.#s = 0;
    if (this.#E && this.#g) {
      var r = this.#g;
      var a;
      while (a = r?.shift()) {
        this.#a?.(...a);
      }
    }
  }
}

var ee = {
  exports: {}
};

var te = {
  32: 16777619n,
  64: 1099511628211n,
  128: 309485009821345068724781371n,
  256: 374144419156711147060143317175368453031918731002211n,
  512: 35835915874844867368919076489095108449946327955754392558399825615420669938882575126094039892345713852759n,
  1024: 5016456510113118655434598811035278955030765345404790744303017523831112055108147451509157692220295382716162651878526895249385292291816524375083746691371804094271873160484737966720260389217684476157468082573n
};

var ie = {
  32: 2166136261n,
  64: 14695981039346656037n,
  128: 144066263297769815596495629667062367629n,
  256: 100029257958052580907070968620625704837092796014241193945225284501741471925557n,
  512: 9659303129496669498009435400716310466090418745672637896108374329434462657994582932197716438449813051892206539805784495328239340083876191928701583869517785n,
  1024: 14197795064947621068722070641403218320880622795441933960878474914617582723252296732303717722150864096521202355549365628174669108571814760471015076148029755969804077320157692458563003215304957150157403644460363550505412711285966361610267868082893823963790439336411086884584107735010676915n
};

ee.exports = function fnv1a(e) {
  var t = Number(ie[32]);
  var i = !1;
  for (var r = 0; r < e.length; r++) {
    var a = e.charCodeAt(r);
    if (a > 127 && !i) {
      a = (e = unescape(encodeURIComponent(e))).charCodeAt(r);
      i = !0;
    }
    t ^= a;
    t += (t << 1) + (t << 4) + (t << 7) + (t << 8) + (t << 24);
  }
  return t >>> 0;
};

ee.exports.bigInt = function bigInt(e, {size: t = 32} = {}) {
  if (!te[t]) {
    throw new Error("The `size` option must be one of 32, 64, 128, 256, 512, or 1024");
  }
  var i = ie[t];
  var r = te[t];
  var a = !1;
  for (var n = 0; n < e.length; n++) {
    var s = e.charCodeAt(n);
    if (s > 127 && !a) {
      s = (e = unescape(encodeURIComponent(e))).charCodeAt(n);
      a = !0;
    }
    i ^= BigInt(s);
    i = BigInt.asUintN(t, i * r);
  }
  return i;
};

var re = getDefaultExportFromCjs(ee.exports);

var ae = 52005;

var unwrapAbstractType = e => e.isUnionOrIntersection() ? e.types.find(e => e.flags & k.TypeFlags.Object) || e : e;

var getVariableDeclaration = e => {
  var t = e;
  var i = new Set;
  while (t.parent && !i.has(t)) {
    i.add(t);
    if (k.isBlock(t)) {
      return;
    } else if (k.isVariableDeclaration(t = t.parent)) {
      return t;
    }
  }
};

var ne = new Set([ "map", "filter", "forEach", "reduce", "every", "some", "find", "flatMap", "sort" ]);

var resolveDataType = (e, t) => {
  var i;
  var r = t.getTypeAtLocation(e.parent);
  if ("target" in r) {
    var a = r.resolvedTypeArguments;
    i = a && a.length > 1 ? a[0] : void 0;
  }
  if (!i) {
    var n = r.getProperty("__apiType");
    if (n) {
      var s = t.getTypeOfSymbol(n);
      var o = r.getCallSignatures()[0];
      if (s.isUnionOrIntersection()) {
        for (var l of s.types) {
          if (o = l.getCallSignatures()[0]) {
            i = o.getReturnType();
            break;
          }
        }
      }
      i = o && o.getReturnType();
    }
  }
  return i;
};

function getFragmentsInSource(t, i, r) {
  var a = [];
  var n = findAllCallExpressions(t, r, !1);
  var s = i.getSymbolAtLocation(t);
  if (!s) {
    return [];
  }
  var o = i.getExportsOfModule(s).map(e => e.name);
  n.nodes.filter(e => {
    var t = e.node.parent;
    while (t && !k.isSourceFile(t) && !k.isVariableDeclaration(t)) {
      t = t.parent;
    }
    if (k.isVariableDeclaration(t)) {
      return o.includes(t.name.getText());
    } else {
      return !1;
    }
  }).forEach(t => {
    var i = resolveTemplate(t.node, 0, r).combinedText;
    try {
      var n = x(i, {
        noLocation: !0
      });
      if (n.definitions.every(t => t.kind === e.FRAGMENT_DEFINITION)) {
        a = a.concat(n.definitions);
      }
    } catch (e) {
      return;
    }
  });
  return a;
}

function getPersistedCodeFixAtPosition(e, t, i) {
  var r = i.config.templateIsCallExpression ?? !0;
  var a = i.languageService.getProgram()?.getTypeChecker();
  if (!r) {
    return;
  }
  var n = getSource(i, e);
  if (!n) {
    return;
  }
  var s = findNode(n, t);
  if (!s) {
    return;
  }
  var o = s;
  if (k.isVariableStatement(o)) {
    o = o.declarationList.declarations.find(e => k.isVariableDeclaration(e) && e.initializer && k.isCallExpression(e.initializer)) || s;
  } else if (k.isVariableDeclarationList(o)) {
    o = o.declarations.find(e => k.isVariableDeclaration(e) && e.initializer && k.isCallExpression(e.initializer)) || s;
  } else if (k.isVariableDeclaration(o) && o.initializer && k.isCallExpression(o.initializer)) {
    o = o.initializer;
  } else {
    while (o && !k.isCallExpression(o)) {
      o = o.parent;
    }
  }
  if (!isTadaPersistedCall(o, a)) {
    return;
  }
  var l, u = e;
  if (o.typeArguments) {
    var [c] = o.typeArguments;
    if (!c || !k.isTypeQueryNode(c)) {
      return;
    }
    var {node: f, filename: h} = getDocumentReferenceFromTypeQuery(c, e, i);
    l = f;
    u = h;
  } else if (o.arguments[1]) {
    if (!k.isIdentifier(o.arguments[1]) && !k.isCallExpression(o.arguments[1])) {
      return;
    }
    var {node: d, filename: v} = getDocumentReferenceFromDocumentNode(o.arguments[1], e, i);
    l = d;
    u = v;
  }
  if (!l) {
    return;
  }
  var g = l;
  if (!(g && k.isCallExpression(g) && g.arguments[0] && k.isStringLiteralLike(g.arguments[0]))) {
    return;
  }
  var m = generateHashForDocument(i, g.arguments[0], u, g.arguments[1] && k.isArrayLiteralExpression(g.arguments[1]) ? g.arguments[1] : void 0);
  var S = o.arguments[0];
  if (!S) {
    return {
      span: {
        start: o.arguments.pos,
        length: 1
      },
      replacement: `"sha256:${m}")`
    };
  } else if (k.isStringLiteral(S) && S.getText() !== `"sha256:${m}"`) {
    return {
      span: {
        start: S.getStart(),
        length: S.end - S.getStart()
      },
      replacement: `"sha256:${m}"`
    };
  } else if (k.isIdentifier(S)) {
    return {
      span: {
        start: S.getStart(),
        length: S.end - S.getStart()
      },
      replacement: `"sha256:${m}"`
    };
  } else {
    return;
  }
}

var generateHashForDocument = (e, t, i, r) => {
  if (r) {
    var a = [];
    unrollTadaFragments(r, a, e);
    var n = resolveTemplate(t, 0, e).combinedText;
    var s = parse(n);
    var o = new Set;
    for (var l of s.definitions) {
      if (l.kind === z && !o.has(l)) {
        stripUnmaskDirectivesFromDefinition(l);
      }
    }
    a.map(e => {
      stripUnmaskDirectivesFromDefinition(e);
      return print(e);
    }).filter((e, t, i) => i.indexOf(e) === t).forEach(e => {
      n = `${n}\n\n${e}`;
    });
    var u = print(parse(n));
    return b("sha256").update(u).digest("hex");
  } else {
    var c = getSource(e, i);
    var {fragments: f} = findAllCallExpressions(c, e);
    var h = resolveTemplate(t, 0, e).combinedText;
    var d = parse(h);
    var v = new Set;
    for (var g of d.definitions) {
      if (g.kind === z && !v.has(g)) {
        stripUnmaskDirectivesFromDefinition(g);
      }
    }
    var m = new Set;
    visit(d, {
      FragmentDefinition: e => {
        f.push(e);
      },
      FragmentSpread: e => {
        m.add(e.name.value);
      }
    });
    var S = h;
    var y = new Set;
    var x = [ ...m ];
    var T;
    while (T = x.shift()) {
      y.add(T);
      var E = f.find(e => e.name.value === T);
      if (!E) {
        e.project.projectService.logger.info(`[GraphQLSP] could not find fragment for spread ${T}!`);
        return;
      }
      stripUnmaskDirectivesFromDefinition(E);
      visit(E, {
        FragmentSpread: e => {
          if (!y.has(e.name.value)) {
            x.push(e.name.value);
          }
        }
      });
      S = `${S}\n\n${print(E)}`;
    }
    return b("sha256").update(print(parse(S))).digest("hex");
  }
};

var getDocumentReferenceFromTypeQuery = (e, t, i) => {
  var r = i.languageService.getProgram()?.getTypeChecker();
  if (!r) {
    return {
      node: null,
      filename: t
    };
  }
  var a;
  if (k.isIdentifier(e.exprName)) {
    a = e.exprName;
  } else if (k.isQualifiedName(e.exprName)) {
    a = e.exprName.right;
  }
  if (!a) {
    return {
      node: null,
      filename: t
    };
  }
  var n = getValueOfIdentifier(a, r);
  if (!n || !isGraphQLCall(n, r)) {
    return {
      node: null,
      filename: t
    };
  }
  return {
    node: n,
    filename: n.getSourceFile().fileName
  };
};

var getDocumentReferenceFromDocumentNode = (e, t, i) => {
  if (k.isIdentifier(e)) {
    var r = i.languageService.getProgram()?.getTypeChecker();
    if (!r) {
      return {
        node: null,
        filename: t
      };
    }
    var a = getValueOfIdentifier(e, r);
    if (!a || !isGraphQLCall(a, r)) {
      return {
        node: null,
        filename: t
      };
    }
    return {
      node: a,
      filename: a.getSourceFile().fileName
    };
  } else {
    return {
      node: e,
      filename: t
    };
  }
};

var stripUnmaskDirectivesFromDefinition = e => {
  e.directives = e.directives?.filter(e => "_unmask" !== e.name.value);
};

var se = new Set([ "populate", "client", "unmask", "_unmask", "_optional", "_relayPagination", "_simplePagination", "_required", "optional", "required", "arguments", "argumentDefinitions", "connection", "refetchable", "relay", "required", "inline" ]);

var oe = 520100;

var le = 520101;

var ue = 520102;

var ce = 520103;

var fe = [ 52001, 52004, 52003, ae, 52006, 52007, 52008, oe, le, ue, ce ];

var getMisconfigurationDiagnostics = (e, t, i, r) => {
  if (!i.errors || !t.length) {
    return [];
  }
  var a = [];
  var n = t[0].node;
  var s = [ i.errors.config, ...i.errors.load.values(), ...i.errors.write.values() ].filter(Boolean);
  var o = r.languageService.getProgram();
  if (o) {
    for (var [l, u] of i.outputLocations) {
      if (Date.now() - u > 3e4 && !o.getSourceFile(l)) {
        s.push(`The generated typings file "${l}" is not part of the TypeScript project. Check that it is matched by the "include" patterns of your tsconfig.json.`);
      }
    }
  }
  for (var c of s) {
    a.push({
      category: k.DiagnosticCategory.Error,
      code: 52006,
      file: e,
      messageText: c,
      start: n.getStart(),
      length: n.getEnd() - n.getStart()
    });
  }
  if (!!i.current || Object.values(i.multi).some(Boolean)) {
    var f = Object.keys(i.multi);
    for (var {node: h, schema: d} of t) {
      if (d && !f.includes(d)) {
        a.push({
          category: k.DiagnosticCategory.Error,
          code: 52008,
          file: e,
          messageText: `This document refers to the schema named "${d}", which isn't configured. ` + (f.length ? `Configured schemas are: ${f.join(", ")}.` : 'No named schemas are configured in the "schemas" option.'),
          start: h.getStart(),
          length: h.getEnd() - h.getStart()
        });
      }
    }
  }
  return a;
};

var he = new LRUCache({
  ttl: 9e5,
  max: 5e3
});

function getGraphQLDiagnostics(t, i, r) {
  var a = r.config.templateIsCallExpression ?? !0;
  var n = getSource(r, t);
  if (!n) {
    return;
  }
  var s, o = [];
  if (a) {
    var l = findAllCallExpressions(n, r);
    o = l.fragments;
    s = l.nodes;
  } else {
    s = findAllTaggedTemplateNodes(n).map(e => ({
      node: e,
      schema: null
    }));
  }
  var u = s.map(({node: e}) => {
    if ((k.isNoSubstitutionTemplateLiteral(e) || k.isTemplateExpression(e)) && !a) {
      if (k.isTaggedTemplateExpression(e.parent)) {
        e = e.parent;
      } else {
        return;
      }
    }
    return resolveTemplate(e, 0, r).combinedText;
  });
  var c = re(a ? n.getText() + o.map(e => print(e)).join("-") + i.version : u.join("-") + i.version);
  var f;
  if (he.has(c)) {
    f = he.get(c);
  } else {
    f = runDiagnostics(n, {
      nodes: s,
      fragments: o
    }, i, r);
    he.set(c, f);
  }
  f = [ ...getMisconfigurationDiagnostics(n, s, i, r), ...f ];
  if (!s.length) {
    var h = ((e, t, i) => {
      if (!new RegExp(`\\b(?:${Array.from(U).join("|")})\\s*${t ? "`" : "\\("}`).test(e.getText())) {
        return null;
      }
      var r;
      var a;
      if (t) {
        r = findAllTaggedTemplateNodes(e)[0];
        a = 'Found GraphQL documents in tagged templates, but GraphQLSP is configured to search for graphql()/gql() calls. If you use tagged templates, set "templateIsCallExpression": false in the plugin configuration in your tsconfig.json.';
      } else {
        r = findAllCallExpressions(e, i, {
          searchExternal: !1,
          collectFragments: !1
        }).nodes.find(e => /^[\s,]*(?:query|mutation|subscription|fragment|\{)/.test(e.node.text))?.node;
        a = 'Found GraphQL documents in graphql()/gql() calls, but GraphQLSP is configured to search for tagged templates. If you use call expressions, remove "templateIsCallExpression": false from the plugin configuration in your tsconfig.json.';
      }
      if (!r) {
        return null;
      }
      return {
        category: k.DiagnosticCategory.Warning,
        code: 52007,
        file: e,
        messageText: a,
        start: r.getStart(),
        length: r.getEnd() - r.getStart()
      };
    })(n, a, r);
    if (h) {
      f.unshift(h);
    }
  }
  var d = r.config.shouldCheckForColocatedFragments ?? !0;
  var v = [];
  if (a) {
    var g = findAllPersistedCallExpressions(n, r).map(e => {
      var {node: i} = e;
      if (!i.typeArguments && !i.arguments[1]) {
        return {
          category: k.DiagnosticCategory.Warning,
          code: oe,
          file: n,
          messageText: "Missing generic pointing at the GraphQL document.",
          start: i.getStart(),
          length: i.getEnd() - i.getStart()
        };
      }
      var a, s, o, l, u = t;
      var c = i.typeArguments && i.typeArguments[0];
      if (c) {
        o = c.getStart();
        l = c.getEnd() - c.getStart();
        if (!k.isTypeQueryNode(c)) {
          return {
            category: k.DiagnosticCategory.Warning,
            code: oe,
            file: n,
            messageText: "Provided generic should be a typeQueryNode in the shape of graphql.persisted<typeof document>.",
            start: o,
            length: l
          };
        }
        var {node: f, filename: h} = getDocumentReferenceFromTypeQuery(c, t, r);
        a = f;
        u = h;
        s = c.getText();
      } else if (i.arguments[1]) {
        o = i.arguments[1].getStart();
        l = i.arguments[1].getEnd() - i.arguments[1].getStart();
        if (!k.isIdentifier(i.arguments[1]) && !k.isCallExpression(i.arguments[1])) {
          return {
            category: k.DiagnosticCategory.Warning,
            code: oe,
            file: n,
            messageText: 'Provided argument should be an identifier or invocation of "graphql" in the shape of graphql.persisted(hash, document).',
            start: o,
            length: l
          };
        }
        var {node: d, filename: v} = getDocumentReferenceFromDocumentNode(i.arguments[1], t, r);
        a = d;
        u = v;
        s = i.arguments[1].getText();
      }
      if (!a) {
        return {
          category: k.DiagnosticCategory.Warning,
          code: ue,
          file: n,
          messageText: `Can't find reference to "${s}".`,
          start: o,
          length: l
        };
      }
      var g = a;
      if (!(g && k.isCallExpression(g) && g.arguments[0] && k.isStringLiteralLike(g.arguments[0]))) {
        return {
          category: k.DiagnosticCategory.Warning,
          code: ue,
          file: n,
          messageText: `Referenced type "${s}" is not a GraphQL document.`,
          start: o,
          length: l
        };
      }
      if (!i.arguments[0]) {
        return {
          category: k.DiagnosticCategory.Warning,
          code: le,
          file: n,
          messageText: "The call-expression is missing a hash for the persisted argument.",
          start: i.arguments.pos,
          length: i.arguments.end - i.arguments.pos
        };
      }
      var m = i.arguments[0].getText().slice(1, -1);
      if (m.startsWith("sha256:")) {
        var S = generateHashForDocument(r, g.arguments[0], u, g.arguments[1] && k.isArrayLiteralExpression(g.arguments[1]) ? g.arguments[1] : void 0);
        if (!S) {
          return null;
        }
        if (`sha256:${S}` !== m) {
          return {
            category: k.DiagnosticCategory.Warning,
            code: ce,
            file: n,
            messageText: "The persisted document's hash is outdated",
            start: i.arguments.pos,
            length: i.arguments.end - i.arguments.pos
          };
        }
      }
      return null;
    }).filter(Boolean);
    f.push(...g);
  }
  if (a && d) {
    var m = ((e, t) => {
      var i = findAllImports(e);
      var r = t.languageService.getProgram()?.getTypeChecker();
      var a = {};
      if (!r) {
        return a;
      }
      if (i.length) {
        i.forEach(e => {
          if (!e.importClause) {
            return;
          }
          if (e.importClause.name) {
            var i = getDeclarationOfIdentifier(e.importClause.name, r);
            if (i) {
              var n = i.getSourceFile();
              if (n.fileName.includes("node_modules")) {
                return;
              }
              if (!n) {
                return;
              }
              var s = getFragmentsInSource(n, r, t).map(e => e.name.value);
              var o = e.moduleSpecifier.getText();
              var l = a[o];
              if (s.length && l) {
                l.fragments = l.fragments.concat(s);
              } else if (s.length && !l) {
                a[o] = l = {
                  start: e.moduleSpecifier.getStart(),
                  length: e.moduleSpecifier.getText().length,
                  fragments: s
                };
              }
            }
          }
          if (e.importClause.namedBindings && k.isNamespaceImport(e.importClause.namedBindings)) {
            var u = getDeclarationOfIdentifier(e.importClause.namedBindings.name, r);
            if (u) {
              var c = u.getSourceFile();
              if (c.fileName.includes("node_modules")) {
                return;
              }
              if (!c) {
                return;
              }
              var f = getFragmentsInSource(c, r, t).map(e => e.name.value);
              var h = e.moduleSpecifier.getText();
              var d = a[h];
              if (f.length && d) {
                d.fragments = d.fragments.concat(f);
              } else if (f.length && !d) {
                a[h] = d = {
                  start: e.moduleSpecifier.getStart(),
                  length: e.moduleSpecifier.getText().length,
                  fragments: f
                };
              }
            }
          } else if (e.importClause.namedBindings && k.isNamedImportBindings(e.importClause.namedBindings)) {
            e.importClause.namedBindings.elements.forEach(i => {
              var n = i.name || i.propertyName;
              if (!n) {
                return;
              }
              var s = getDeclarationOfIdentifier(n, r);
              if (s) {
                var o = s.getSourceFile();
                if (o.fileName.includes("node_modules")) {
                  return;
                }
                if (!o) {
                  return;
                }
                var l = getFragmentsInSource(o, r, t).map(e => e.name.value);
                var u = e.moduleSpecifier.getText();
                var c = a[u];
                if (l.length && c) {
                  c.fragments = c.fragments.concat(l);
                } else if (l.length && !c) {
                  a[u] = c = {
                    start: e.moduleSpecifier.getStart(),
                    length: e.moduleSpecifier.getText().length,
                    fragments: l
                  };
                }
              }
            });
          }
        });
      }
      return a;
    })(n, r);
    var S = r.languageService.getProgram()?.getTypeChecker();
    var y = new Set;
    s.forEach(({node: e}) => {
      try {
        var t = x(e.getText().slice(1, -1), {
          noLocation: !0
        });
        A(t, {
          FragmentSpread: e => {
            y.add(e.name.value);
          }
        });
      } catch (e) {}
    });
    var T = function getDirectlyUsedFragments(t, i, r) {
      var a = new Set;
      if (!r) {
        return a;
      }
      var n = new Set;
      i.forEach(({tadaFragmentRefs: e}) => {
        if (e) {
          e.forEach(e => n.add(e));
        }
      });
      var s = new Map;
      var visit = e => {
        if (k.isIdentifier(e)) {
          var t = r.getSymbolAtLocation(e);
          if (t) {
            var i = s.get(t);
            if (i) {
              i.push(e);
            } else {
              s.set(t, [ e ]);
            }
          }
        }
        k.forEachChild(e, visit);
      };
      visit(t);
      var o = [];
      findAllImports(t).forEach(e => {
        var t = e.importClause;
        if (!t) {
          return;
        }
        if (t.name) {
          o.push(t.name);
        }
        var i = t.namedBindings;
        if (i) {
          if (k.isNamespaceImport(i)) {
            o.push(i.name);
          } else {
            i.elements.forEach(e => o.push(e.name));
          }
        }
      });
      o.forEach(t => {
        var i = r.getSymbolAtLocation(t);
        if (!i) {
          return;
        }
        if (!(s.get(i) || []).some(e => e !== t && !n.has(e))) {
          return;
        }
        (function getFragmentNamesForIdentifier(t, i) {
          var r = getValueOfIdentifier(t, i);
          if (!r || !k.isCallExpression(r)) {
            return [];
          }
          var a = r.arguments[0];
          if (!a || !k.isStringLiteralLike(a)) {
            return [];
          }
          try {
            return x(a.getText().slice(1, -1), {
              noLocation: !0
            }).definitions.filter(t => t.kind === e.FRAGMENT_DEFINITION).map(e => e.name.value);
          } catch (e) {
            return [];
          }
        })(t, r).forEach(e => a.add(e));
      });
      return a;
    }(n, s, S);
    Object.keys(m).forEach(e => {
      var {fragments: t, start: i, length: r} = m[e];
      var a = Array.from(new Set(t.filter(e => !y.has(e) && !T.has(e))));
      if (a.length) {
        v.push({
          file: n,
          length: r,
          start: i,
          category: k.DiagnosticCategory.Warning,
          code: 52003,
          messageText: `Unused co-located fragment definition(s) "${a.join(", ")}" in ${e}`
        });
      }
    });
    return [ ...f, ...v ];
  } else {
    return f;
  }
}

var runDiagnostics = (t, {nodes: i, fragments: r}, a, n) => {
  var s = n.config.templateIsCallExpression ?? !0;
  var o = n.languageService.getProgram()?.getTypeChecker();
  var l = i.map(t => {
    var i = t.node;
    if (!s && (k.isNoSubstitutionTemplateLiteral(i) || k.isTemplateExpression(i))) {
      if (k.isTaggedTemplateExpression(i.parent)) {
        i = i.parent;
      } else {
        return;
      }
    }
    var {combinedText: l, resolvedSpans: u} = resolveTemplate(i, 0, n);
    var c = l.split("\n");
    var f = !1;
    if (k.isAsExpression(i.parent)) {
      if (k.isExpressionStatement(i.parent.parent)) {
        f = !0;
      }
    } else if (k.isExpressionStatement(i.parent)) {
      f = !0;
    }
    var h = i.getStart() + (s ? 0 : i.tag.getText().length + (f ? 2 : 0));
    var d = h + i.getText().length;
    var v = [ ...r ];
    if (null != t.tadaFragmentRefs) {
      var g = new Set;
      for (var m of t.tadaFragmentRefs) {
        unrollFragment(m, 0, o).forEach(e => g.add(e.name.value));
      }
      v = v.filter(e => g.has(e.name.value));
    }
    if (s) {
      try {
        var S = x(l, {
          noLocation: !0
        }).definitions.filter(t => t.kind === e.FRAGMENT_DEFINITION);
        v = v.filter(t => !S.some(i => i.kind === e.FRAGMENT_DEFINITION && i.name.value === t.name.value));
      } catch (e) {}
    }
    var y = t.schema && a.multi[t.schema] ? a.multi[t.schema]?.schema : a.current?.schema;
    if (!y) {
      return;
    }
    var T = new Set([ ...se, ...n.config.clientDirectives || [] ]);
    var E = getDiagnostics(l, y, void 0, void 0, v).filter(e => {
      if (!e.message.includes("Unknown directive")) {
        return !0;
      }
      var [t] = e.message.split("(");
      var i = t && /Unknown directive "@([^)]+)"/g.exec(t);
      if (!i) {
        return !0;
      }
      var r = i[1];
      return r && !T.has(r);
    }).map(e => {
      var {start: t, end: i} = e.range;
      var r = h + t.line;
      for (var a = 0; a <= t.line && a < c.length; a++) {
        if (a === t.line) {
          r += t.character;
        } else if (c[a]) {
          r += c[a].length;
        }
      }
      var n = h + i.line;
      for (var s = 0; s <= i.line && s < c.length; s++) {
        if (s === i.line) {
          n += i.character;
        } else if (c[s]) {
          n += c[s].length;
        }
      }
      var o = u.find(e => r >= e.new.start && n <= e.new.start + e.new.length);
      if (o) {
        return {
          ...e,
          start: o.original.start,
          length: o.original.length
        };
      } else if (r > d) {
        var l = u.filter(e => e.new.start + e.new.length < r).reduce((e, t) => e + (t.new.length - t.original.length), 0);
        r -= l;
        n -= l;
        return {
          ...e,
          start: r + 1,
          length: n - r
        };
      } else {
        return {
          ...e,
          start: r + 1,
          length: n - r
        };
      }
    }).filter(e => e.start + e.length <= d);
    return E;
  }).flat().filter(Boolean);
  var u = l.map(e => ({
    file: t,
    length: e.length,
    start: e.start,
    category: 2 === e.severity ? k.DiagnosticCategory.Warning : k.DiagnosticCategory.Error,
    code: "number" == typeof e.code ? e.code : 2 === e.severity ? 52004 : 52001,
    messageText: e.message.split("\n")[0]
  }));
  if (s) {
    var c = ((e, t, i) => {
      var r = [];
      if (!(i.config.trackFieldUsage ?? 1)) {
        return r;
      }
      var a = new Set([ "id", "_id", "__typename", ...i.config.reservedKeys ?? [] ]);
      var n = i.languageService.getProgram()?.getTypeChecker();
      if (!n) {
        return;
      }
      try {
        var s = 0;
        var makeTrieNode = e => ({
          id: s++,
          children: new Map,
          isLeaf: e,
          used: !1
        });
        var o = [];
        var _loop = function() {
          var e = l.getText();
          if (e.includes("mutation") || e.includes("subscription")) {
            return 0;
          }
          var t;
          try {
            t = x(e.slice(1, -1));
          } catch (e) {
            return 0;
          }
          var i = makeTrieNode(!1);
          var r = [];
          var n = new Map;
          var s = new Map;
          var u = [];
          var c = [ i ];
          A(t, {
            Field: {
              enter(e) {
                var t = e.alias ? e.alias.value : e.name.value;
                var i = u.length ? `${u.join(".")}.${t}` : t;
                var o = c[c.length - 1];
                if (!e.selectionSet && !a.has(e.name.value)) {
                  r.push(i);
                  s.set(i, {
                    start: e.name.loc.start,
                    length: e.name.loc.end - e.name.loc.start
                  });
                  var l = o.children.get(t);
                  if (!l) {
                    o.children.set(t, l = makeTrieNode(!0));
                  } else {
                    l.isLeaf = !0;
                  }
                  n.set(i, l);
                } else if (e.selectionSet) {
                  u.push(t);
                  s.set(i, {
                    start: e.name.loc.start,
                    length: e.name.loc.end - e.name.loc.start
                  });
                  var f = o.children.get(t);
                  if (!f) {
                    o.children.set(t, f = makeTrieNode(!1));
                  }
                  c.push(f);
                }
              },
              leave(e) {
                if (e.selectionSet) {
                  u.pop();
                  c.pop();
                }
              }
            }
          });
          o.push({
            root: i,
            allPaths: r,
            leafByPath: n,
            fieldToLoc: s,
            templateNode: l,
            accessCount: 0
          });
        };
        for (var l of t) {
          if (0 === _loop()) {
            continue;
          }
        }
        if (!o.length) {
          return r;
        }
        var u = new Map;
        var c = [];
        var indexWalk = e => {
          if (k.isTypeNode(e)) {
            return;
          }
          if (k.isIdentifier(e)) {
            var t = e.parent;
            var i = t && (k.isVariableDeclaration(t) || k.isBindingElement(t) || k.isParameter(t) || k.isFunctionDeclaration(t) || k.isFunctionExpression(t) || k.isClassDeclaration(t) || k.isImportSpecifier(t) || k.isImportClause(t) || k.isNamespaceImport(t) || k.isExportSpecifier(t)) && t.name === e;
            var r = t && (k.isPropertyAccessExpression(t) && t.name === e || k.isQualifiedName(t) && t.right === e || k.isPropertyAssignment(t) && t.name === e || k.isBindingElement(t) && t.propertyName === e || k.isJsxAttribute(t) && t.name === e || k.isPropertyDeclaration(t) && t.name === e || k.isMethodDeclaration(t) && t.name === e || k.isPropertySignature(t) && t.name === e || k.isMethodSignature(t) && t.name === e || k.isEnumMember(t) && t.name === e);
            if (!i && !r) {
              var a = u.get(e.text);
              if (!a) {
                u.set(e.text, a = []);
              }
              a.push(e);
            }
            return;
          }
          if (k.isVariableDeclaration(e) && e.initializer) {
            var n = e.initializer;
            while (k.isParenthesizedExpression(n) || k.isAwaitExpression(n) || k.isNonNullExpression(n) || k.isAsExpression(n)) {
              n = n.expression;
            }
            if (k.isCallExpression(n)) {
              c.push(e);
            }
          }
          k.forEachChild(e, indexWalk);
        };
        indexWalk(e);
        var f = new Map;
        var getRefSymbol = e => {
          if (f.has(e)) {
            return f.get(e);
          }
          var t = k.isShorthandPropertyAssignment(e.parent) && e.parent.name === e ? n.getShorthandAssignmentValueSymbol(e.parent) : n.getSymbolAtLocation(e);
          f.set(e, t);
          return t;
        };
        var h = [];
        var d = new Map;
        var addEntry = (e, t, i, r) => {
          if (!e) {
            return;
          }
          var a = d.get(e);
          if (!a) {
            d.set(e, a = new Set);
          }
          var n = `${t.id}${r ? "s" : ""}`;
          if (a.has(n)) {
            return;
          }
          a.add(n);
          h.push({
            symbol: e,
            node: t,
            doc: i,
            suppressReturnBail: r
          });
        };
        var trackBinding = (e, t, i, r) => {
          if (k.isIdentifier(e)) {
            addEntry(n.getSymbolAtLocation(e), t, i, r);
          } else if (k.isObjectBindingPattern(e)) {
            for (var a of e.elements) {
              var s = void 0;
              if (a.propertyName) {
                if (k.isIdentifier(a.propertyName) || k.isStringLiteral(a.propertyName)) {
                  s = a.propertyName.text;
                }
              } else if (k.isIdentifier(a.name)) {
                s = a.name.text;
              }
              var o = s && t.children.get(s) || t;
              trackBinding(a.name, o, i, r);
            }
          } else {
            for (var l of e.elements) {
              if (k.isOmittedExpression(l)) {
                continue;
              }
              trackBinding(l.name, t, i, r);
            }
          }
        };
        var markSubtreeUsed = e => {
          for (var t of e.children.values()) {
            t.used = !0;
            markSubtreeUsed(t);
          }
        };
        var handleArrayChain = (e, t, i, r) => {
          var a = e;
          while (1) {
            var s = a.expression.name.text;
            var o = a.arguments[0];
            if (o && k.isIdentifier(o)) {
              var l = getValueOfIdentifier(o, n);
              if (l && (k.isFunctionDeclaration(l) || k.isFunctionExpression(l) || k.isArrowFunction(l))) {
                o = l;
              }
            }
            if (o && (k.isFunctionDeclaration(o) || k.isFunctionExpression(o) || k.isArrowFunction(o))) {
              var u = o.parameters["reduce" === s ? 1 : 0];
              if (u) {
                trackBinding(u.name, i, r, !0);
              }
            }
            var c = a.parent;
            if (k.isPropertyAccessExpression(c) && c.expression === a && ne.has(c.name.text) && k.isCallExpression(c.parent) && c.parent.expression === c) {
              a = c.parent;
              continue;
            }
            break;
          }
          if (k.isVariableDeclaration(e.parent) && "some" !== t && "every" !== t) {
            trackBinding(e.parent.name, i, r, !0);
          }
        };
        var walkUseChain = (e, t) => {
          var i = t.doc;
          var r = e;
          var a = t.node;
          var s = !0;
          while (1) {
            var o = r.parent;
            if (!o) {
              break;
            }
            if (k.isPropertyAccessExpression(o)) {
              if (o.expression !== r) {
                break;
              }
              var l = o.name.text;
              var u = o.parent;
              if (k.isCallExpression(u) && u.expression === o) {
                if ("at" === l) {
                  r = u;
                  continue;
                } else if (ne.has(l)) {
                  handleArrayChain(u, l, a, i);
                  i.accessCount++;
                  return;
                }
              }
              a = a.children.get(l) || a;
              r = o;
              continue;
            }
            if (k.isElementAccessExpression(o)) {
              if (o.expression !== r) {
                break;
              }
              if (k.isStringLiteral(o.argumentExpression)) {
                a = a.children.get(o.argumentExpression.text) || a;
              }
              r = o;
              continue;
            }
            if (k.isNonNullExpression(o) || k.isParenthesizedExpression(o) || k.isAsExpression(o) || k.isAwaitExpression(o) || k.isSatisfiesExpression && k.isSatisfiesExpression(o)) {
              r = o;
              continue;
            }
            if (k.isBinaryExpression(o)) {
              if (o.operatorToken.kind === k.SyntaxKind.EqualsToken) {
                if (o.left === r) {
                  return;
                }
                if (k.isIdentifier(o.left)) {
                  addEntry(n.getSymbolAtLocation(o.left), a, i, !1);
                  i.accessCount++;
                  return;
                }
                if (s) {
                  if (a.isLeaf) {
                    a.used = !0;
                  }
                  markSubtreeUsed(a);
                  i.accessCount++;
                  return;
                }
                break;
              }
              if (o.operatorToken.kind !== k.SyntaxKind.BarBarToken && o.operatorToken.kind !== k.SyntaxKind.QuestionQuestionToken) {
                s = !1;
              }
              r = o;
              continue;
            }
            if (k.isConditionalExpression(o)) {
              if (o.condition === r) {
                break;
              }
              s = !1;
              r = o;
              continue;
            }
            if (k.isVariableDeclaration(o)) {
              trackBinding(o.name, a, i, !1);
              return;
            }
            if (k.isForOfStatement(o) && o.expression === r) {
              if (k.isVariableDeclarationList(o.initializer)) {
                for (var c of o.initializer.declarations) {
                  trackBinding(c.name, a, i, !1);
                }
              } else if (k.isIdentifier(o.initializer)) {
                addEntry(n.getSymbolAtLocation(o.initializer), a, i, !1);
              }
              i.accessCount++;
              return;
            }
            if (k.isReturnStatement(o) || k.isArrowFunction(o) && o.body === r) {
              if (a.isLeaf) {
                a.used = !0;
              }
              if (!t.suppressReturnBail) {
                markSubtreeUsed(a);
              }
              i.accessCount++;
              return;
            }
            if (k.isCallExpression(o) && o.expression !== r) {
              if (a.isLeaf) {
                a.used = !0;
              }
              if (s) {
                markSubtreeUsed(a);
              }
              i.accessCount++;
              return;
            }
            if (s && (k.isSpreadElement(o) || k.isSpreadAssignment(o) || k.isJsxSpreadAttribute(o) || k.isShorthandPropertyAssignment(o) || k.isPropertyAssignment(o) && o.initializer === r)) {
              if (a.isLeaf) {
                a.used = !0;
              }
              markSubtreeUsed(a);
              i.accessCount++;
              return;
            }
            break;
          }
          if (a.isLeaf) {
            a.used = !0;
          }
          i.accessCount++;
        };
        var v = new Map;
        for (var g of o) {
          var m = g.templateNode;
          var S = void 0;
          var y = m.parent.parent;
          while (k.isParenthesizedExpression(y) || k.isAsExpression(y) || k.isNonNullExpression(y) || k.isSatisfiesExpression && k.isSatisfiesExpression(y)) {
            y = y.parent;
          }
          if (k.isVariableDeclaration(y)) {
            S = y;
          }
          if (S && k.isIdentifier(S.name)) {
            var T = n.getSymbolAtLocation(S.name);
            if (T) {
              var E = u.get(S.name.text) || [];
              for (var b of E) {
                if (getRefSymbol(b) !== T) {
                  continue;
                }
                var C = getVariableDeclaration(b);
                if (!C || C === S) {
                  continue;
                }
                trackBinding(C.name, g.root, g, !1);
              }
            }
          } else if (!S) {
            var w = getVariableDeclaration(m);
            if (w && k.isIdentifier(w.name)) {
              addEntry(n.getSymbolAtLocation(w.name), g.root, g, !1);
            }
          }
          var D = resolveDataType(m, n);
          if (D) {
            for (var L of c) {
              if (L === S) {
                continue;
              }
              var F = v.get(L);
              if (!F) {
                var N = unwrapAbstractType(n.getTypeAtLocation(L.initializer));
                F = {
                  type: N
                };
                if (N.flags & k.TypeFlags.Object) {
                  var _ = N.getProperty("0");
                  if (_) {
                    F.tupleType = n.getTypeOfSymbol(_);
                  }
                  var I = (F.tupleType || N).getProperty("data");
                  if (I) {
                    F.dataType = unwrapAbstractType(n.getTypeOfSymbol(I));
                  }
                }
                v.set(L, F);
              }
              if (F.type === D || F.tupleType === D || F.dataType === D) {
                trackBinding(L.name, g.root, g, !1);
              }
            }
          }
        }
        for (var O = 0; O < h.length; O++) {
          var z = h[O];
          var P = u.get(z.symbol.name);
          if (!P) {
            continue;
          }
          for (var V of P) {
            if (getRefSymbol(V) !== z.symbol) {
              continue;
            }
            walkUseChain(V, z);
          }
        }
        var _loop2 = function(t) {
          if (!t.accessCount) {
            return 1;
          }
          var i = t.templateNode;
          var a = t.fieldToLoc;
          var n = t.allPaths.filter(e => !t.leafByPath.get(e).used);
          var s = new Set;
          var o = {};
          var l = new Set;
          n.forEach(e => {
            var t = e.split(".");
            t.pop();
            var i = t.join(".");
            if (a.get(i)) {
              s.add(i);
              if (o[i]) {
                o[i].add(e);
              } else {
                o[i] = new Set([ e ]);
              }
            } else {
              l.add(e);
            }
          });
          s.forEach(t => {
            var n = a.get(t);
            var s = o[t];
            r.push({
              file: e,
              length: n.length,
              start: i.getStart() + n.start + 1,
              category: k.DiagnosticCategory.Warning,
              code: ae,
              messageText: `Field(s) ${[ ...s ].map(e => `'${e}'`).join(", ")} are not used.`
            });
          });
          l.forEach(t => {
            var n = a.get(t);
            r.push({
              file: e,
              length: n.length,
              start: i.getStart() + n.start + 1,
              category: k.DiagnosticCategory.Warning,
              code: ae,
              messageText: `Field ${t} is not used.`
            });
          });
        };
        for (var R of o) {
          if (_loop2(R)) {
            continue;
          }
        }
      } catch (e) {
        console.error("[GraphQLSP]: ", e.message, e.stack);
      }
      return r;
    })(t, i.map(e => e.node), n) || [];
    if (!c) {
      return u;
    }
    return [ ...u, ...c ];
  } else {
    return u;
  }
};

export { fe as A, CharacterStream as C, bubbleUpTemplate as a, bubbleUpCallExpression as b, getSchemaName as c, getAllFragments as d, isGraphQLTag as e, findNode as f, getSource as g, init as h, isGraphQLCall as i, U as j, getGraphQLDiagnostics as k, getPersistedCodeFixAtPosition as l, findAllPersistedCallExpressions as m, findAllCallExpressions as n, onlineParser as o, print as p, getDocumentReferenceFromTypeQuery as q, resolveTemplate as r, getDocumentReferenceFromDocumentNode as s, k as t, unrollTadaFragments as u };
